// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/connector/private-key-provider.connector.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateKeyProviderConnector = void 0;
var ethers = require("ethers6");
var PrivateKeyProviderConnector = /** @class */ (function () {
    function PrivateKeyProviderConnector(privateKey, provider, ethersSigner) {
        this.privateKey = privateKey;
        this.provider = provider;
        this.isEthers = provider instanceof ethers.JsonRpcProvider;
        if (this.isEthers) {
            this.ethersSigner = ethersSigner;
        }
    }
    PrivateKeyProviderConnector.prototype.contractEncodeABI = function (abi, address, methodName, methodParams) {
        var _a;
        if (this.isEthers && this.ethersSigner) {
            var contract = new ethers.Contract(address || ethers.ZeroAddress, abi, this.ethersSigner);
            return contract.interface.encodeFunctionData(methodName, methodParams);
        }
        else {
            var contract = new this.provider.eth.Contract(abi, address === null ? undefined : address);
            return (_a = contract.methods)[methodName].apply(_a, methodParams).encodeABI();
        }
    };
    PrivateKeyProviderConnector.prototype.signTypedData = function (_walletAddress_1, typedData_1) {
        return __awaiter(this, arguments, void 0, function (_walletAddress, typedData, _typedDataHash) {
            var domain, types, message, EIP712Domain, restTypes, signer, domain, types, message, EIP712Domain, restTypes, signature;
            if (_typedDataHash === void 0) { _typedDataHash = ''; }
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.isEthers && this.ethersSigner)) return [3 /*break*/, 2];
                        domain = typedData.domain, types = typedData.types, message = typedData.message;
                        EIP712Domain = types.EIP712Domain, restTypes = __rest(types, ["EIP712Domain"]);
                        return [4 /*yield*/, this.ethersSigner.signTypedData(domain, restTypes, message)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        signer = new ethers.Wallet(this.privateKey);
                        domain = typedData.domain, types = typedData.types, message = typedData.message;
                        EIP712Domain = types.EIP712Domain, restTypes = __rest(types, ["EIP712Domain"]);
                        return [4 /*yield*/, signer.signTypedData(domain, restTypes, message)];
                    case 3:
                        signature = _a.sent();
                        return [2 /*return*/, signature];
                }
            });
        });
    };
    PrivateKeyProviderConnector.prototype.ethCall = function (contractAddress, callData) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.isEthers) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.provider.call({
                                to: contractAddress,
                                data: callData,
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2: return [2 /*return*/, this.provider.eth.call({
                            to: contractAddress,
                            data: callData,
                        })];
                }
            });
        });
    };
    PrivateKeyProviderConnector.prototype.decodeABIParameter = function (type, hex) {
        if (this.isEthers) {
            // For ethers v6, we use Interface to decode parameters
            var iface = new ethers.Interface(["function dummy(".concat(type, ")")]);
            try {
                return iface.decodeFunctionResult("dummy", hex)[0];
            }
            catch (error) {
                // Fallback to ethers.utils.defaultAbiCoder if Interface fails
                return ethers.AbiCoder.defaultAbiCoder().decode([type], hex)[0];
            }
        }
        else {
            return this.provider.eth.abi.decodeParameter(type, hex);
        }
    };
    return PrivateKeyProviderConnector;
}());
exports.PrivateKeyProviderConnector = PrivateKeyProviderConnector;
