// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/connector/web3-provider.connector.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3ProviderConnector = void 0;
var ethers = require("ethers6");
var Web3ProviderConnector = /** @class */ (function () {
    function Web3ProviderConnector(provider, signer) {
        this.provider = provider;
        this.signer = signer;
        this.providerType = this.detectProviderType(provider);
        console.log('providerType', this.providerType);
        if (this.providerType === 'ethers-provider') {
            var ethersVersion = getEthersVersion(this.provider);
            console.log('ethersVersion', ethersVersion);
        }
    }
    Web3ProviderConnector.prototype.getSigner = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.signer) {
                            return [2 /*return*/, this.signer];
                        }
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.provider.getSigner(address)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        if (this.providerType === 'web3-browser' || this.providerType === 'web3-private-key') {
                            return [2 /*return*/, this.provider];
                        }
                        _a.label = 3;
                    case 3: throw new Error('Unsupported provider type');
                }
            });
        });
    };
    Web3ProviderConnector.prototype.contractEncodeABI = function (abi, address, methodName, methodParams) {
        var _a;
        if (this.providerType === 'ethers-provider') {
            var contract = new ethers.Contract(address || ethers.ZeroAddress, abi, this.provider);
            return contract.interface.encodeFunctionData(methodName, methodParams);
        }
        else if (this.providerType === 'web3-browser' || this.providerType === 'web3-private-key') {
            var contract = new this.provider.eth.Contract(abi, address === null ? undefined : address);
            return (_a = contract.methods)[methodName].apply(_a, methodParams).encodeABI();
        }
    };
    Web3ProviderConnector.prototype.signTypedData = function (walletAddress, typedData) {
        return __awaiter(this, void 0, void 0, function () {
            var signer, domain, types, message, EIP712Domain, restTypes, walletAccount, signer, domain, types, message, EIP712Domain, restTypes, signature, extendedWeb3, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 10, , 11]);
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.getSigner(walletAddress)];
                    case 1:
                        signer = _a.sent();
                        domain = typedData.domain, types = typedData.types, message = typedData.message;
                        EIP712Domain = types.EIP712Domain, restTypes = __rest(types, ["EIP712Domain"]);
                        if (!signer._signTypedData) return [3 /*break*/, 3];
                        return [4 /*yield*/, signer._signTypedData(domain, restTypes, message)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3: return [4 /*yield*/, signer.signTypedData(domain, restTypes, message)];
                    case 4: return [2 /*return*/, _a.sent()];
                    case 5: return [3 /*break*/, 9];
                    case 6:
                        if (!(this.providerType === 'web3-private-key')) return [3 /*break*/, 8];
                        walletAccount = this.provider.eth.accounts.wallet[walletAddress];
                        if (!walletAccount) {
                            throw new Error('Wallet account not found');
                        }
                        signer = new ethers.Wallet(walletAccount.privateKey);
                        domain = typedData.domain, types = typedData.types, message = typedData.message;
                        EIP712Domain = types.EIP712Domain, restTypes = __rest(types, ["EIP712Domain"]);
                        return [4 /*yield*/, signer.signTypedData(domain, restTypes, message)];
                    case 7:
                        signature = _a.sent();
                        return [2 /*return*/, signature];
                    case 8:
                        if (this.providerType === 'web3-browser') {
                            extendedWeb3 = this.provider.extend({
                                methods: [
                                    {
                                        name: 'signTypedDataV4',
                                        call: 'eth_signTypedData_v4',
                                        params: 2,
                                    },
                                ],
                            });
                            return [2 /*return*/, extendedWeb3.signTypedDataV4(walletAddress, JSON.stringify(typedData))];
                        }
                        _a.label = 9;
                    case 9: throw new Error('Unsupported provider type for signing');
                    case 10:
                        error_1 = _a.sent();
                        console.error('Sign typed data error:', error_1);
                        throw error_1;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    Web3ProviderConnector.prototype.signMessage = function (walletAddress, message) {
        return __awaiter(this, void 0, void 0, function () {
            var signer, walletAccount, signer, extendedWeb3, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 10, , 11]);
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.getSigner(walletAddress)];
                    case 1:
                        signer = _a.sent();
                        if (!signer.signMessage) return [3 /*break*/, 3];
                        return [4 /*yield*/, signer.signMessage(message)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3: throw new Error('Signer does not support signMessage');
                    case 4: return [3 /*break*/, 9];
                    case 5:
                        if (!(this.providerType === 'web3-private-key')) return [3 /*break*/, 7];
                        walletAccount = this.provider.eth.accounts.wallet[walletAddress];
                        if (!walletAccount) {
                            throw new Error('Wallet account not found');
                        }
                        signer = new ethers.Wallet(walletAccount.privateKey);
                        return [4 /*yield*/, signer.signMessage(message)];
                    case 6: return [2 /*return*/, _a.sent()];
                    case 7:
                        if (!(this.providerType === 'web3-browser')) return [3 /*break*/, 9];
                        extendedWeb3 = this.provider.extend({
                            methods: [
                                {
                                    name: 'personalSign',
                                    call: 'personal_sign',
                                    params: 2,
                                },
                            ],
                        });
                        return [4 /*yield*/, extendedWeb3.personalSign(message, walletAddress)];
                    case 8: return [2 /*return*/, _a.sent()];
                    case 9: throw new Error('Unsupported provider type for signing');
                    case 10:
                        error_2 = _a.sent();
                        console.error('Sign message error:', error_2);
                        throw error_2;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    Web3ProviderConnector.prototype.ethCall = function (contractAddress, callData) {
        return __awaiter(this, void 0, void 0, function () {
            var error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.provider.call({
                                to: contractAddress,
                                data: callData,
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        if (!(this.providerType === 'web3-browser' || this.providerType === 'web3-private-key')) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.provider.eth.call({
                                to: contractAddress,
                                data: callData,
                            })];
                    case 3: return [2 /*return*/, _a.sent()];
                    case 4: throw new Error('Unsupported provider type for eth_call');
                    case 5:
                        error_3 = _a.sent();
                        console.error('Eth call error:', error_3);
                        throw error_3;
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    Web3ProviderConnector.prototype.decodeABIParameter = function (type, hex) {
        try {
            if (this.providerType === 'ethers-provider') {
                var iface = new ethers.Interface(["function dummy(".concat(type, ")")]);
                try {
                    return iface.decodeFunctionResult("dummy", hex)[0];
                }
                catch (error) {
                    return ethers.AbiCoder.defaultAbiCoder().decode([type], hex)[0];
                }
            }
            else if (this.providerType === 'web3-browser' || this.providerType === 'web3-private-key') {
                return this.provider.eth.abi.decodeParameter(type, hex);
            }
            throw new Error('Unsupported provider type for decoding');
        }
        catch (error) {
            console.error('Decode parameter error:', error);
            throw error;
        }
    };
    Web3ProviderConnector.prototype.waitForConfirmation = function (txHash_1) {
        return __awaiter(this, arguments, void 0, function (txHash, confirmations, interval) {
            var receipt;
            if (confirmations === void 0) { confirmations = 1; }
            if (interval === void 0) { interval = 1500; }
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        receipt = null;
                        _a.label = 1;
                    case 1:
                        if (!!receipt) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.provider.eth.getTransactionReceipt(txHash)];
                    case 2:
                        receipt = _a.sent();
                        if (receipt && receipt.confirmations >= confirmations) {
                            return [2 /*return*/, receipt];
                        }
                        return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(resolve, interval); })];
                    case 3:
                        _a.sent();
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Web3ProviderConnector.prototype.sendTransaction = function (transaction) {
        return __awaiter(this, void 0, void 0, function () {
            var gasPrice, feeData, signer, gasLimit, gas, signer, tx, error_4;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 17, , 18]);
                        if (!!transaction.gasPrice) return [3 /*break*/, 5];
                        gasPrice = 0;
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.provider.getFeeData()];
                    case 1:
                        feeData = _a.sent();
                        gasPrice = feeData.gasPrice ? Number(feeData.gasPrice) : 0;
                        return [3 /*break*/, 4];
                    case 2:
                        if (!(this.providerType === 'web3-private-key' || this.providerType === 'web3-browser')) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.provider.eth.getGasPrice()];
                    case 3:
                        gasPrice = _a.sent();
                        _a.label = 4;
                    case 4:
                        transaction.gasPrice = parseInt(gasPrice * 1.2);
                        _a.label = 5;
                    case 5:
                        if (!(!transaction.gas && !transaction.gasLimit)) return [3 /*break*/, 11];
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 8];
                        return [4 /*yield*/, this.getSigner(transaction.from)];
                    case 6:
                        signer = _a.sent();
                        return [4 /*yield*/, signer.estimateGas(__assign({}, transaction))];
                    case 7:
                        gasLimit = _a.sent();
                        transaction.gasLimit = Number(gasLimit);
                        return [3 /*break*/, 11];
                    case 8:
                        if (!(this.providerType === 'web3-private-key' || this.providerType === 'web3-browser')) return [3 /*break*/, 10];
                        return [4 /*yield*/, this.provider.eth.estimateGas(__assign({}, transaction))];
                    case 9:
                        gas = _a.sent();
                        transaction.gas = parseInt(gas * 1.2);
                        return [3 /*break*/, 11];
                    case 10:
                        if (this.provider.request) {
                            // transaction.gas = '0x5208';
                        }
                        _a.label = 11;
                    case 11:
                        if (!(this.providerType === 'ethers-provider')) return [3 /*break*/, 15];
                        return [4 /*yield*/, this.getSigner(transaction.from)];
                    case 12:
                        signer = _a.sent();
                        return [4 /*yield*/, signer.sendTransaction(transaction)];
                    case 13:
                        tx = _a.sent();
                        return [4 /*yield*/, tx.wait()];
                    case 14: return [2 /*return*/, _a.sent()];
                    case 15:
                        if (this.providerType === 'web3-private-key' || this.providerType === 'web3-browser') {
                            return [2 /*return*/, new Promise(function (resolve, reject) {
                                    _this.provider.eth.sendTransaction(transaction)
                                        .on('receipt', resolve)
                                        .on('error', reject);
                                })];
                        }
                        _a.label = 16;
                    case 16: throw new Error('Unsupported provider type for sending transaction');
                    case 17:
                        error_4 = _a.sent();
                        console.error('Send transaction error:', error_4);
                        throw error_4;
                    case 18: return [2 /*return*/];
                }
            });
        });
    };
    Web3ProviderConnector.prototype.detectProviderType = function (providerOrWrapper) {
        if (providerOrWrapper instanceof ethers.JsonRpcProvider || typeof providerOrWrapper.send === 'function') {
            return 'ethers-provider';
        }
        if (providerOrWrapper.eth) {
            var accounts = providerOrWrapper.eth.accounts.wallet;
            if (accounts && accounts.length > 0) {
                return 'web3-private-key';
            }
            return 'web3-browser';
        }
        return 'unknown';
    };
    Web3ProviderConnector.prototype.sanitizeTransaction = function (tx) {
        var bigFields = ['gasPrice', 'value', 'gasLimit', 'maxFeePerGas', 'maxPriorityFeePerGas'];
        for (var _i = 0, bigFields_1 = bigFields; _i < bigFields_1.length; _i++) {
            var key = bigFields_1[_i];
            if (tx[key] !== undefined) {
                if (typeof tx[key] === 'number') {
                    tx[key] = "0x".concat(BigInt(tx[key]).toString(16));
                }
                else if (typeof tx[key] === 'string' && /^\d+$/.test(tx[key])) {
                    tx[key] = "0x".concat(BigInt(tx[key]).toString(16));
                }
            }
        }
        return tx;
    };
    return Web3ProviderConnector;
}());
exports.Web3ProviderConnector = Web3ProviderConnector;
function getEthersVersion(provider) {
    if (provider._isProvider) {
        return 'isV5';
    }
    else {
        return 'isV6';
    }
}
