// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/connector/web3-provider.connector copy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3ProviderConnector = void 0;
var Web3ProviderConnector = /** @class */ (function () {
    function Web3ProviderConnector(web3Provider) {
        this.web3Provider = web3Provider;
    }
    Web3ProviderConnector.prototype.contractEncodeABI = function (abi, address, methodName, methodParams) {
        var _a;
        var contract = new this.web3Provider.eth.Contract(abi, address === null ? undefined : address);
        return (_a = contract.methods)[methodName].apply(_a, methodParams).encodeABI();
    };
    Web3ProviderConnector.prototype.signTypedData = function (walletAddress, typedData, 
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    _typedDataHash) {
        var extendedWeb3 = this.web3Provider.extend({
            methods: [
                {
                    name: 'signTypedDataV4',
                    call: 'eth_signTypedData_v4',
                    params: 2,
                },
            ],
        });
        return extendedWeb3.signTypedDataV4(walletAddress, JSON.stringify(typedData));
    };
    Web3ProviderConnector.prototype.ethCall = function (contractAddress, callData) {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    };
    Web3ProviderConnector.prototype.decodeABIParameter = function (type, hex) {
        return this.web3Provider.eth.abi.decodeParameter(type, hex);
    };
    return Web3ProviderConnector;
}());
exports.Web3ProviderConnector = Web3ProviderConnector;
