// @cowprotocol/sdk-common/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  AbstractProviderAdapter: () => AbstractProviderAdapter,
  AbstractSigner: () => AbstractSigner,
  AdapterContext: () => AdapterContext,
  AdapterUtils: () => AdapterUtils,
  ContractFactory: () => ContractFactory,
  CowError: () => CowError,
  ERC20_ALLOWANCE_ABI: () => ERC20_ALLOWANCE_ABI,
  ERC20_APPROVE_ABI: () => ERC20_APPROVE_ABI,
  EthFlowAbi: () => EthFlowAbi,
  GPV2SettlementAbi: () => GPV2SettlementAbi,
  MAX_UINT256: () => MAX_UINT256,
  MAX_UINT32: () => MAX_UINT32,
  ONE: () => ONE,
  TTLCache: () => TTLCache,
  ZERO: () => ZERO,
  ZERO_ADDRESS: () => ZERO_ADDRESS,
  ZERO_HASH: () => ZERO_HASH,
  applyPercentage: () => applyPercentage,
  areAddressesEqual: () => areAddressesEqual,
  areTokensEqual: () => areTokensEqual,
  bpsToPercentage: () => bpsToPercentage,
  enableLogging: () => enableLogging,
  getAddressKey: () => getAddressKey,
  getBtcAddressKey: () => getBtcAddressKey,
  getEvmAddressKey: () => getEvmAddressKey,
  getGlobalAdapter: () => getGlobalAdapter,
  getSolAddressKey: () => getSolAddressKey,
  getTokenId: () => getTokenId,
  isBtcAddress: () => isBtcAddress,
  isCoWSettlementContract: () => isCoWSettlementContract,
  isCoWVaultRelayerContract: () => isCoWVaultRelayerContract,
  isEvmAddress: () => isEvmAddress,
  isNativeToken: () => isNativeToken,
  isSolanaAddress: () => isSolanaAddress,
  isValidPrivateKey: () => isValidPrivateKey,
  isWrappedNativeToken: () => isWrappedNativeToken,
  jsonWithBigintReplacer: () => jsonWithBigintReplacer,
  log: () => log,
  normalizePrivateKey: () => normalizePrivateKey,
  percentageToBps: () => percentageToBps,
  setGlobalAdapter: () => setGlobalAdapter
});
module.exports = __toCommonJS(src_exports);

// src/adapters/AbstractProviderAdapter.ts
var AbstractProviderAdapter = class {
  ZERO_ADDRESS;
};

// src/adapters/types/AdapterUtils.ts
var AdapterUtils = class {
};

// src/adapters/types/constants.ts
var MAX_UINT256 = BigInt("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");

// src/abi/EthFlowAbi.ts
var EthFlowAbi = [
  {
    inputs: [
      {
        components: [
          {
            internalType: "contract IERC20",
            name: "buyToken",
            type: "address"
          },
          {
            internalType: "address",
            name: "receiver",
            type: "address"
          },
          {
            internalType: "uint256",
            name: "sellAmount",
            type: "uint256"
          },
          {
            internalType: "uint256",
            name: "buyAmount",
            type: "uint256"
          },
          {
            internalType: "bytes32",
            name: "appData",
            type: "bytes32"
          },
          {
            internalType: "uint256",
            name: "feeAmount",
            type: "uint256"
          },
          {
            internalType: "uint32",
            name: "validTo",
            type: "uint32"
          },
          {
            internalType: "bool",
            name: "partiallyFillable",
            type: "bool"
          },
          {
            internalType: "int64",
            name: "quoteId",
            type: "int64"
          }
        ],
        internalType: "struct EthFlowOrder.Data",
        name: "order",
        type: "tuple"
      }
    ],
    name: "createOrder",
    outputs: [
      {
        internalType: "bytes32",
        name: "orderHash",
        type: "bytes32"
      }
    ],
    stateMutability: "payable",
    type: "function"
  },
  {
    inputs: [
      {
        components: [
          {
            internalType: "contract IERC20",
            name: "buyToken",
            type: "address"
          },
          {
            internalType: "address",
            name: "receiver",
            type: "address"
          },
          {
            internalType: "uint256",
            name: "sellAmount",
            type: "uint256"
          },
          {
            internalType: "uint256",
            name: "buyAmount",
            type: "uint256"
          },
          {
            internalType: "bytes32",
            name: "appData",
            type: "bytes32"
          },
          {
            internalType: "uint256",
            name: "feeAmount",
            type: "uint256"
          },
          {
            internalType: "uint32",
            name: "validTo",
            type: "uint32"
          },
          {
            internalType: "bool",
            name: "partiallyFillable",
            type: "bool"
          },
          {
            internalType: "int64",
            name: "quoteId",
            type: "int64"
          }
        ],
        internalType: "struct EthFlowOrder.Data",
        name: "order",
        type: "tuple"
      }
    ],
    name: "invalidateOrder",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "",
        type: "bytes32"
      }
    ],
    name: "orders",
    outputs: [
      {
        internalType: "address",
        name: "owner",
        type: "address"
      },
      {
        internalType: "uint32",
        name: "validTo",
        type: "uint32"
      }
    ],
    stateMutability: "view",
    type: "function"
  }
];

// src/abi/GPV2SettlementAbi.ts
var GPV2SettlementAbi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "owner",
        type: "address"
      },
      {
        indexed: false,
        internalType: "contract IERC20",
        name: "sellToken",
        type: "address"
      },
      {
        indexed: false,
        internalType: "contract IERC20",
        name: "buyToken",
        type: "address"
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "sellAmount",
        type: "uint256"
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "buyAmount",
        type: "uint256"
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "feeAmount",
        type: "uint256"
      },
      {
        indexed: false,
        internalType: "bytes",
        name: "orderUid",
        type: "bytes"
      }
    ],
    name: "Trade",
    type: "event"
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "orderUid",
        type: "bytes"
      },
      {
        internalType: "bool",
        name: "signed",
        type: "bool"
      }
    ],
    name: "setPreSignature",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "orderUid",
        type: "bytes"
      }
    ],
    name: "invalidateOrder",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [],
    name: "domainSeparator",
    outputs: [{ name: "", type: "bytes32" }],
    stateMutability: "nonpayable",
    type: "function"
  }
];

// src/abi/Erc20.ts
var ERC20_APPROVE_ABI = [
  {
    constant: false,
    inputs: [
      {
        name: "_spender",
        type: "address"
      },
      {
        name: "_value",
        type: "uint256"
      }
    ],
    name: "approve",
    outputs: [
      {
        name: "",
        type: "bool"
      }
    ],
    payable: false,
    stateMutability: "nonpayable",
    type: "function"
  }
];
var ERC20_ALLOWANCE_ABI = [
  {
    constant: true,
    inputs: [
      {
        name: "_owner",
        type: "address"
      },
      {
        name: "_spender",
        type: "address"
      }
    ],
    name: "allowance",
    outputs: [
      {
        name: "",
        type: "uint256"
      }
    ],
    payable: false,
    stateMutability: "view",
    type: "function"
  }
];

// src/adapters/context.ts
var AdapterContext = class _AdapterContext {
  static _instance;
  _adapter;
  constructor() {
  }
  static getInstance() {
    if (!_AdapterContext._instance) {
      _AdapterContext._instance = new _AdapterContext();
    }
    return _AdapterContext._instance;
  }
  setAdapter(adapter) {
    this._adapter = adapter;
  }
  getAdapter() {
    if (!this._adapter) {
      throw new Error(
        "Provider adapter is not configurated. Configure with CowSdk or using AdapterContext.getInstance().setAdapter()"
      );
    }
    return this._adapter;
  }
};
function getGlobalAdapter() {
  return AdapterContext.getInstance().getAdapter();
}
function setGlobalAdapter(adapter) {
  const instance = AdapterContext.getInstance();
  instance.setAdapter(adapter);
}

// src/adapters/types/ContractFactory.ts
var ContractFactory = class {
  static createEthFlowContract(address, signer) {
    const adapter = getGlobalAdapter();
    return {
      address,
      estimateGas: {
        createOrder: async (order, options) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(EthFlowAbi, "createOrder", [order]),
            value: options?.value || "0"
          };
          return await signer.estimateGas(txParams);
        },
        invalidateOrder: async (order) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(EthFlowAbi, "invalidateOrder", [order])
          };
          return await signer.estimateGas(txParams);
        },
        orders: async (orderHash) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(EthFlowAbi, "orders", [orderHash])
          };
          return await signer.estimateGas(txParams);
        }
      },
      interface: {
        encodeFunctionData: (functionName, args) => {
          return adapter.utils.encodeFunction(EthFlowAbi, functionName, args);
        },
        decodeFunctionData: (functionName, data) => {
          return adapter.utils.decodeFunctionData(EthFlowAbi, functionName, data);
        },
        parseLog(event) {
          return adapter.utils.createInterface(EthFlowAbi).parseLog(event);
        },
        getEventTopic(eventFragment) {
          return adapter.utils.createInterface(EthFlowAbi).getEventTopic(eventFragment);
        }
      },
      functions: {
        createOrder: async (order, options) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(EthFlowAbi, "createOrder", [order]),
            value: options?.value || "0"
          };
          return await signer.sendTransaction(txParams);
        },
        invalidateOrder: async (order) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(EthFlowAbi, "invalidateOrder", [order])
          };
          return await signer.sendTransaction(txParams);
        },
        orders: async (orderHash) => {
          return await adapter.readContract({
            address,
            abi: EthFlowAbi,
            functionName: "orders",
            args: [orderHash]
          });
        }
      },
      createOrder: async (order, options) => {
        const txParams = {
          to: address,
          data: adapter.utils.encodeFunction(EthFlowAbi, "createOrder", [order]),
          value: options?.value || "0"
        };
        return await signer.sendTransaction(txParams);
      },
      invalidateOrder: async (order) => {
        const txParams = {
          to: address,
          data: adapter.utils.encodeFunction(EthFlowAbi, "invalidateOrder", [order])
        };
        return await signer.sendTransaction(txParams);
      },
      orders: async (orderHash) => {
        return await adapter.readContract({
          address,
          abi: EthFlowAbi,
          functionName: "orders",
          args: [orderHash]
        });
      }
    };
  }
  static createSettlementContract(address, signer) {
    const adapter = getGlobalAdapter();
    return {
      address,
      estimateGas: {
        setPreSignature: async (orderUid, signed) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(GPV2SettlementAbi, "setPreSignature", [orderUid, signed])
          };
          return await signer.estimateGas(txParams);
        },
        invalidateOrder: async (orderUid) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(GPV2SettlementAbi, "invalidateOrder", [orderUid])
          };
          return await signer.estimateGas(txParams);
        },
        domainSeparator: async () => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(GPV2SettlementAbi, "domainSeparator", [])
          };
          return await signer.estimateGas(txParams);
        }
      },
      interface: {
        encodeFunctionData: (functionName, args) => {
          return adapter.utils.encodeFunction(GPV2SettlementAbi, functionName, args);
        },
        decodeFunctionData: (functionName, data) => {
          return adapter.utils.decodeFunctionData(GPV2SettlementAbi, functionName, data);
        },
        parseLog(event) {
          return adapter.utils.createInterface(GPV2SettlementAbi).parseLog(event);
        },
        getEventTopic(eventFragment) {
          return adapter.utils.createInterface(GPV2SettlementAbi).getEventTopic(eventFragment);
        }
      },
      functions: {
        setPreSignature: async (orderUid, signed) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(GPV2SettlementAbi, "setPreSignature", [orderUid, signed])
          };
          return await signer.sendTransaction(txParams);
        },
        invalidateOrder: async (orderUid) => {
          const txParams = {
            to: address,
            data: adapter.utils.encodeFunction(GPV2SettlementAbi, "invalidateOrder", [orderUid])
          };
          return await signer.sendTransaction(txParams);
        },
        domainSeparator: async () => {
          return await adapter.readContract({
            address,
            abi: GPV2SettlementAbi,
            functionName: "domainSeparator",
            args: []
          });
        }
      },
      setPreSignature: async (orderUid, signed) => {
        const txParams = {
          to: address,
          data: adapter.utils.encodeFunction(GPV2SettlementAbi, "setPreSignature", [orderUid, signed])
        };
        return await signer.sendTransaction(txParams);
      },
      invalidateOrder: async (orderUid) => {
        const txParams = {
          to: address,
          data: adapter.utils.encodeFunction(GPV2SettlementAbi, "invalidateOrder", [orderUid])
        };
        return await signer.sendTransaction(txParams);
      },
      domainSeparator: async () => {
        return await adapter.readContract({
          address,
          abi: GPV2SettlementAbi,
          functionName: "domainSeparator",
          args: []
        });
      }
    };
  }
};

// src/adapters/AbstractSigner.ts
var AbstractSigner = class {
};

// src/adapters/constants.ts
var ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
var ZERO = BigInt(0);
var ONE = BigInt(1);
var ZERO_HASH = "0x0000000000000000000000000000000000000000000000000000000000000000";
var MAX_UINT32 = BigInt(2) ** BigInt(32) - BigInt(1);

// src/utils/cowError.ts
var CowError = class extends Error {
  error_code;
  constructor(message, error_code) {
    super(message);
    this.error_code = error_code;
  }
};

// src/utils/log.ts
var logEnabled = false;
function log(text) {
  if (!logEnabled)
    return;
  console.log(`[COW SDK] ${text}`);
}
function enableLogging(enabled) {
  logEnabled = enabled;
}

// src/utils/serialize.ts
var jsonWithBigintReplacer = (_key, value) => {
  if (typeof value === "bigint") {
    return value.toString();
  }
  if (typeof value === "object" && value !== null && "_isBigNumber" in value) {
    return value.toString();
  }
  return value;
};

// src/utils/math.ts
var SCALE = 1e6;
var SCALE_BIGINT = BigInt(SCALE);
var BPS_FACTOR = 10000n;
function percentageToBps(percentage) {
  const bps = typeof percentage === "bigint" ? Number(percentage * BPS_FACTOR) : percentage * Number(BPS_FACTOR);
  return Math.round(bps);
}
function bpsToPercentage(bps) {
  return bps / 100;
}
function applyPercentage(value, percentage) {
  const valueMultiplied = value * BigInt(Math.floor(percentage * SCALE)) / SCALE_BIGINT;
  const roundUp = valueMultiplied % 100n >= 50n ? 1n : 0n;
  return valueMultiplied / 100n + roundUp;
}

// src/utils/privateKey.ts
function normalizePrivateKey(privateKey) {
  if (!privateKey || typeof privateKey !== "string") {
    throw new CowError("Private key must be a non-empty string");
  }
  const cleanKey = privateKey.startsWith("0x") ? privateKey.slice(2) : privateKey;
  if (!/^[a-fA-F0-9]{64}$/.test(cleanKey)) {
    throw new CowError(
      "Invalid private key format: must be exactly 64 hexadecimal characters (with or without 0x prefix)"
    );
  }
  return `0x${cleanKey}`;
}
function isValidPrivateKey(privateKey) {
  try {
    normalizePrivateKey(privateKey);
    return true;
  } catch {
    return false;
  }
}

// src/utils/ttlCache.ts
var MemoryStorage = class {
  storage = /* @__PURE__ */ new Map();
  getItem(key) {
    return this.storage.get(key) ?? null;
  }
  setItem(key, value) {
    this.storage.set(key, value);
  }
  removeItem(key) {
    this.storage.delete(key);
  }
  // Additional methods for memory storage management
  size(prefix) {
    if (!prefix)
      return this.storage.size;
    let count = 0;
    for (const key of this.storage.keys()) {
      if (key.startsWith(prefix))
        count++;
    }
    return count;
  }
  clear(prefix) {
    if (!prefix) {
      this.storage.clear();
      return;
    }
    const keysToDelete = [];
    for (const key of this.storage.keys()) {
      if (key.startsWith(prefix))
        keysToDelete.push(key);
    }
    keysToDelete.forEach((key) => this.storage.delete(key));
  }
};
var LocalStorageWrapper = class {
  fallback = new MemoryStorage();
  getItem(key) {
    try {
      return typeof localStorage !== "undefined" ? localStorage.getItem(key) : this.fallback.getItem(key);
    } catch {
      return this.fallback.getItem(key);
    }
  }
  setItem(key, value) {
    try {
      if (typeof localStorage !== "undefined") {
        localStorage.setItem(key, value);
      } else {
        this.fallback.setItem(key, value);
      }
    } catch {
      this.fallback.setItem(key, value);
    }
  }
  removeItem(key) {
    try {
      if (typeof localStorage !== "undefined") {
        localStorage.removeItem(key);
      } else {
        this.fallback.removeItem(key);
      }
    } catch {
      this.fallback.removeItem(key);
    }
  }
};
var DEFAULT_KEY_PREFIX = "ttl-cache";
var DEFAULT_TTL = 2 * 60 * 1e3;
var TTLCache = class {
  storage;
  keyPrefix;
  ttl;
  // Track keys written by this instance (covers wrapper fallback/SSR)
  keys = /* @__PURE__ */ new Set();
  get isMemoryStorage() {
    return this.storage instanceof MemoryStorage;
  }
  constructor(keyPrefix = DEFAULT_KEY_PREFIX, useLocalStorage = true, ttl = DEFAULT_TTL) {
    this.keyPrefix = keyPrefix;
    this.ttl = ttl;
    this.storage = useLocalStorage ? new LocalStorageWrapper() : new MemoryStorage();
    if (useLocalStorage && typeof localStorage === "undefined") {
      this.storage = new MemoryStorage();
    }
  }
  /**
   * Set a value in the cache with TTL
   */
  set(key, value) {
    if (this.ttl <= 0)
      return;
    const entry = {
      value,
      timestamp: Date.now(),
      ttl: this.ttl
    };
    const storageKey = this.getStorageKey(key);
    try {
      this.storage.setItem(storageKey, JSON.stringify(entry));
    } catch (error) {
      console.warn("TTLCache: Failed to store cache entry", error);
    }
    this.keys.add(storageKey);
  }
  /**
   * Get a value from the cache, automatically checking TTL
   */
  get(key) {
    const storageKey = this.getStorageKey(key);
    const item = this.storage.getItem(storageKey);
    if (!item) {
      return void 0;
    }
    try {
      const entry = JSON.parse(item);
      const now = Date.now();
      if (now - entry.timestamp > entry.ttl) {
        this.storage.removeItem(storageKey);
        this.keys.delete(storageKey);
        return void 0;
      }
      return entry.value;
    } catch (error) {
      console.warn("TTLCache: Failed to parse cache entry", error);
      this.storage.removeItem(storageKey);
      this.keys.delete(storageKey);
      return void 0;
    }
  }
  /**
   * Remove a specific key from the cache
   */
  delete(key) {
    const storageKey = this.getStorageKey(key);
    this.storage.removeItem(storageKey);
    this.keys.delete(storageKey);
  }
  /**
   * Clear all cache entries with this cache's prefix
   */
  clear() {
    const prefix = this.keyPrefix + ":";
    if (this.isMemoryStorage) {
      ;
      this.storage.clear(prefix);
      this.keys.clear();
      return;
    }
    if (typeof localStorage !== "undefined") {
      try {
        const keysToDelete = [];
        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          if (key && key.startsWith(prefix)) {
            keysToDelete.push(key);
          }
        }
        keysToDelete.forEach((key) => localStorage.removeItem(key));
      } catch {
        console.warn("TTLCache: Failed to clear localStorage cache");
      }
    }
    for (const key of Array.from(this.keys)) {
      try {
        this.storage.removeItem(key);
      } catch (error) {
        console.warn("TTLCache: Failed to remove cache key", error);
      }
      this.keys.delete(key);
    }
  }
  /**
   * Get the number of cache entries (approximate for localStorage)
   */
  size() {
    const prefix = this.keyPrefix + ":";
    if (this.isMemoryStorage && this.storage instanceof MemoryStorage) {
      return this.storage.size(prefix);
    }
    if (typeof localStorage === "undefined") {
      return this.keys.size;
    }
    if (typeof localStorage !== "undefined") {
      try {
        let count = 0;
        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          if (key && key.startsWith(prefix)) {
            count++;
          }
        }
        return count;
      } catch {
        return this.keys.size;
      }
    }
    return this.keys.size;
  }
  /**
   * Clean up expired entries
   */
  cleanup() {
    const prefix = this.keyPrefix + ":";
    if (typeof localStorage !== "undefined") {
      try {
        const keysToDelete = [];
        const now2 = Date.now();
        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          if (key && key.startsWith(prefix)) {
            const item = localStorage.getItem(key);
            if (item) {
              try {
                const entry = JSON.parse(item);
                if (now2 - entry.timestamp > entry.ttl) {
                  keysToDelete.push(key);
                }
              } catch {
                keysToDelete.push(key);
              }
            }
          }
        }
        keysToDelete.forEach((key) => localStorage.removeItem(key));
      } catch {
        console.warn("TTLCache: Failed to cleanup expired entries");
      }
    }
    const now = Date.now();
    for (const key of Array.from(this.keys)) {
      const item = this.storage.getItem(key);
      if (!item) {
        this.keys.delete(key);
        continue;
      }
      try {
        const entry = JSON.parse(item);
        if (now - entry.timestamp > entry.ttl) {
          this.storage.removeItem(key);
          this.keys.delete(key);
        }
      } catch {
        this.storage.removeItem(key);
        this.keys.delete(key);
      }
    }
  }
  getStorageKey(key) {
    return `${this.keyPrefix}:${key}`;
  }
};

// src/utils/address.ts
var EVM_ADDRESS_PATTERN = /^0x[a-fA-F0-9]{40}$/;
var BTC_LEGACY_ADDRESS_PATTERN = /^[13][a-km-zA-HJ-NP-Z1-9]{24,33}$/;
var BTC_BECH32_MAINNET_PATTERN = /^(bc1[a-z0-9]{39,59}|BC1[A-Z0-9]{39,59})$/;
var SOL_ADDRESS_PATTERN = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/;
function isEvmAddress(address) {
  if (typeof address !== "string")
    return false;
  if (!address.startsWith("0x"))
    return false;
  if (address.length !== 42)
    return false;
  return EVM_ADDRESS_PATTERN.test(address);
}
function isBtcAddress(address) {
  if (typeof address !== "string")
    return false;
  if (address.length < 25 || address.length > 62)
    return false;
  if (BTC_LEGACY_ADDRESS_PATTERN.test(address)) {
    return true;
  }
  if (BTC_BECH32_MAINNET_PATTERN.test(address)) {
    return true;
  }
  return false;
}
function getEvmAddressKey(address) {
  return `${address.toLowerCase()}`;
}
function getBtcAddressKey(address) {
  return address;
}
function isSolanaAddress(address) {
  if (typeof address !== "string")
    return false;
  if (address.length < 32 || address.length > 44)
    return false;
  return SOL_ADDRESS_PATTERN.test(address);
}
function getSolAddressKey(address) {
  return address;
}
function getAddressKey(address) {
  if (isEvmAddress(address))
    return getEvmAddressKey(address);
  return address;
}

// src/utils/token.ts
var import_sdk_config = require("@cowprotocol/sdk-config");
function getTokenId(token) {
  const addressKey = getAddressKey(token.address);
  return `${token.chainId}:${addressKey}`;
}
function areTokensEqual(a, b) {
  if (!a || !b)
    return false;
  return getTokenId(a) === getTokenId(b);
}
function areAddressesEqual(a, b) {
  if (!a || !b)
    return false;
  const aIsEvm = isEvmAddress(a);
  const bIsEvm = isEvmAddress(b);
  if (aIsEvm && bIsEvm) {
    return getEvmAddressKey(a) === getEvmAddressKey(b);
  }
  return a === b;
}
function isNativeToken(token) {
  const chainInfo = (0, import_sdk_config.getChainInfo)(token.chainId);
  return areAddressesEqual(chainInfo?.nativeCurrency.address, token.address);
}
function isWrappedNativeToken(token) {
  if (!(0, import_sdk_config.isSupportedChain)(token.chainId)) {
    return false;
  }
  return areAddressesEqual(import_sdk_config.WRAPPED_NATIVE_CURRENCIES[token.chainId]?.address, token.address);
}

// src/utils/protocolContracts.ts
var import_sdk_config2 = require("@cowprotocol/sdk-config");
function isCoWSettlementContract(_address, chainId) {
  const address = getAddressKey(_address);
  return areAddressesEqual(import_sdk_config2.COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS[chainId], address) || areAddressesEqual(import_sdk_config2.COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING[chainId], address);
}
function isCoWVaultRelayerContract(_address, chainId) {
  const address = getAddressKey(_address);
  return areAddressesEqual(import_sdk_config2.COW_PROTOCOL_VAULT_RELAYER_ADDRESS[chainId], address) || areAddressesEqual(import_sdk_config2.COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING[chainId], address);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AbstractProviderAdapter,
  AbstractSigner,
  AdapterContext,
  AdapterUtils,
  ContractFactory,
  CowError,
  ERC20_ALLOWANCE_ABI,
  ERC20_APPROVE_ABI,
  EthFlowAbi,
  GPV2SettlementAbi,
  MAX_UINT256,
  MAX_UINT32,
  ONE,
  TTLCache,
  ZERO,
  ZERO_ADDRESS,
  ZERO_HASH,
  applyPercentage,
  areAddressesEqual,
  areTokensEqual,
  bpsToPercentage,
  enableLogging,
  getAddressKey,
  getBtcAddressKey,
  getEvmAddressKey,
  getGlobalAdapter,
  getSolAddressKey,
  getTokenId,
  isBtcAddress,
  isCoWSettlementContract,
  isCoWVaultRelayerContract,
  isEvmAddress,
  isNativeToken,
  isSolanaAddress,
  isValidPrivateKey,
  isWrappedNativeToken,
  jsonWithBigintReplacer,
  log,
  normalizePrivateKey,
  percentageToBps,
  setGlobalAdapter
});
