// @cowprotocol/sdk-order-book/dist/index.mjs
// src/api.ts
import "cross-fetch/polyfill";
import { RateLimiter } from "limiter";
import {
  DEFAULT_COW_API_CONTEXT,
  ENVS_LIST,
  SupportedChainId
} from "@cowprotocol/sdk-config";
import { CowError, jsonWithBigintReplacer, log } from "@cowprotocol/sdk-common";

// src/request.ts
import { backOff } from "exponential-backoff";
var OrderBookApiError = class extends Error {
  /**
   * Error thrown when the CoW Protocol OrderBook API returns an error.
   * @param response The response from the CoW Protocol OrderBook API.
   * @param body The body of the response.
   * @constructor
   */
  constructor(response, body) {
    super(typeof body === "string" ? body : response.statusText);
    this.response = response;
    this.body = body;
  }
};
var REQUEST_TIMEOUT = 408;
var TOO_EARLY = 425;
var TOO_MANY_REQUESTS = 429;
var INTERNAL_SERVER_ERROR = 500;
var BAD_GATEWAY = 502;
var SERVICE_UNAVAILABLE = 503;
var GATEWAY_TIMEOUT = 504;
var STATUS_CODES_TO_RETRY = [
  REQUEST_TIMEOUT,
  TOO_EARLY,
  TOO_MANY_REQUESTS,
  INTERNAL_SERVER_ERROR,
  BAD_GATEWAY,
  SERVICE_UNAVAILABLE,
  GATEWAY_TIMEOUT
];
var DEFAULT_BACKOFF_OPTIONS = {
  numOfAttempts: 10,
  maxDelay: Infinity,
  jitter: "none",
  retry: (error) => {
    if (error instanceof OrderBookApiError) {
      return STATUS_CODES_TO_RETRY.includes(error.response.status);
    }
    return true;
  }
};
var DEFAULT_LIMITER_OPTIONS = {
  tokensPerInterval: 5,
  interval: "second"
};
var getResponseBody = async (response) => {
  if (response.status !== 204) {
    try {
      const contentType = response.headers.get("Content-Type");
      if (contentType) {
        if (contentType.toLowerCase().startsWith("application/json")) {
          return await response.json();
        } else {
          return await response.text();
        }
      }
    } catch (error) {
      console.error(error);
    }
  }
  return void 0;
};
async function request(baseUrl, { path, query, method, body }, rateLimiter, backoffOpts, additionalHeaders) {
  const queryString = query ? "?" + query : "";
  const headers = {
    Accept: "application/json",
    "Content-Type": "application/json",
    ...additionalHeaders
  };
  const url = `${baseUrl}${path}${queryString}`;
  const bodyContent = (() => {
    if (!body)
      return void 0;
    return typeof body === "string" ? body : JSON.stringify(body);
  })();
  const init = {
    method,
    body: bodyContent,
    headers
  };
  return backOff(async () => {
    await rateLimiter.removeTokens(1);
    const response = await fetch(url, init);
    const responseBody = await getResponseBody(response);
    if (response.status >= 200 && response.status < 300) {
      return responseBody;
    }
    return Promise.reject(new OrderBookApiError(response, responseBody));
  }, backoffOpts);
}

// src/transformOrder.ts
import { ETH_ADDRESS } from "@cowprotocol/sdk-config";
function transformOrder(order) {
  return transformEthFlowOrder(addTotalFeeToOrder(order));
}
function addTotalFeeToOrder(dto) {
  const { executedFeeAmount, executedFee } = dto;
  const _executedFeeAmount = BigInt(executedFeeAmount || "0");
  const _executedFee = BigInt(executedFee || "0");
  const totalFee = String(_executedFeeAmount + _executedFee);
  return {
    ...dto,
    totalFee
  };
}
function transformEthFlowOrder(order) {
  const { ethflowData } = order;
  if (!ethflowData) {
    return order;
  }
  const { userValidTo: validTo } = ethflowData;
  const owner = order.onchainUser || order.owner;
  const sellToken = ETH_ADDRESS;
  return { ...order, validTo, owner, sellToken };
}

// src/api.ts
var PROD_BASE_URL = "https://api.cow.fi";
var STAGING_BASE_URL = "https://barn.api.cow.fi";
var PARTNER_PROD_BASE_URL = "https://partners.cow.fi";
var PARTNER_STAGING_BASE_URL = "https://partners.barn.cow.fi";
var ORDER_BOOK_PROD_CONFIG = {
  [SupportedChainId.MAINNET]: `${PROD_BASE_URL}/mainnet`,
  [SupportedChainId.GNOSIS_CHAIN]: `${PROD_BASE_URL}/xdai`,
  [SupportedChainId.ARBITRUM_ONE]: `${PROD_BASE_URL}/arbitrum_one`,
  [SupportedChainId.BASE]: `${PROD_BASE_URL}/base`,
  [SupportedChainId.SEPOLIA]: `${PROD_BASE_URL}/sepolia`,
  [SupportedChainId.POLYGON]: `${PROD_BASE_URL}/polygon`,
  [SupportedChainId.AVALANCHE]: `${PROD_BASE_URL}/avalanche`,
  [SupportedChainId.BNB]: `${PROD_BASE_URL}/bnb`,
  [SupportedChainId.LINEA]: `${PROD_BASE_URL}/linea`,
  [SupportedChainId.PLASMA]: `${PROD_BASE_URL}/plasma`,
  [SupportedChainId.INK]: `${PROD_BASE_URL}/ink`
};
var ORDER_BOOK_STAGING_CONFIG = {
  [SupportedChainId.MAINNET]: `${STAGING_BASE_URL}/mainnet`,
  [SupportedChainId.GNOSIS_CHAIN]: `${STAGING_BASE_URL}/xdai`,
  [SupportedChainId.ARBITRUM_ONE]: `${STAGING_BASE_URL}/arbitrum_one`,
  [SupportedChainId.BASE]: `${STAGING_BASE_URL}/base`,
  [SupportedChainId.SEPOLIA]: `${STAGING_BASE_URL}/sepolia`,
  [SupportedChainId.POLYGON]: `${STAGING_BASE_URL}/polygon`,
  [SupportedChainId.AVALANCHE]: `${STAGING_BASE_URL}/avalanche`,
  [SupportedChainId.BNB]: `${STAGING_BASE_URL}/bnb`,
  [SupportedChainId.LINEA]: `${STAGING_BASE_URL}/linea`,
  [SupportedChainId.PLASMA]: `${STAGING_BASE_URL}/plasma`,
  [SupportedChainId.INK]: `${STAGING_BASE_URL}/ink`
};
var ORDER_BOOK_PARTNER_PROD_CONFIG = {
  [SupportedChainId.MAINNET]: `${PARTNER_PROD_BASE_URL}/mainnet`,
  [SupportedChainId.GNOSIS_CHAIN]: `${PARTNER_PROD_BASE_URL}/xdai`,
  [SupportedChainId.ARBITRUM_ONE]: `${PARTNER_PROD_BASE_URL}/arbitrum_one`,
  [SupportedChainId.BASE]: `${PARTNER_PROD_BASE_URL}/base`,
  [SupportedChainId.SEPOLIA]: `${PARTNER_PROD_BASE_URL}/sepolia`,
  [SupportedChainId.POLYGON]: `${PARTNER_PROD_BASE_URL}/polygon`,
  [SupportedChainId.AVALANCHE]: `${PARTNER_PROD_BASE_URL}/avalanche`,
  [SupportedChainId.BNB]: `${PARTNER_PROD_BASE_URL}/bnb`,
  [SupportedChainId.LINEA]: `${PARTNER_PROD_BASE_URL}/linea`,
  [SupportedChainId.PLASMA]: `${PARTNER_PROD_BASE_URL}/plasma`,
  [SupportedChainId.INK]: `${PARTNER_PROD_BASE_URL}/ink`
};
var ORDER_BOOK_PARTNER_STAGING_CONFIG = {
  [SupportedChainId.MAINNET]: `${PARTNER_STAGING_BASE_URL}/mainnet`,
  [SupportedChainId.GNOSIS_CHAIN]: `${PARTNER_STAGING_BASE_URL}/xdai`,
  [SupportedChainId.ARBITRUM_ONE]: `${PARTNER_STAGING_BASE_URL}/arbitrum_one`,
  [SupportedChainId.BASE]: `${PARTNER_STAGING_BASE_URL}/base`,
  [SupportedChainId.SEPOLIA]: `${PARTNER_STAGING_BASE_URL}/sepolia`,
  [SupportedChainId.POLYGON]: `${PARTNER_STAGING_BASE_URL}/polygon`,
  [SupportedChainId.AVALANCHE]: `${PARTNER_STAGING_BASE_URL}/avalanche`,
  [SupportedChainId.BNB]: `${PARTNER_STAGING_BASE_URL}/bnb`,
  [SupportedChainId.LINEA]: `${PARTNER_STAGING_BASE_URL}/linea`,
  [SupportedChainId.PLASMA]: `${PARTNER_STAGING_BASE_URL}/plasma`,
  [SupportedChainId.INK]: `${PARTNER_STAGING_BASE_URL}/ink`
};
function cleanObjectFromUndefinedValues(obj) {
  return Object.keys(obj).reduce(
    (acc, key) => {
      const val = obj[key];
      if (typeof val !== "undefined")
        acc[key] = val;
      return acc;
    },
    {}
  );
}
var OrderBookApi = class {
  context;
  rateLimiter;
  /**
   * Creates a new instance of the CoW Protocol OrderBook API client.
   * @param context - The API context to use. If not provided, the default context will be used.
   */
  constructor(context = {}) {
    this.context = { ...DEFAULT_COW_API_CONTEXT, ...context };
    this.rateLimiter = new RateLimiter(context.limiterOpts || DEFAULT_LIMITER_OPTIONS);
  }
  /**
   * Get the version of the API.
   * @param contextOverride Optional context override for this request.
   * @returns The version of the API.
   * @see {@link https://api.cow.fi/docs/#/default/get_api_v1_version}
   */
  getVersion(contextOverride = {}) {
    return this.fetch({ path: "/api/v1/version", method: "GET" }, contextOverride);
  }
  /**
   * Get all the trades for either an `owner` **OR** `orderUid`.
   *
   * Given that an order *may* be partially fillable, it is possible that a discrete order (`orderUid`)
   * may have *multiple* trades. Therefore, this method returns a list of trades, either for *all* the orders
   * of a given `owner`, or for a discrete order (`orderUid`).
   * @param request Either an `owner` or an `orderUid` **MUST** be specified.
   * @param contextOverride Optional context override for this request.
   * @returns A list of trades matching the request.
   */
  getTrades(request2, contextOverride = {}) {
    if (request2.owner && request2.orderUid) {
      return Promise.reject(new CowError("Cannot specify both owner and orderId"));
    } else if (!request2.owner && !request2.orderUid) {
      return Promise.reject(new CowError("Must specify either owner or orderId"));
    }
    const { offset = 0, limit = 10, ...rest } = request2;
    const params = { ...rest, offset: offset.toString(), limit: limit.toString() };
    const query = new URLSearchParams(cleanObjectFromUndefinedValues(params));
    return this.fetch({ path: "/api/v2/trades", method: "GET", query }, contextOverride);
  }
  /**
   * Get a list of orders for a given `owner`.
   * @param request The request parameters with `request.offset = 0` and `request.limit = 1000` by default.
   * @param contextOverride Optional context override for this request.
   * @returns A list of orders matching the request.
   * @see {@link GetOrdersRequest}
   * @see {@link EnrichedOrder}
   */
  getOrders({ owner, offset = 0, limit = 1e3 }, contextOverride = {}) {
    const query = new URLSearchParams(
      cleanObjectFromUndefinedValues({ offset: offset.toString(), limit: limit.toString() })
    );
    return this.fetch(
      { path: `/api/v1/account/${owner}/orders`, method: "GET", query },
      contextOverride
    ).then((orders) => {
      return orders.map(transformOrder);
    });
  }
  /**
   * Get a list of orders from a given settlement transaction hash.
   * @param txHash The transaction hash.
   * @param contextOverride Optional context override for this request.
   * @returns A list of orders matching the request.
   * @see {@link EnrichedOrder}
   */
  getTxOrders(txHash, contextOverride = {}) {
    return this.fetch(
      { path: `/api/v1/transactions/${txHash}/orders`, method: "GET" },
      contextOverride
    ).then((orders) => {
      return orders.map(transformOrder);
    });
  }
  /**
   * Get an order by its unique identifier, `orderUid`.
   * @param orderUid The unique identifier of the order.
   * @param contextOverride Optional context override for this request.
   * @returns The order matching the request.
   */
  getOrder(orderUid, contextOverride = {}) {
    return this.fetch({ path: `/api/v1/orders/${orderUid}`, method: "GET" }, contextOverride).then((order) => {
      return transformOrder(order);
    });
  }
  /**
   * Get the order status while open
   */
  getOrderCompetitionStatus(orderUid, contextOverride = {}) {
    return this.fetch({ path: `/api/v1/orders/${orderUid}/status`, method: "GET" }, contextOverride);
  }
  /**
   * Attempt to get an order by its unique identifier, `orderUid`, from multiple environments.
   *
   * **NOTE**: The environment refers to either `prod` or `staging`. This allows a conveience method to
   * attempt to get an order from both environments, in the event that the order is not found in the
   * environment specified in the context.
   * @param orderUid The unique identifier of the order.
   * @param contextOverride Optional context override for this request.
   * @returns The order matching the request.
   * @throws {OrderBookApiError} If the order is not found in any of the environments.
   */
  getOrderMultiEnv(orderUid, contextOverride = {}) {
    const { env } = this.getContextWithOverride(contextOverride);
    const otherEnvs = ENVS_LIST.filter((i) => i !== env);
    let attemptsCount = 0;
    const fallback = (error) => {
      const nextEnv = otherEnvs[attemptsCount];
      if (error instanceof OrderBookApiError && error.response.status === 404 && nextEnv) {
        attemptsCount++;
        return this.getOrder(orderUid, { ...contextOverride, env: nextEnv }).catch(fallback);
      }
      return Promise.reject(error);
    };
    return this.getOrder(orderUid, { ...contextOverride, env }).catch(fallback);
  }
  /**
   * Get a quote for an order.
   * This allows for the calculation of the total cost of an order, including fees, before signing and submitting.
   * @param requestBody The parameters for the order quote request.
   * @param contextOverride Optional context override for this request.
   * @returns A hydrated order matching the request ready to be signed.
   */
  getQuote(requestBody, contextOverride = {}) {
    return this.fetch({ path: "/api/v1/quote", method: "POST", body: requestBody }, contextOverride);
  }
  /**
   * Cancel one or more orders.
   *
   * **NOTE**: Cancellation is on a best-effort basis. Orders that are already in the process of being settled
   * (ie. transaction has been submitted to chain by the solver) cannot not be cancelled.
   * **CAUTION**: This method can only be used to cancel orders that were signed using `EIP-712` or `eth_sign (EIP-191)`.
   * @param requestBody Orders to be cancelled and signed instructions to cancel them.
   * @param contextOverride Optional context override for this request.
   * @returns A list of order unique identifiers that were successfully cancelled.
   */
  sendSignedOrderCancellations(requestBody, contextOverride = {}) {
    return this.fetch({ path: "/api/v1/orders", method: "DELETE", body: requestBody }, contextOverride);
  }
  /**
   * Submit an order to the order book.
   * @param requestBody The signed order to be submitted.
   * @param contextOverride Optional context override for this request.
   * @returns The unique identifier of the order.
   */
  sendOrder(requestBody, contextOverride = {}) {
    return this.fetch({ path: "/api/v1/orders", method: "POST", body: requestBody }, contextOverride);
  }
  /**
   * Get the native price of a token.
   *
   * **NOTE**: The native price is the price of the token in the native currency of the chain. For example, on Ethereum
   * this would be the price of the token in ETH.
   * @param tokenAddress The address of the ERC-20 token.
   * @param contextOverride Optional context override for this request.
   * @returns The native price of the token.
   */
  getNativePrice(tokenAddress, contextOverride = {}) {
    return this.fetch({ path: `/api/v1/token/${tokenAddress}/native_price`, method: "GET" }, contextOverride);
  }
  /**
   * Given a user's address, get the total surplus that they have earned.
   * @param address The user's address
   * @param contextOverride Optional context override for this request.
   * @returns Calculated user's surplus
   */
  getTotalSurplus(address, contextOverride = {}) {
    return this.fetch({ path: `/api/v1/users/${address}/total_surplus`, method: "GET" }, contextOverride);
  }
  /**
   * Retrieve the full app data for a given app data hash.
   * @param appDataHash `bytes32` hash of the app data
   * @param contextOverride Optional context override for this request.
   * @returns Full app data that was uploaded
   */
  getAppData(appDataHash, contextOverride = {}) {
    return this.fetch({ path: `/api/v1/app_data/${appDataHash}`, method: "GET" }, contextOverride);
  }
  /**
   * Upload the full app data that corresponds to a given app data hash.
   * @param appDataHash `bytes32` hash of the app data
   * @param fullAppData Full app data to be uploaded
   * @param contextOverride Optional context override for this request.
   * @returns The string encoding of the full app data that was uploaded.
   */
  uploadAppData(appDataHash, fullAppData, contextOverride = {}) {
    return this.fetch(
      { path: `/api/v1/app_data/${appDataHash}`, method: "PUT", body: { fullAppData } },
      contextOverride
    );
  }
  /**
   * Given an auction id or tx hash, get the details of the solver competition for that auction.
   * @param auctionIdorTx auction id or tx hash corresponding to the auction
   * @param contextOverride Optional context override for this request.
   * @returns An object containing the solver competition details
   */
  getSolverCompetition(auctionIdorTx, contextOverride = {}) {
    return this.fetch(
      {
        path: `/api/v1/solver_competition${typeof auctionIdorTx === "string" ? "/by_tx_hash" : ""}/${auctionIdorTx}`,
        method: "GET"
      },
      contextOverride
    );
  }
  /**
   * Generate an API endpoint for an order by its unique identifier, `orderUid`.
   * @param orderUid The unique identifier of the order.
   * @param contextOverride Optional context override for this request.
   * @returns The API endpoint to get the order.
   */
  getOrderLink(orderUid, contextOverride) {
    const context = this.getContextWithOverride(contextOverride ?? {});
    return this.getApiBaseUrls(context)[context.chainId] + `/api/v1/orders/${orderUid}`;
  }
  /**
   * Apply an override to the context for a request.
   * @param contextOverride Optional context override for this request.
   * @returns New context with the override applied.
   */
  getContextWithOverride(contextOverride = {}) {
    return { ...this.context, ...contextOverride };
  }
  /**
   * Get the base URLs for the API endpoints given the context.
   * Uses partner URLs when apiKey is set (and no custom baseUrls override).
   * @param context The merged API context for the request.
   * @returns The base URLs for the API endpoints.
   */
  getApiBaseUrls(context) {
    if (context.baseUrls)
      return context.baseUrls;
    if (context.apiKey) {
      return context.env === "prod" ? ORDER_BOOK_PARTNER_PROD_CONFIG : ORDER_BOOK_PARTNER_STAGING_CONFIG;
    }
    return context.env === "prod" ? ORDER_BOOK_PROD_CONFIG : ORDER_BOOK_STAGING_CONFIG;
  }
  /**
   * Make a request to the API.
   * @param params The parameters for the request.
   * @param contextOverride Optional context override for this request.
   * @returns The response from the API.
   */
  fetch(params, contextOverride = {}) {
    const context = this.getContextWithOverride(contextOverride);
    const { chainId, backoffOpts: _backoffOpts, apiKey } = context;
    const baseUrl = this.getApiBaseUrls(context)[chainId];
    const backoffOpts = _backoffOpts || DEFAULT_BACKOFF_OPTIONS;
    const rateLimiter = contextOverride.limiterOpts ? new RateLimiter(contextOverride.limiterOpts) : this.rateLimiter;
    const additionalHeaders = apiKey ? { "X-API-Key": apiKey } : void 0;
    log(`Fetching OrderBook API: ${baseUrl}${params.path}. Params: ${JSON.stringify(params, jsonWithBigintReplacer)}`);
    return request(baseUrl, params, rateLimiter, backoffOpts, additionalHeaders);
  }
};

// src/generated/models/BuyTokenDestination.ts
var BuyTokenDestination = /* @__PURE__ */ ((BuyTokenDestination2) => {
  BuyTokenDestination2["ERC20"] = "erc20";
  BuyTokenDestination2["INTERNAL"] = "internal";
  return BuyTokenDestination2;
})(BuyTokenDestination || {});

// src/generated/models/CompetitionOrderStatus.ts
var CompetitionOrderStatus;
((CompetitionOrderStatus2) => {
  let type;
  ((type2) => {
    type2["OPEN"] = "open";
    type2["SCHEDULED"] = "scheduled";
    type2["ACTIVE"] = "active";
    type2["SOLVED"] = "solved";
    type2["EXECUTING"] = "executing";
    type2["TRADED"] = "traded";
    type2["CANCELLED"] = "cancelled";
  })(type = CompetitionOrderStatus2.type || (CompetitionOrderStatus2.type = {}));
})(CompetitionOrderStatus || (CompetitionOrderStatus = {}));

// src/generated/models/EcdsaSigningScheme.ts
var EcdsaSigningScheme = /* @__PURE__ */ ((EcdsaSigningScheme2) => {
  EcdsaSigningScheme2["EIP712"] = "eip712";
  EcdsaSigningScheme2["ETHSIGN"] = "ethsign";
  return EcdsaSigningScheme2;
})(EcdsaSigningScheme || {});

// src/generated/models/OnchainOrderData.ts
var OnchainOrderData;
((OnchainOrderData2) => {
  let placementError;
  ((placementError2) => {
    placementError2["QUOTE_NOT_FOUND"] = "QuoteNotFound";
    placementError2["VALID_TO_TOO_FAR_IN_FUTURE"] = "ValidToTooFarInFuture";
    placementError2["PRE_VALIDATION_ERROR"] = "PreValidationError";
  })(placementError = OnchainOrderData2.placementError || (OnchainOrderData2.placementError = {}));
})(OnchainOrderData || (OnchainOrderData = {}));

// src/generated/models/OrderCancellationError.ts
var OrderCancellationError;
((OrderCancellationError2) => {
  let errorType;
  ((errorType2) => {
    errorType2["INVALID_SIGNATURE"] = "InvalidSignature";
    errorType2["WRONG_OWNER"] = "WrongOwner";
    errorType2["ORDER_NOT_FOUND"] = "OrderNotFound";
    errorType2["ALREADY_CANCELLED"] = "AlreadyCancelled";
    errorType2["ORDER_FULLY_EXECUTED"] = "OrderFullyExecuted";
    errorType2["ORDER_EXPIRED"] = "OrderExpired";
    errorType2["ON_CHAIN_ORDER"] = "OnChainOrder";
  })(errorType = OrderCancellationError2.errorType || (OrderCancellationError2.errorType = {}));
})(OrderCancellationError || (OrderCancellationError = {}));

// src/generated/models/OrderClass.ts
var OrderClass = /* @__PURE__ */ ((OrderClass2) => {
  OrderClass2["MARKET"] = "market";
  OrderClass2["LIMIT"] = "limit";
  OrderClass2["LIQUIDITY"] = "liquidity";
  return OrderClass2;
})(OrderClass || {});

// src/generated/models/OrderKind.ts
var OrderKind = /* @__PURE__ */ ((OrderKind2) => {
  OrderKind2["BUY"] = "buy";
  OrderKind2["SELL"] = "sell";
  return OrderKind2;
})(OrderKind || {});

// src/generated/models/OrderPostError.ts
var OrderPostError;
((OrderPostError2) => {
  let errorType;
  ((errorType2) => {
    errorType2["DUPLICATED_ORDER"] = "DuplicatedOrder";
    errorType2["QUOTE_NOT_FOUND"] = "QuoteNotFound";
    errorType2["QUOTE_NOT_VERIFIED"] = "QuoteNotVerified";
    errorType2["INVALID_QUOTE"] = "InvalidQuote";
    errorType2["MISSING_FROM"] = "MissingFrom";
    errorType2["WRONG_OWNER"] = "WrongOwner";
    errorType2["INVALID_EIP1271SIGNATURE"] = "InvalidEip1271Signature";
    errorType2["INSUFFICIENT_BALANCE"] = "InsufficientBalance";
    errorType2["INSUFFICIENT_ALLOWANCE"] = "InsufficientAllowance";
    errorType2["INVALID_SIGNATURE"] = "InvalidSignature";
    errorType2["SELL_AMOUNT_OVERFLOW"] = "SellAmountOverflow";
    errorType2["TRANSFER_SIMULATION_FAILED"] = "TransferSimulationFailed";
    errorType2["ZERO_AMOUNT"] = "ZeroAmount";
    errorType2["INCOMPATIBLE_SIGNING_SCHEME"] = "IncompatibleSigningScheme";
    errorType2["TOO_MANY_LIMIT_ORDERS"] = "TooManyLimitOrders";
    errorType2["TOO_MUCH_GAS"] = "TooMuchGas";
    errorType2["UNSUPPORTED_BUY_TOKEN_DESTINATION"] = "UnsupportedBuyTokenDestination";
    errorType2["UNSUPPORTED_SELL_TOKEN_SOURCE"] = "UnsupportedSellTokenSource";
    errorType2["UNSUPPORTED_ORDER_TYPE"] = "UnsupportedOrderType";
    errorType2["INSUFFICIENT_VALID_TO"] = "InsufficientValidTo";
    errorType2["EXCESSIVE_VALID_TO"] = "ExcessiveValidTo";
    errorType2["INVALID_NATIVE_SELL_TOKEN"] = "InvalidNativeSellToken";
    errorType2["SAME_BUY_AND_SELL_TOKEN"] = "SameBuyAndSellToken";
    errorType2["UNSUPPORTED_TOKEN"] = "UnsupportedToken";
    errorType2["INVALID_APP_DATA"] = "InvalidAppData";
    errorType2["APP_DATA_HASH_MISMATCH"] = "AppDataHashMismatch";
    errorType2["APPDATA_FROM_MISMATCH"] = "AppdataFromMismatch";
    errorType2["OLD_ORDER_ACTIVELY_BID_ON"] = "OldOrderActivelyBidOn";
  })(errorType = OrderPostError2.errorType || (OrderPostError2.errorType = {}));
})(OrderPostError || (OrderPostError = {}));

// src/generated/models/OrderQuoteSideKindBuy.ts
var OrderQuoteSideKindBuy = /* @__PURE__ */ ((OrderQuoteSideKindBuy2) => {
  OrderQuoteSideKindBuy2["BUY"] = "buy";
  return OrderQuoteSideKindBuy2;
})(OrderQuoteSideKindBuy || {});

// src/generated/models/OrderQuoteSideKindSell.ts
var OrderQuoteSideKindSell = /* @__PURE__ */ ((OrderQuoteSideKindSell2) => {
  OrderQuoteSideKindSell2["SELL"] = "sell";
  return OrderQuoteSideKindSell2;
})(OrderQuoteSideKindSell || {});

// src/generated/models/OrderStatus.ts
var OrderStatus = /* @__PURE__ */ ((OrderStatus2) => {
  OrderStatus2["PRESIGNATURE_PENDING"] = "presignaturePending";
  OrderStatus2["OPEN"] = "open";
  OrderStatus2["FULFILLED"] = "fulfilled";
  OrderStatus2["CANCELLED"] = "cancelled";
  OrderStatus2["EXPIRED"] = "expired";
  return OrderStatus2;
})(OrderStatus || {});

// src/generated/models/PriceEstimationError.ts
var PriceEstimationError;
((PriceEstimationError2) => {
  let errorType;
  ((errorType2) => {
    errorType2["QUOTE_NOT_VERIFIED"] = "QuoteNotVerified";
    errorType2["UNSUPPORTED_TOKEN"] = "UnsupportedToken";
    errorType2["ZERO_AMOUNT"] = "ZeroAmount";
    errorType2["UNSUPPORTED_ORDER_TYPE"] = "UnsupportedOrderType";
  })(errorType = PriceEstimationError2.errorType || (PriceEstimationError2.errorType = {}));
})(PriceEstimationError || (PriceEstimationError = {}));

// src/generated/models/PriceQuality.ts
var PriceQuality = /* @__PURE__ */ ((PriceQuality2) => {
  PriceQuality2["FAST"] = "fast";
  PriceQuality2["OPTIMAL"] = "optimal";
  PriceQuality2["VERIFIED"] = "verified";
  return PriceQuality2;
})(PriceQuality || {});

// src/generated/models/SellTokenSource.ts
var SellTokenSource = /* @__PURE__ */ ((SellTokenSource2) => {
  SellTokenSource2["ERC20"] = "erc20";
  SellTokenSource2["INTERNAL"] = "internal";
  SellTokenSource2["EXTERNAL"] = "external";
  return SellTokenSource2;
})(SellTokenSource || {});

// src/generated/models/SigningScheme.ts
var SigningScheme = /* @__PURE__ */ ((SigningScheme2) => {
  SigningScheme2["EIP712"] = "eip712";
  SigningScheme2["ETHSIGN"] = "ethsign";
  SigningScheme2["PRESIGN"] = "presign";
  SigningScheme2["EIP1271"] = "eip1271";
  return SigningScheme2;
})(SigningScheme || {});

// src/quoteAmountsAndCosts/quoteAmountsAndCosts.const.ts
var HUNDRED_THOUSANDS = 1e5;
var ONE_HUNDRED_BPS = BigInt(100 * 100);

// src/quoteAmountsAndCosts/getProtocolFeeAmount.ts
var PROTOCOL_FEE_BPS_SCALE = BigInt(HUNDRED_THOUSANDS);
function getProtocolFeeAmount(params) {
  const { orderParams, protocolFeeBps } = params;
  const isSell = orderParams.kind === "sell" /* SELL */;
  if (protocolFeeBps <= 0) {
    return 0n;
  }
  const { sellAmount: sellAmountStr, buyAmount: buyAmountStr, feeAmount: feeAmountStr } = orderParams;
  const sellAmount = BigInt(sellAmountStr);
  const buyAmount = BigInt(buyAmountStr);
  const feeAmount = BigInt(feeAmountStr);
  const protocolFeeScale = PROTOCOL_FEE_BPS_SCALE;
  const protocolFeeBpsBig = BigInt(Math.round(protocolFeeBps * HUNDRED_THOUSANDS));
  if (protocolFeeBpsBig <= 0n) {
    return 0n;
  }
  if (isSell) {
    const denominator = ONE_HUNDRED_BPS * protocolFeeScale - protocolFeeBpsBig;
    return buyAmount * protocolFeeBpsBig / denominator;
  } else {
    const denominator = ONE_HUNDRED_BPS * protocolFeeScale + protocolFeeBpsBig;
    return (sellAmount + feeAmount) * protocolFeeBpsBig / denominator;
  }
}

// src/quoteAmountsAndCosts/getQuoteAmountsAfterPartnerFee.ts
function getQuoteAmountsAfterPartnerFee(params) {
  const { afterNetworkCosts, beforeAllFees, isSell, partnerFeeBps } = params;
  const surplusAmountForPartnerFee = isSell ? beforeAllFees.buyAmount : beforeAllFees.sellAmount;
  const partnerFeeAmount = partnerFeeBps > 0 ? surplusAmountForPartnerFee * BigInt(partnerFeeBps) / ONE_HUNDRED_BPS : BigInt(0);
  const afterPartnerFees = isSell ? {
    sellAmount: afterNetworkCosts.sellAmount,
    buyAmount: afterNetworkCosts.buyAmount - partnerFeeAmount
  } : {
    sellAmount: afterNetworkCosts.sellAmount + partnerFeeAmount,
    buyAmount: afterNetworkCosts.buyAmount
  };
  return {
    partnerFeeAmount,
    afterPartnerFees
  };
}

// src/quoteAmountsAndCosts/getQuoteAmountsAfterSlippage.ts
function getQuoteAmountsAfterSlippage(params) {
  const { afterPartnerFees, isSell, slippageBps } = params;
  const getSlippageAmount = (amount) => amount * BigInt(slippageBps) / ONE_HUNDRED_BPS;
  const afterSlippage = isSell ? {
    sellAmount: afterPartnerFees.sellAmount,
    buyAmount: afterPartnerFees.buyAmount - getSlippageAmount(afterPartnerFees.buyAmount)
  } : {
    sellAmount: afterPartnerFees.sellAmount + getSlippageAmount(afterPartnerFees.sellAmount),
    buyAmount: afterPartnerFees.buyAmount
  };
  return {
    afterSlippage
  };
}

// src/quoteAmountsAndCosts/getQuoteAmountsAndCosts.ts
function getQuoteAmountsAndCosts(params) {
  const { orderParams, slippagePercentBps } = params;
  const partnerFeeBps = params.partnerFeeBps ?? 0;
  const protocolFeeBps = params.protocolFeeBps ?? 0;
  const isSell = orderParams.kind === "sell" /* SELL */;
  const sellAmount = BigInt(orderParams.sellAmount);
  const buyAmount = BigInt(orderParams.buyAmount);
  const networkCostAmount = BigInt(orderParams.feeAmount);
  const networkCostAmountInBuyCurrency = buyAmount * networkCostAmount / sellAmount;
  const protocolFeeAmount = getProtocolFeeAmount({
    orderParams,
    protocolFeeBps
  });
  const beforeAllFees = isSell ? {
    // Only in case of SELL order, network costs are already added to sellAmount which comes from /quote API
    sellAmount: sellAmount + networkCostAmount,
    /**
     * Important!
     * For SELL orders, /quote API returns sellAmount AFTER network costs
     * buyAmount is correlated to that sellAmount, hence, it doesn't include network costs as well
     * Because of that, we add network costs here as well to have both amounts consistent
     */
    buyAmount: buyAmount + networkCostAmountInBuyCurrency + protocolFeeAmount
  } : {
    sellAmount: sellAmount - protocolFeeAmount,
    buyAmount
  };
  const afterProtocolFees = isSell ? {
    sellAmount: beforeAllFees.sellAmount,
    buyAmount: beforeAllFees.buyAmount - protocolFeeAmount
  } : {
    sellAmount,
    buyAmount: beforeAllFees.buyAmount
  };
  const afterNetworkCosts = isSell ? {
    // For SELL order, the /quote API response sellAmount is already after network costs
    // buyAmount is relative to sellAmount, hence it's also after network costs
    sellAmount,
    buyAmount
  } : {
    // For BUY order, the /quote API response sellAmount is only after protocolFee, so we need to add networks costs to the amount
    sellAmount: sellAmount + networkCostAmount,
    buyAmount: afterProtocolFees.buyAmount
  };
  const { afterPartnerFees, partnerFeeAmount } = getQuoteAmountsAfterPartnerFee({
    afterNetworkCosts,
    beforeAllFees,
    isSell,
    partnerFeeBps
  });
  const { afterSlippage } = getQuoteAmountsAfterSlippage({
    afterPartnerFees,
    isSell,
    slippageBps: slippagePercentBps
  });
  const amountsToSign = isSell ? {
    sellAmount: beforeAllFees.sellAmount,
    buyAmount: afterSlippage.buyAmount
  } : {
    sellAmount: afterSlippage.sellAmount,
    buyAmount: beforeAllFees.buyAmount
  };
  return {
    isSell,
    costs: {
      networkFee: {
        amountInSellCurrency: networkCostAmount,
        amountInBuyCurrency: networkCostAmountInBuyCurrency
      },
      partnerFee: {
        amount: partnerFeeAmount,
        bps: partnerFeeBps
      },
      protocolFee: {
        amount: protocolFeeAmount,
        bps: protocolFeeBps
      }
    },
    beforeAllFees,
    beforeNetworkCosts: afterProtocolFees,
    afterProtocolFees,
    afterNetworkCosts,
    afterPartnerFees,
    afterSlippage,
    amountsToSign
  };
}
export {
  BuyTokenDestination,
  CompetitionOrderStatus,
  DEFAULT_BACKOFF_OPTIONS,
  DEFAULT_LIMITER_OPTIONS,
  EcdsaSigningScheme,
  ORDER_BOOK_PARTNER_PROD_CONFIG,
  ORDER_BOOK_PARTNER_STAGING_CONFIG,
  ORDER_BOOK_PROD_CONFIG,
  ORDER_BOOK_STAGING_CONFIG,
  OnchainOrderData,
  OrderBookApi,
  OrderBookApiError,
  OrderCancellationError,
  OrderClass,
  OrderKind,
  OrderPostError,
  OrderQuoteSideKindBuy,
  OrderQuoteSideKindSell,
  OrderStatus,
  PriceEstimationError,
  PriceQuality,
  SellTokenSource,
  SigningScheme,
  getProtocolFeeAmount,
  getQuoteAmountsAfterPartnerFee,
  getQuoteAmountsAfterSlippage,
  getQuoteAmountsAndCosts,
  request
};
