// @cowprotocol/sdk-trading/dist/index.mjs
// src/postCoWProtocolTrade.ts
import { SigningScheme as SigningScheme3 } from "@cowprotocol/sdk-order-book";

// src/consts.ts
import { EcdsaSigningScheme, SigningScheme } from "@cowprotocol/sdk-order-book";
import { mapSupportedNetworks } from "@cowprotocol/sdk-config";
var DEFAULT_QUOTE_VALIDITY = 60 * 30;
var DEFAULT_SLIPPAGE_BPS = 50;
var ETH_FLOW_DEFAULT_SLIPPAGE_BPS = mapSupportedNetworks(DEFAULT_SLIPPAGE_BPS);
var SIGN_SCHEME_MAP = {
  [EcdsaSigningScheme.EIP712]: SigningScheme.EIP712,
  [EcdsaSigningScheme.ETHSIGN]: SigningScheme.ETHSIGN
};
var GAS_LIMIT_DEFAULT = BigInt(15e4);

// src/postCoWProtocolTrade.ts
import { OrderSigningUtils as OrderSigningUtils2 } from "@cowprotocol/sdk-order-signing";

// src/getOrderToSign.ts
import {
  BuyTokenDestination,
  getQuoteAmountsAndCosts,
  SellTokenSource
} from "@cowprotocol/sdk-order-book";

// src/utils/getPartnerFeeBps.ts
function getPartnerFeeBps(partnerFee) {
  if (!partnerFee) {
    return void 0;
  }
  if ("volumeBps" in partnerFee) {
    return partnerFee.volumeBps;
  }
  if (Array.isArray(partnerFee)) {
    for (const fee of partnerFee) {
      if ("volumeBps" in fee) {
        return fee.volumeBps;
      }
    }
  }
  return void 0;
}

// src/utils/slippage.ts
var SCALE = 10n ** 6n;
function getSlippagePercent(params) {
  const { sellAmountBeforeNetworkCosts, sellAmountAfterNetworkCosts, slippage, isSell } = params;
  const sellAmount = isSell ? sellAmountAfterNetworkCosts : sellAmountBeforeNetworkCosts;
  if (sellAmount <= 0n) {
    throw new Error("sellAmount must be greater than 0: " + sellAmount);
  }
  if (slippage < 0n) {
    throw new Error("slippage must be non-negative: " + slippage);
  }
  if (isSell) {
    const percentageInScale = SCALE - SCALE * (sellAmount - slippage) / sellAmount;
    return Number(percentageInScale) / Number(SCALE);
  } else {
    const percentageInScale = SCALE * (sellAmount + slippage) / sellAmount - SCALE;
    return Number(percentageInScale) / Number(SCALE);
  }
}
function getDefaultSlippageBps(chainId, isEthFlow) {
  if (isEthFlow) {
    return ETH_FLOW_DEFAULT_SLIPPAGE_BPS[chainId];
  } else {
    return DEFAULT_SLIPPAGE_BPS;
  }
}

// src/utils/order.ts
function getOrderDeadlineFromNow(validFor) {
  return BigInt(Math.floor(Date.now() / 1e3) + validFor);
}
function unsignedOrderForSigning(order) {
  return {
    ...order,
    sellTokenBalance: order.sellTokenBalance,
    buyTokenBalance: order.buyTokenBalance
  };
}

// src/getOrderToSign.ts
function getOrderToSign({
  chainId,
  from,
  networkCostsAmount = "0",
  isEthFlow,
  applyCostsSlippageAndFees = true,
  protocolFeeBps
}, limitOrderParams, appDataKeccak256) {
  const {
    sellAmount,
    buyAmount,
    sellToken,
    buyToken,
    kind,
    partiallyFillable = false,
    slippageBps = getDefaultSlippageBps(chainId, isEthFlow),
    partnerFee,
    validFor = DEFAULT_QUOTE_VALIDITY
  } = limitOrderParams;
  const receiver = limitOrderParams.receiver || from;
  const validTo = limitOrderParams.validTo || Number(getOrderDeadlineFromNow(validFor));
  const orderParams = {
    sellToken,
    buyToken,
    sellAmount,
    buyAmount,
    receiver,
    validTo,
    kind,
    feeAmount: networkCostsAmount,
    appData: appDataKeccak256,
    partiallyFillable
  };
  let sellAmountToUse = sellAmount;
  let buyAmountToUse = buyAmount;
  if (applyCostsSlippageAndFees) {
    const { isSell, beforeAllFees, afterSlippage } = getQuoteAmountsAndCosts({
      orderParams,
      slippagePercentBps: slippageBps,
      partnerFeeBps: getPartnerFeeBps(partnerFee),
      protocolFeeBps
    });
    sellAmountToUse = isSell ? beforeAllFees.sellAmount.toString() : afterSlippage.sellAmount.toString();
    buyAmountToUse = isSell ? afterSlippage.buyAmount.toString() : beforeAllFees.buyAmount.toString();
  }
  return {
    sellToken,
    buyToken,
    sellAmount: sellAmountToUse,
    buyAmount: buyAmountToUse,
    validTo,
    kind,
    partiallyFillable,
    appData: appDataKeccak256,
    receiver,
    feeAmount: "0",
    sellTokenBalance: SellTokenSource.ERC20,
    buyTokenBalance: BuyTokenDestination.ERC20
  };
}

// src/postSellNativeCurrencyOrder.ts
import { getGlobalAdapter as getGlobalAdapter2 } from "@cowprotocol/sdk-common";
import { SigningScheme as SigningScheme2 } from "@cowprotocol/sdk-order-book";

// src/calculateUniqueOrderId.ts
import { OrderSigningUtils } from "@cowprotocol/sdk-order-signing";
import {
  BARN_ETH_FLOW_ADDRESSES,
  ETH_FLOW_ADDRESSES,
  MAX_VALID_TO_EPOCH,
  WRAPPED_NATIVE_CURRENCIES
} from "@cowprotocol/sdk-config";
async function calculateUniqueOrderId(chainId, order, checkEthFlowOrderExists, options) {
  const { env, ethFlowContractOverride } = options ?? {};
  const { orderDigest, orderId } = await OrderSigningUtils.generateOrderId(
    chainId,
    {
      ...unsignedOrderForSigning(order),
      validTo: MAX_VALID_TO_EPOCH,
      sellToken: WRAPPED_NATIVE_CURRENCIES[chainId].address
    },
    {
      owner: ethFlowContractOverride?.[chainId] ?? (env === "staging" ? BARN_ETH_FLOW_ADDRESSES[chainId] : ETH_FLOW_ADDRESSES[chainId])
    },
    options
  );
  if (checkEthFlowOrderExists && await checkEthFlowOrderExists(orderId, orderDigest)) {
    console.error("ETH FLOW", "[calculateUniqueOrderId] \u274C Collision detected: " + orderId, {
      sellAmount: order.sellAmount,
      fee: order.feeAmount
    });
    return calculateUniqueOrderId(
      chainId,
      adjustAmounts(order),
      checkEthFlowOrderExists,
      options
    );
  }
  return orderId;
}
function adjustAmounts(order) {
  const buyAmount = BigInt(order.buyAmount);
  return { ...order, buyAmount: (buyAmount - BigInt(1)).toString() };
}

// src/getEthFlowTransaction.ts
import {
  BARN_ETH_FLOW_ADDRESSES as BARN_ETH_FLOW_ADDRESSES2,
  ETH_FLOW_ADDRESSES as ETH_FLOW_ADDRESSES2
} from "@cowprotocol/sdk-config";

// src/utils/misc.ts
import { ETH_ADDRESS, WRAPPED_NATIVE_CURRENCIES as WRAPPED_NATIVE_CURRENCIES2 } from "@cowprotocol/sdk-config";
import { areAddressesEqual } from "@cowprotocol/sdk-common";
function swapParamsToLimitOrderParams(params, quoteResponse) {
  return {
    ...params,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    quoteId: quoteResponse.id,
    sellAmount: quoteResponse.quote.sellAmount,
    buyAmount: quoteResponse.quote.buyAmount
  };
}
function getIsEthFlowOrder(params) {
  return areAddressesEqual(params.sellToken, ETH_ADDRESS);
}
function calculateGasMargin(value) {
  return value + value * BigInt(20) / BigInt(100);
}
function mapQuoteAmountsAndCosts(value, mapper) {
  const {
    costs: { networkFee, partnerFee, protocolFee }
  } = value;
  function serializeAmounts(value2) {
    return {
      sellAmount: mapper(value2.sellAmount),
      buyAmount: mapper(value2.buyAmount)
    };
  }
  const beforeNetworkCosts = serializeAmounts(value.beforeNetworkCosts);
  return {
    ...value,
    costs: {
      ...value.costs,
      networkFee: {
        ...networkFee,
        amountInSellCurrency: mapper(networkFee.amountInSellCurrency),
        amountInBuyCurrency: mapper(networkFee.amountInBuyCurrency)
      },
      partnerFee: {
        ...partnerFee,
        amount: mapper(partnerFee.amount)
      },
      protocolFee: {
        ...protocolFee,
        amount: mapper(protocolFee.amount)
      }
    },
    beforeAllFees: serializeAmounts(value.beforeAllFees),
    beforeNetworkCosts,
    afterProtocolFees: beforeNetworkCosts,
    afterNetworkCosts: serializeAmounts(value.afterNetworkCosts),
    afterPartnerFees: serializeAmounts(value.afterPartnerFees),
    afterSlippage: serializeAmounts(value.afterSlippage),
    amountsToSign: serializeAmounts(value.amountsToSign)
  };
}
function getTradeParametersAfterQuote({
  quoteParameters,
  sellToken
}) {
  return { ...quoteParameters, sellToken };
}
function adjustEthFlowOrderParams(chainId, params) {
  return {
    ...params,
    sellToken: WRAPPED_NATIVE_CURRENCIES2[chainId].address
  };
}

// src/getEthFlowTransaction.ts
import {
  getGlobalAdapter,
  ContractFactory
} from "@cowprotocol/sdk-common";
async function getEthFlowTransaction(appDataKeccak256, _params, chainId, additionalParams = {}, paramSigner) {
  const signer = paramSigner ? getGlobalAdapter().createSigner(paramSigner) : getGlobalAdapter().signer;
  const { networkCostsAmount = "0", checkEthFlowOrderExists } = additionalParams;
  const from = await signer.getAddress();
  const slippageBps = _params.slippageBps ?? getDefaultSlippageBps(chainId, true);
  const params = {
    ...adjustEthFlowOrderParams(chainId, _params),
    slippageBps
  };
  const { quoteId } = params;
  const protocolOptions = {
    env: params.env,
    settlementContractOverride: params.settlementContractOverride,
    ethFlowContractOverride: params.ethFlowContractOverride
  };
  const contract = getEthFlowContract(signer, chainId, protocolOptions);
  const orderToSign = getOrderToSign(
    {
      chainId,
      isEthFlow: true,
      from,
      networkCostsAmount
    },
    params,
    appDataKeccak256
  );
  const orderId = await calculateUniqueOrderId(chainId, orderToSign, checkEthFlowOrderExists, protocolOptions);
  const ethOrderParams = {
    buyToken: orderToSign.buyToken,
    receiver: orderToSign.receiver,
    sellAmount: orderToSign.sellAmount,
    buyAmount: orderToSign.buyAmount,
    feeAmount: orderToSign.feeAmount,
    partiallyFillable: orderToSign.partiallyFillable,
    quoteId,
    appData: appDataKeccak256,
    validTo: orderToSign.validTo.toString()
  };
  const estimatedGas = contract.estimateGas.createOrder ? await contract.estimateGas.createOrder(ethOrderParams, { value: orderToSign.sellAmount }).then((res) => {
    return BigInt(res.toHexString ? res.toHexString() : `0x${res.toString(16)}`);
  }).catch((error) => {
    console.error(error);
    return GAS_LIMIT_DEFAULT;
  }) : GAS_LIMIT_DEFAULT;
  const data = contract.interface.encodeFunctionData("createOrder", [ethOrderParams]);
  return {
    orderId,
    orderToSign,
    transaction: {
      data,
      gasLimit: "0x" + calculateGasMargin(estimatedGas).toString(16),
      to: contract.address,
      value: "0x" + BigInt(orderToSign.sellAmount).toString(16)
    }
  };
}
function getEthFlowContract(signer, chainId, options) {
  const { env, ethFlowContractOverride } = options ?? {};
  const address = ethFlowContractOverride?.[chainId] ?? (env === "staging" ? BARN_ETH_FLOW_ADDRESSES2[chainId] : ETH_FLOW_ADDRESSES2[chainId]);
  return ContractFactory.createEthFlowContract(address, signer);
}

// src/postSellNativeCurrencyOrder.ts
import { log } from "@cowprotocol/sdk-common";
async function postSellNativeCurrencyOrder(orderBookApi, appData, _params, additionalParams = {}, paramSigner) {
  const signer = paramSigner ? getGlobalAdapter2().createSigner(paramSigner) : getGlobalAdapter2().signer;
  const { appDataKeccak256, fullAppData } = appData;
  const { orderId, transaction, orderToSign } = await getEthFlowTransaction(
    appDataKeccak256,
    {
      ..._params,
      env: _params.env ?? orderBookApi.context.env,
      ethFlowContractOverride: _params.ethFlowContractOverride
    },
    orderBookApi.context.chainId,
    additionalParams,
    signer
  );
  log("Uploading app-data");
  await orderBookApi.uploadAppData(appDataKeccak256, fullAppData);
  log("Sending on-chain order transaction");
  const txReceipt = await signer.sendTransaction(transaction);
  log(`On-chain order transaction sent, txHash: ${txReceipt.hash}, order: ${orderId}`);
  return { txHash: txReceipt.hash, orderId, orderToSign, signature: "", signingScheme: SigningScheme2.EIP1271 };
}

// src/postCoWProtocolTrade.ts
import { log as log2 } from "@cowprotocol/sdk-common";

// src/utils/resolveSigner.ts
import { getGlobalAdapter as getGlobalAdapter3 } from "@cowprotocol/sdk-common";
function resolveSigner(signer) {
  const adapter = getGlobalAdapter3();
  return signer ? adapter.createSigner(signer) : adapter.signer;
}

// src/postCoWProtocolTrade.ts
async function postCoWProtocolTrade(orderBookApi, appData, params, additionalParams = {}, paramSigner) {
  const signer = resolveSigner(paramSigner);
  const {
    networkCostsAmount = "0",
    signingScheme: _signingScheme = SigningScheme3.EIP712,
    customEIP1271Signature,
    applyCostsSlippageAndFees
  } = additionalParams;
  const isEthFlow = getIsEthFlowOrder(params);
  if (isEthFlow) {
    const quoteId2 = params.quoteId;
    if (typeof quoteId2 === "number") {
      return postSellNativeCurrencyOrder(orderBookApi, appData, { ...params, quoteId: quoteId2 }, additionalParams, paramSigner);
    } else {
      throw new Error("quoteId is required for EthFlow orders");
    }
  }
  const { quoteId = null, owner } = params;
  const { appDataKeccak256, fullAppData } = appData;
  const env = params.env ?? orderBookApi.context.env;
  const settlementContractOverride = params.settlementContractOverride;
  const chainId = orderBookApi.context.chainId;
  const from = owner || await signer.getAddress();
  const orderToSign = getOrderToSign(
    { chainId, from, networkCostsAmount, isEthFlow, applyCostsSlippageAndFees },
    params,
    appData.appDataKeccak256
  );
  log2("Uploading app-data");
  await orderBookApi.uploadAppData(appDataKeccak256, fullAppData);
  log2("Signing order...");
  const { signature, signingScheme } = await (async () => {
    if (_signingScheme === SigningScheme3.PRESIGN) {
      return { signature: from, signingScheme: SigningScheme3.PRESIGN };
    } else {
      const isEip1271 = _signingScheme === SigningScheme3.EIP1271;
      if (isEip1271 && customEIP1271Signature) {
        return {
          signature: await customEIP1271Signature(orderToSign, signer),
          signingScheme: _signingScheme
        };
      }
      const signingResult = await OrderSigningUtils2.signOrder(
        orderToSign,
        chainId,
        signer,
        { env, settlementContractOverride }
      );
      if (isEip1271) {
        return {
          signature: OrderSigningUtils2.getEip1271Signature(orderToSign, signingResult.signature),
          signingScheme: _signingScheme
        };
      }
      return { signature: signingResult.signature, signingScheme: SIGN_SCHEME_MAP[signingResult.signingScheme] };
    }
  })();
  const orderBody = {
    ...orderToSign,
    from,
    signature,
    signingScheme,
    quoteId,
    appData: fullAppData,
    appDataHash: appDataKeccak256
  };
  log2("Posting order...");
  const orderId = await orderBookApi.sendOrder(orderBody);
  log2(`Order created, id: ${orderId}`);
  return { orderId, signature, signingScheme, orderToSign };
}

// src/getQuote.ts
import { getGlobalAdapter as getGlobalAdapter5, log as log4 } from "@cowprotocol/sdk-common";
import {
  getQuoteAmountsAndCosts as getQuoteAmountsAndCosts4,
  OrderBookApi as OrderBookApi3,
  OrderQuoteSideKindBuy,
  OrderQuoteSideKindSell,
  PriceQuality as PriceQuality2,
  SigningScheme as SigningScheme4
} from "@cowprotocol/sdk-order-book";

// src/appDataUtils.ts
import {
  MetadataApi,
  stringifyDeterministic
} from "@cowprotocol/sdk-app-data";
import { getGlobalAdapter as getGlobalAdapter4 } from "@cowprotocol/sdk-common";
import deepmerge from "deepmerge";

// ../sdk/package.json
var package_default = {
  name: "@cowprotocol/cow-sdk",
  version: "8.0.4",
  license: "(MIT OR Apache-2.0)",
  description: "CoW Protocol SDK - get quote, configure your order, and trade",
  repository: {
    type: "git",
    url: "https://github.com/cowprotocol/cow-sdk.git",
    directory: "packages/sdk"
  },
  main: "./dist/index.js",
  module: "./dist/index.mjs",
  types: "./dist/index.d.ts",
  exports: {
    ".": {
      types: "./dist/index.d.ts",
      import: "./dist/index.mjs",
      require: "./dist/index.js"
    }
  },
  sideEffects: false,
  files: [
    "dist/**"
  ],
  publishConfig: {
    access: "public"
  },
  scripts: {
    build: "tsup src/index.ts --format esm,cjs --dts --treeshake",
    postbuild: "pnpm copy-package-json && pnpm copy-md-files",
    "copy-package-json": "npx shx cp package.json dist",
    "copy-md-files": 'npx cpx "*.md" dist',
    lint: "eslint src/**/*.ts",
    typecheck: "tsc --noEmit",
    clean: "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    prepublishOnly: "npm run build"
  },
  devDependencies: {
    "@cow-sdk/typescript-config": "workspace:*",
    "@types/node": "^20.17.31",
    jest: "^29.7.0",
    shx: "^0.3.4",
    "ts-node": "^10.8.2",
    tsup: "^7.2.0",
    typescript: "^5.2.2"
  },
  dependencies: {
    "@cowprotocol/sdk-app-data": "workspace:*",
    "@cowprotocol/sdk-common": "workspace:*",
    "@cowprotocol/sdk-config": "workspace:*",
    "@cowprotocol/sdk-contracts-ts": "workspace:*",
    "@cowprotocol/sdk-order-book": "workspace:*",
    "@cowprotocol/sdk-order-signing": "workspace:*",
    "@cowprotocol/sdk-trading": "workspace:*"
  },
  peerDependencies: {
    "@openzeppelin/merkle-tree": "^1.x",
    "cross-fetch": "^3.x",
    "ipfs-only-hash": "^4.x",
    multiformats: "^9.x"
  },
  peerDependenciesMeta: {
    "cross-fetch": {
      optional: false
    },
    "ipfs-only-hash": {
      optional: true
    },
    multiformats: {
      optional: true
    },
    "@openzeppelin/merkle-tree": {
      optional: true
    }
  },
  keywords: [
    "cow",
    "cow-protocol",
    "sdk",
    "crypto",
    "typescript",
    "subgraph"
  ],
  bugs: {
    url: "https://github.com/cowprotocol/cow-sdk/issues"
  },
  homepage: "https://github.com/cowprotocol/cow-sdk#readme"
};

// src/appDataUtils.ts
function getDefaultUtmParams() {
  return {
    utmCampaign: "developer-cohort",
    utmContent: "",
    utmMedium: `cow-sdk@${package_default.version}`,
    utmSource: "cowmunity",
    utmTerm: "js"
  };
}
async function buildAppData({ slippageBps, appCode, orderClass: orderClassName, partnerFee }, advancedParams) {
  const quoteParams = { slippageBips: slippageBps };
  const orderClass = { orderClass: orderClassName };
  const metadataApiSdk = new MetadataApi(getGlobalAdapter4());
  const shouldAddDefaultUtm = !advancedParams?.metadata?.utm;
  const baseMetadata = {
    quote: quoteParams,
    orderClass,
    partnerFee,
    ...shouldAddDefaultUtm ? { utm: getDefaultUtmParams() } : {}
  };
  const doc = await metadataApiSdk.generateAppDataDoc(
    deepmerge(
      {
        appCode,
        metadata: baseMetadata
      },
      advancedParams || {}
    )
  );
  const { fullAppData, appDataKeccak256 } = await generateAppDataFromDoc(doc);
  return { doc, fullAppData, appDataKeccak256 };
}
async function generateAppDataFromDoc(doc) {
  const adapter = getGlobalAdapter4();
  const fullAppData = await stringifyDeterministic(doc);
  const appDataKeccak256 = adapter.utils.keccak256(adapter.utils.toUtf8Bytes(fullAppData));
  return { fullAppData, appDataKeccak256 };
}
async function mergeAppDataDoc(_doc, appDataOverride) {
  const doc = {
    ..._doc,
    metadata: {
      ..._doc.metadata,
      ...appDataOverride.metadata?.hooks ? { hooks: {} } : {},
      ...appDataOverride.metadata?.userConsents ? { userConsents: [] } : {}
    }
  };
  const appData = appDataOverride ? deepmerge(doc, appDataOverride) : doc;
  const { fullAppData, appDataKeccak256 } = await generateAppDataFromDoc(appData);
  return {
    fullAppData,
    appDataKeccak256,
    doc: appData
  };
}

// src/getOrderTypedData.ts
import { COW_EIP712_TYPES, ORDER_PRIMARY_TYPE, OrderSigningUtils as OrderSigningUtils3 } from "@cowprotocol/sdk-order-signing";
var EIP712DomainTypes = [
  { name: "name", type: "string" },
  { name: "version", type: "string" },
  { name: "chainId", type: "uint256" },
  { name: "verifyingContract", type: "address" }
];
async function getOrderTypedData(chainId, orderToSign, options) {
  const domain = await OrderSigningUtils3.getDomain(chainId, options);
  return {
    domain,
    primaryType: ORDER_PRIMARY_TYPE,
    types: {
      ...COW_EIP712_TYPES,
      EIP712Domain: EIP712DomainTypes
    },
    message: orderToSign
  };
}

// src/resolveSlippageSuggestion.ts
import { bpsToPercentage, log as log3 } from "@cowprotocol/sdk-common";
import { getQuoteAmountsAndCosts as getQuoteAmountsAndCosts3, PriceQuality } from "@cowprotocol/sdk-order-book";

// src/suggestSlippageBps.ts
import { percentageToBps } from "@cowprotocol/sdk-common";
import { getQuoteAmountsAndCosts as getQuoteAmountsAndCosts2, OrderKind } from "@cowprotocol/sdk-order-book";

// src/suggestSlippageFromFee.ts
import { applyPercentage } from "@cowprotocol/sdk-common";
function suggestSlippageFromFee(params) {
  const { feeAmount, multiplyingFactorPercent } = params;
  if (feeAmount < 0n) {
    throw new Error("Fee amount must be non-negative: " + feeAmount);
  }
  if (multiplyingFactorPercent < 0) {
    throw new Error("multiplyingFactorPercent must be non-negative: " + multiplyingFactorPercent);
  }
  return applyPercentage(feeAmount, multiplyingFactorPercent);
}

// src/suggestSlippageFromVolume.ts
import { applyPercentage as applyPercentage2 } from "@cowprotocol/sdk-common";
function suggestSlippageFromVolume(params) {
  const { sellAmountBeforeNetworkCosts, sellAmountAfterNetworkCosts, isSell, slippagePercent } = params;
  const sellAmount = isSell ? sellAmountAfterNetworkCosts : sellAmountBeforeNetworkCosts;
  if (sellAmount <= 0n) {
    throw new Error("sellAmount must be greater than 0: " + sellAmount);
  }
  if (slippagePercent < 0) {
    throw new Error("slippagePercent must be non-negative: " + slippagePercent);
  }
  return applyPercentage2(sellAmount, slippagePercent);
}

// src/suggestSlippageBps.ts
var MAX_SLIPPAGE_BPS = 1e4;
var SLIPPAGE_FEE_MULTIPLIER_PERCENT = 50;
var SLIPPAGE_VOLUME_MULTIPLIER_PERCENT = 0.5;
function suggestSlippageBps(params) {
  const { quote, trader, isEthFlow, volumeMultiplierPercent = SLIPPAGE_VOLUME_MULTIPLIER_PERCENT } = params;
  const isSell = quote.quote.kind === OrderKind.SELL;
  const {
    beforeNetworkCosts: { sellAmount: sellAmountBeforeNetworkCosts },
    afterNetworkCosts: { sellAmount: sellAmountAfterNetworkCosts }
  } = getQuoteAmountsAndCosts2({
    orderParams: quote.quote,
    protocolFeeBps: quote.protocolFeeBps ? Number(quote.protocolFeeBps) : 0,
    partnerFeeBps: void 0,
    slippagePercentBps: 0
  });
  const { feeAmount: feeAmountString } = quote.quote;
  const feeAmount = BigInt(feeAmountString);
  const slippageBpsFromFee = suggestSlippageFromFee({
    feeAmount,
    multiplyingFactorPercent: SLIPPAGE_FEE_MULTIPLIER_PERCENT
  });
  const slippageBpsFromVolume = suggestSlippageFromVolume({
    isSell,
    sellAmountBeforeNetworkCosts,
    sellAmountAfterNetworkCosts,
    slippagePercent: volumeMultiplierPercent
  });
  const totalSlippageBps = slippageBpsFromFee + slippageBpsFromVolume;
  const slippagePercent = getSlippagePercent({
    isSell,
    sellAmountBeforeNetworkCosts,
    sellAmountAfterNetworkCosts,
    slippage: totalSlippageBps
  });
  const slippageBps = percentageToBps(slippagePercent);
  const lowerCap = isEthFlow ? ETH_FLOW_DEFAULT_SLIPPAGE_BPS[trader.chainId] : 0;
  return Math.max(Math.min(slippageBps, MAX_SLIPPAGE_BPS), lowerCap);
}

// src/resolveSlippageSuggestion.ts
async function resolveSlippageSuggestion(chainId, tradeParameters, trader, quote, isEthFlow, advancedSettings) {
  const suggestSlippageParams = {
    isEthFlow,
    quote,
    tradeParameters,
    trader,
    advancedSettings
  };
  const getSlippageSuggestion = advancedSettings?.getSlippageSuggestion;
  const priceQuality = advancedSettings?.quoteRequest?.priceQuality ?? PriceQuality.OPTIMAL;
  const defaultSuggestion = suggestSlippageBps(suggestSlippageParams);
  if (priceQuality === PriceQuality.FAST || !getSlippageSuggestion) {
    return { slippageBps: defaultSuggestion };
  }
  const amountsAndCosts = getQuoteAmountsAndCosts3({
    orderParams: quote.quote,
    slippagePercentBps: 0,
    partnerFeeBps: getPartnerFeeBps(tradeParameters.partnerFee),
    protocolFeeBps: quote.protocolFeeBps ? Number(quote.protocolFeeBps) : void 0
  });
  const { isSell, beforeAllFees, afterSlippage } = amountsAndCosts;
  try {
    const suggestedSlippage = await getSlippageSuggestion({
      chainId,
      sellToken: tradeParameters.sellToken,
      buyToken: tradeParameters.buyToken,
      sellAmount: isSell ? beforeAllFees.sellAmount : afterSlippage.sellAmount,
      buyAmount: isSell ? afterSlippage.buyAmount : beforeAllFees.buyAmount
    });
    const suggestedSlippageBps = suggestedSlippage.slippageBps;
    return {
      slippageBps: suggestedSlippageBps ? suggestSlippageBps({
        ...suggestSlippageParams,
        volumeMultiplierPercent: bpsToPercentage(suggestedSlippageBps)
      }) : defaultSuggestion
    };
  } catch (e) {
    log3(`getSlippageSuggestion() error: ${e.message || String(e)}`);
    return { slippageBps: defaultSuggestion };
  }
}

// src/getQuote.ts
var ETH_FLOW_AUX_QUOTE_PARAMS = {
  signingScheme: SigningScheme4.EIP1271,
  onchainOrder: true,
  // Ethflow orders are subsidized in the backend.
  // This means we can assume the verification gas costs are zero for the quote/fee estimation
  verificationGasLimit: 0
};
var PROTOCOL_FEE_BPS_MIN = 1e-4;
async function getQuoteRaw(_tradeParameters, trader, advancedSettings, _orderBookApi) {
  const { appCode, chainId, account: from } = trader;
  const isEthFlow = getIsEthFlowOrder(_tradeParameters);
  const tradeParameters = isEthFlow ? adjustEthFlowOrderParams(chainId, _tradeParameters) : _tradeParameters;
  const {
    sellToken,
    buyToken,
    amount,
    kind,
    partnerFee,
    validFor,
    validTo,
    slippageBps,
    env = "prod"
  } = tradeParameters;
  if (validTo !== void 0 && validFor !== void 0) {
    throw new Error(
      "Cannot specify both validFor and validTo. Use validFor for relative time or validTo for absolute time."
    );
  }
  const effectiveValidFor = validFor ?? DEFAULT_QUOTE_VALIDITY;
  log4(
    `getQuote for: Swap ${amount} ${sellToken} for ${buyToken} on chain ${chainId} with ${slippageBps !== void 0 ? `${slippageBps} BPS` : "AUTO"} slippage`
  );
  const orderBookApi = _orderBookApi || new OrderBookApi3({ chainId, env });
  const receiver = tradeParameters.receiver || from;
  const isSell = kind === "sell";
  log4("Building app data...");
  const defaultSlippageBps = getDefaultSlippageBps(chainId, isEthFlow);
  const slippageBpsOrDefault = slippageBps ?? defaultSlippageBps;
  const buildAppDataParams = {
    slippageBps: slippageBpsOrDefault,
    orderClass: "market",
    appCode,
    partnerFee
  };
  const appDataInfo = await buildAppData(buildAppDataParams, advancedSettings?.appData);
  const { appDataKeccak256, fullAppData } = appDataInfo;
  log4(`App data: appDataKeccak256=${appDataKeccak256} fullAppData=${fullAppData}`);
  const quoteRequest = {
    from,
    sellToken,
    buyToken,
    receiver,
    ...typeof validTo === "number" ? { validTo } : { validFor: effectiveValidFor },
    appData: fullAppData,
    appDataHash: appDataKeccak256,
    priceQuality: PriceQuality2.OPTIMAL,
    // Do not change this parameter because we rely on the fact that quote has id
    signingScheme: SigningScheme4.EIP712,
    ...isEthFlow ? ETH_FLOW_AUX_QUOTE_PARAMS : {},
    ...isSell ? { kind: OrderQuoteSideKindSell.SELL, sellAmountBeforeFee: amount } : { kind: OrderQuoteSideKindBuy.BUY, buyAmountAfterFee: amount },
    ...advancedSettings?.quoteRequest
  };
  log4("Getting quote...");
  const quote = await orderBookApi.getQuote(quoteRequest);
  if (quote.protocolFeeBps) {
    const protocolFeeBps = Number(quote.protocolFeeBps);
    if (Number.isNaN(protocolFeeBps) || protocolFeeBps < PROTOCOL_FEE_BPS_MIN) {
      quote.protocolFeeBps = "0";
    }
  }
  const { slippageBps: suggestedSlippageBps } = await resolveSlippageSuggestion(
    chainId,
    tradeParameters,
    trader,
    quote,
    isEthFlow,
    advancedSettings
  );
  const commonResult = {
    isEthFlow,
    quote,
    orderBookApi,
    suggestedSlippageBps: suggestedSlippageBps || defaultSlippageBps
  };
  if (slippageBps === void 0) {
    if (suggestedSlippageBps) {
      log4(
        `Suggested slippage is greater than ${defaultSlippageBps} BPS (default), using the suggested slippage (${suggestedSlippageBps} BPS)`
      );
      const newAppDataInfo = await buildAppData(
        {
          ...buildAppDataParams,
          slippageBps: suggestedSlippageBps
        },
        advancedSettings?.appData
      );
      log4(
        `App data with new suggested slippage: appDataKeccak256=${newAppDataInfo.appDataKeccak256} fullAppData=${newAppDataInfo.fullAppData}`
      );
      return {
        ...commonResult,
        appDataInfo: newAppDataInfo,
        tradeParameters: { ..._tradeParameters, slippageBps: suggestedSlippageBps },
        slippageBps: suggestedSlippageBps
      };
    } else {
      log4(`No suggested slippage. Using the default slippage (${defaultSlippageBps} BPS)`);
    }
  }
  return {
    ...commonResult,
    appDataInfo,
    // Return the original tradeParameters to not expose all intermediate changes
    tradeParameters: _tradeParameters,
    slippageBps: slippageBpsOrDefault
  };
}
async function getQuote(_tradeParameters, trader, advancedSettings, _orderBookApi) {
  const { quote, orderBookApi, tradeParameters, slippageBps, suggestedSlippageBps, appDataInfo, isEthFlow } = await getQuoteRaw(_tradeParameters, trader, advancedSettings, _orderBookApi);
  const { partnerFee } = tradeParameters;
  const { chainId, account: from } = trader;
  const env = _tradeParameters.env ?? trader.env;
  const settlementContractOverride = _tradeParameters.settlementContractOverride ?? trader.settlementContractOverride;
  const amountsAndCosts = getQuoteAmountsAndCosts4({
    orderParams: quote.quote,
    slippagePercentBps: slippageBps,
    partnerFeeBps: getPartnerFeeBps(partnerFee),
    protocolFeeBps: quote.protocolFeeBps ? Number(quote.protocolFeeBps) : void 0
  });
  const orderToSign = getOrderToSign(
    {
      chainId,
      from,
      networkCostsAmount: quote.quote.feeAmount,
      isEthFlow,
      protocolFeeBps: quote.protocolFeeBps ? Number(quote.protocolFeeBps) : void 0
    },
    swapParamsToLimitOrderParams(tradeParameters, quote),
    appDataInfo.appDataKeccak256
  );
  const orderTypedData = await getOrderTypedData(chainId, orderToSign, { env, settlementContractOverride });
  return {
    result: {
      tradeParameters,
      suggestedSlippageBps,
      amountsAndCosts,
      orderToSign,
      quoteResponse: quote,
      appDataInfo,
      orderTypedData
    },
    orderBookApi
  };
}
async function getTrader(swapParameters) {
  const signer = getGlobalAdapter5().signerOrNull();
  const account = swapParameters.owner || await signer?.getAddress();
  return {
    chainId: swapParameters.chainId,
    appCode: swapParameters.appCode,
    account,
    settlementContractOverride: swapParameters.settlementContractOverride
  };
}
async function getQuoteWithSigner(swapParameters, advancedSettings, orderBookApi) {
  const signer = resolveSigner(swapParameters.signer);
  const trader = await getTrader(swapParameters);
  const result = await getQuote(swapParameters, trader, advancedSettings, orderBookApi);
  return {
    result: {
      ...result.result,
      signer
    },
    orderBookApi: result.orderBookApi
  };
}

// src/utils/applySettingsToLimitTradeParameters.ts
function applySettingsToLimitTradeParameters(_params, advancedSettings) {
  const params = { ..._params };
  const appDataOverride = advancedSettings?.appData;
  const appDataSlippageOverride = appDataOverride?.metadata?.quote?.slippageBips;
  const partnerFeeOverride = appDataOverride?.metadata?.partnerFee;
  if (typeof appDataSlippageOverride !== "undefined") {
    params.slippageBps = appDataSlippageOverride;
  }
  if (partnerFeeOverride) {
    params.partnerFee = partnerFeeOverride;
  }
  if (advancedSettings?.quoteRequest) {
    const { receiver, validTo, sellToken, buyToken, from } = advancedSettings.quoteRequest;
    if (receiver)
      params.receiver = receiver;
    if (validTo)
      params.validTo = validTo;
    if (sellToken)
      params.sellToken = sellToken;
    if (buyToken)
      params.buyToken = buyToken;
    if (from)
      params.owner = from;
  }
  if (!params.env) {
    params.env = "prod";
  }
  return params;
}

// src/postSwapOrder.ts
async function postSwapOrder(params, advancedSettings, orderBookApi) {
  return postSwapOrderFromQuote(await getQuoteWithSigner(params, advancedSettings, orderBookApi), advancedSettings);
}
async function postSwapOrderFromQuote({
  orderBookApi,
  result: { signer, appDataInfo: _appDataInfo, quoteResponse, tradeParameters }
}, advancedSettings) {
  const appDataOverride = advancedSettings?.appData;
  const appDataInfo = appDataOverride ? await mergeAppDataDoc(_appDataInfo.doc, appDataOverride) : _appDataInfo;
  const params = applySettingsToLimitTradeParameters(
    swapParamsToLimitOrderParams(tradeParameters, quoteResponse),
    advancedSettings
  );
  return postCoWProtocolTrade(
    orderBookApi,
    appDataInfo,
    params,
    {
      signingScheme: advancedSettings?.quoteRequest?.signingScheme,
      networkCostsAmount: quoteResponse.quote.feeAmount,
      ...advancedSettings?.additionalParams
    },
    signer
  );
}

// src/postLimitOrder.ts
import { OrderBookApi as OrderBookApi4 } from "@cowprotocol/sdk-order-book";
import { log as log5 } from "@cowprotocol/sdk-common";
async function postLimitOrder(_params, advancedSettings, _orderBookApi) {
  const params = applySettingsToLimitTradeParameters(_params, advancedSettings);
  if (!params.slippageBps) {
    params.slippageBps = 0;
  }
  const { appCode, chainId, sellToken, buyToken, sellAmount, buyAmount, partnerFee } = params;
  log5(`Limit order ${sellAmount} ${sellToken} for ${buyAmount} ${buyToken} on chain ${chainId}`);
  const orderBookApi = _orderBookApi || new OrderBookApi4({ chainId, env: params.env });
  log5("Building app data...");
  const appDataInfo = await buildAppData(
    {
      slippageBps: params.slippageBps,
      orderClass: "limit",
      appCode,
      partnerFee
    },
    advancedSettings?.appData
  );
  const additionalParams = {
    ...advancedSettings?.additionalParams ?? {},
    applyCostsSlippageAndFees: advancedSettings?.additionalParams?.applyCostsSlippageAndFees ?? false
  };
  return postCoWProtocolTrade(orderBookApi, appDataInfo, params, additionalParams, params.signer);
}

// src/getSettlementContract.ts
import { ContractFactory as ContractFactory2 } from "@cowprotocol/sdk-common";
import {
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS,
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING
} from "@cowprotocol/sdk-config";
function getSettlementContract(chainId, signer, options) {
  const { env, settlementContractOverride } = options ?? {};
  return ContractFactory2.createSettlementContract(
    settlementContractOverride?.[chainId] ?? (env === "staging" ? COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING[chainId] : COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS[chainId]),
    signer
  );
}

// src/getPreSignTransaction.ts
async function getPreSignTransaction(signer, chainId, orderId, options) {
  const contract = getSettlementContract(chainId, signer, options);
  const preSignatureCall = contract.interface.encodeFunctionData("setPreSignature", [orderId, true]);
  const gas = await contract.estimateGas.setPreSignature?.(orderId, true).catch(() => GAS_LIMIT_DEFAULT) || GAS_LIMIT_DEFAULT;
  return {
    data: preSignatureCall,
    gasLimit: "0x" + calculateGasMargin(gas).toString(16),
    to: contract.address,
    value: "0"
  };
}

// src/tradingSdk.ts
import {
  enableLogging,
  ERC20_ALLOWANCE_ABI,
  ERC20_APPROVE_ABI,
  getGlobalAdapter as getGlobalAdapter6,
  setGlobalAdapter
} from "@cowprotocol/sdk-common";
import { OrderSigningUtils as OrderSigningUtils4 } from "@cowprotocol/sdk-order-signing";

// src/onChainCancellation.ts
var CANCELLATION_GAS_LIMIT_DEFAULT = 150000n;
async function getEthFlowCancellation(ethFlowContract, order) {
  const cancelOrderParams = {
    buyToken: order.buyToken,
    receiver: order.receiver || order.owner,
    sellAmount: order.sellAmount,
    buyAmount: order.buyAmount,
    appData: order.appData.toString(),
    feeAmount: order.feeAmount,
    validTo: order.validTo.toString(),
    partiallyFillable: false,
    quoteId: 0
    // value doesn't matter, set to 0 for reducing gas costs
  };
  return getOnChainCancellation(ethFlowContract, cancelOrderParams);
}
async function getSettlementCancellation(settlementContract, order) {
  const cancelOrderParams = order.uid;
  return getOnChainCancellation(settlementContract, cancelOrderParams);
}
async function getOnChainCancellation(contract, cancelOrderParams) {
  const estimatedGas = await (async () => {
    try {
      if (contract.estimateGas.invalidateOrder) {
        const estimated = await contract.estimateGas.invalidateOrder(cancelOrderParams);
        return BigInt(estimated.toHexString ? estimated.toHexString() : `0x${estimated.toString(16)}`);
      }
      return CANCELLATION_GAS_LIMIT_DEFAULT;
    } catch (error) {
      console.error(error);
      return CANCELLATION_GAS_LIMIT_DEFAULT;
    }
  })();
  const data = contract.interface.encodeFunctionData("invalidateOrder", [cancelOrderParams]);
  return {
    estimatedGas,
    transaction: {
      data,
      gasLimit: "0x" + estimatedGas.toString(16),
      to: contract.address,
      value: "0x0"
    }
  };
}

// src/utils/resolveOrderBookApi.ts
import { OrderBookApi as OrderBookApi5 } from "@cowprotocol/sdk-order-book";
var orderBookApiCache = /* @__PURE__ */ new Map([
  ["prod", /* @__PURE__ */ new Map()],
  ["staging", /* @__PURE__ */ new Map()]
]);
function resolveOrderBookApi(chainId, env, existingOrderBookApi) {
  if (existingOrderBookApi) {
    existingOrderBookApi.context.chainId = chainId;
    existingOrderBookApi.context.env = env;
    return existingOrderBookApi;
  }
  const envCache = orderBookApiCache.get(env) ?? /* @__PURE__ */ new Map();
  const cached = envCache.get(chainId);
  if (cached)
    return cached;
  const orderBookApi = new OrderBookApi5({ chainId, env });
  envCache.set(chainId, orderBookApi);
  orderBookApiCache.set(env, envCache);
  return orderBookApi;
}

// src/tradingSdk.ts
import {
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS,
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING
} from "@cowprotocol/sdk-config";
var TradingSdk = class {
  constructor(traderParams = {}, options = {}, adapter) {
    this.traderParams = traderParams;
    this.options = options;
    if (options.enableLogging !== void 0) {
      enableLogging(options.enableLogging);
    }
    if (adapter) {
      setGlobalAdapter(adapter);
    }
  }
  setTraderParams(params) {
    this.traderParams = { ...this.traderParams, ...params };
    if (this.options.orderBookApi) {
      if (params.chainId) {
        this.options.orderBookApi.context.chainId = params.chainId;
      }
      if (params.env) {
        this.options.orderBookApi.context.env = params.env;
      }
    }
    return this;
  }
  async getQuote(params, advancedSettings) {
    const quoteResults = await getQuoteWithSigner(
      this.mergeParams(params),
      advancedSettings,
      this.resolveOrderBookApi(params)
    );
    return {
      quoteResults: quoteResults.result,
      postSwapOrderFromQuote: async (advancedSettings2, signingStepManager) => {
        await signingStepManager?.beforeOrderSign?.();
        return postSwapOrderFromQuote(
          {
            ...quoteResults,
            result: {
              ...quoteResults.result,
              tradeParameters: getTradeParametersAfterQuote({
                quoteParameters: quoteResults.result.tradeParameters,
                sellToken: params.sellToken
              }),
              // It's important to get a fresh instance of the signer
              // Because quote might be called with another signer
              signer: getGlobalAdapter6().signer
            }
          },
          advancedSettings2
        ).catch((error) => {
          signingStepManager?.onOrderSignError?.();
          throw error;
        }).then((result) => {
          signingStepManager?.afterOrderSign?.();
          return result;
        });
      }
    };
  }
  /**
   * Gets a quote without requiring a signer or wallet connection.
   * Useful for building UIs that preview swap quotes before the user connects a wallet.
   *
   * Unlike {@link getQuote}, this method only requires an `owner` address (the account
   * to quote for) instead of a signer. The returned result contains quote information
   * but no `postSwapOrderFromQuote` helper, since posting requires signing.
   *
   * @param params - Trade parameters with `owner` address. Signer is not needed.
   * @param advancedSettings - Optional advanced settings for the swap quote.
   * @returns Quote results including amounts, costs, slippage, and order data.
   *
   * @example
   * ```typescript
   * // Get a quote without a connected wallet
   * const sdk = new TradingSdk(
   *   { chainId: 1, appCode: 'My App' },
   *   { orderBookApi },
   * )
   *
   * const quoteResults = await sdk.getQuoteOnly({
   *   owner: '0x1234...', // any valid address
   *   kind: OrderKind.SELL,
   *   sellToken: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
   *   sellTokenDecimals: 6,
   *   buyToken: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
   *   buyTokenDecimals: 18,
   *   amount: '1000000000', // 1000 USDC
   * })
   *
   * console.log(quoteResults.amountsAndCosts)
   * ```
   */
  async getQuoteOnly(params, advancedSettings) {
    const quoterParams = this.mergeQuoterParams(params);
    const trader = {
      chainId: quoterParams.chainId,
      appCode: quoterParams.appCode,
      env: quoterParams.env,
      account: quoterParams.owner,
      settlementContractOverride: quoterParams.settlementContractOverride,
      ethFlowContractOverride: quoterParams.ethFlowContractOverride
    };
    const result = await getQuote(quoterParams, trader, advancedSettings, this.resolveOrderBookApi(params));
    return result.result;
  }
  async getQuoteResults(params, advancedSettings) {
    return getQuoteWithSigner(this.mergeParams(params), advancedSettings, this.resolveOrderBookApi(params));
  }
  async postSwapOrder(params, advancedSettings) {
    return postSwapOrder(this.mergeParams(params), advancedSettings, this.resolveOrderBookApi(params));
  }
  async postLimitOrder(params, advancedSettings) {
    return postLimitOrder(this.mergeParams(params), advancedSettings, this.resolveOrderBookApi(params));
  }
  /**
   * Posts a sell order for native currency (e.g., ETH) using the EthFlow contract.
   * This method creates an on-chain transaction for selling native tokens.
   *
   * @param params - The trade parameters including token addresses and amounts
   * @param advancedSettings - Optional advanced settings for the swap
   * @returns Promise resolving to the order posting result with transaction hash and order ID
   *
   * @example
   * ```typescript
   * const parameters: TradeParameters = {
   *   kind: OrderKind.SELL,
   *   sellToken: '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE', // Native ETH
   *   sellTokenDecimals: 18,
   *   buyToken: '0x0625afb445c3b6b7b929342a04a22599fd5dbb59',
   *   buyTokenDecimals: 18,
   *   amount: '100000000000000000', // 0.1 ETH
   * }
   *
   * const { orderId, txHash } = await sdk.postSellNativeCurrencyOrder(parameters)
   * ```
   */
  async postSellNativeCurrencyOrder(params, advancedSettings) {
    const quoteResults = await getQuoteWithSigner(
      this.mergeParams(params),
      advancedSettings,
      this.resolveOrderBookApi(params)
    );
    const { tradeParameters, quoteResponse } = quoteResults.result;
    return postSellNativeCurrencyOrder(
      quoteResults.orderBookApi,
      quoteResults.result.appDataInfo,
      // Quote response always has an id
      swapParamsToLimitOrderParams(
        getTradeParametersAfterQuote({ quoteParameters: tradeParameters, sellToken: params.sellToken }),
        quoteResponse
      ),
      advancedSettings?.additionalParams,
      quoteResults.result.signer
    );
  }
  async getPreSignTransaction(params) {
    const { chainId, env, settlementContractOverride, orderUid, signer: signerLike } = this.mergeParams(params);
    const signer = resolveSigner(signerLike);
    return getPreSignTransaction(signer, chainId, orderUid, {
      env,
      settlementContractOverride
    });
  }
  async getOrder(params) {
    const orderBookApi = this.resolveOrderBookApi(params);
    return orderBookApi.getOrder(params.orderUid);
  }
  async offChainCancelOrder(params) {
    const orderBookApi = this.resolveOrderBookApi(params);
    const { env, chainId, settlementContractOverride, signer: signerLike } = this.mergeParams(params);
    const signer = resolveSigner(signerLike);
    const { orderUid } = params;
    const orderCancellationSigning = await OrderSigningUtils4.signOrderCancellations([orderUid], chainId, signer, {
      env,
      settlementContractOverride
    });
    await orderBookApi.sendSignedOrderCancellations({
      ...orderCancellationSigning,
      orderUids: [orderUid]
    });
    return true;
  }
  async onChainCancelOrder(params, _order) {
    const { env, chainId, settlementContractOverride, ethFlowContractOverride } = this.mergeParams(params);
    const order = _order ?? await this.getOrder(params);
    const isEthFlowOrder = !!order.onchainOrderData;
    const signer = params.signer ? getGlobalAdapter6().createSigner(params.signer) : getGlobalAdapter6().signer;
    const { transaction } = await (isEthFlowOrder ? getEthFlowCancellation(getEthFlowContract(signer, chainId, { env, ethFlowContractOverride }), order) : getSettlementCancellation(getSettlementContract(chainId, signer, { env, settlementContractOverride }), order));
    const txReceipt = await signer.sendTransaction(transaction);
    return txReceipt.hash;
  }
  /**
   * Checks the current allowance for the CoW Protocol Vault Relayer to spend an ERC-20 token.
   *
   * @param params - Parameters including token address and owner address
   * @returns Promise resolving to the current allowance amount as a bigint
   *
   * @example
   * ```typescript
   * const params = {
   *   tokenAddress: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', // USDC
   *   owner: '0x123...',
   *   chainId: 1,
   * }
   *
   * const allowance = await sdk.getCowProtocolAllowance(params)
   * console.log('Current allowance:', allowance.toString())
   * ```
   */
  async getCowProtocolAllowance(params) {
    const { env, chainId } = this.mergeParams(params);
    const adapter = getGlobalAdapter6();
    const vaultRelayerAddress = params.vaultRelayerAddress ?? (env === "staging" ? COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING[chainId] : COW_PROTOCOL_VAULT_RELAYER_ADDRESS[chainId]);
    return await adapter.readContract({
      address: params.tokenAddress,
      abi: ERC20_ALLOWANCE_ABI,
      functionName: "allowance",
      args: [params.owner, vaultRelayerAddress]
    });
  }
  /**
   * Approves the CoW Protocol Vault Relayer to spend a specified amount of an ERC-20 token.
   * This method creates an on-chain approval transaction.
   *
   * @param params - Parameters including token address and amount to approve
   * @returns Promise resolving to the transaction hash of the approval transaction
   *
   * @example
   * ```typescript
   * const params = {
   *   tokenAddress: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', // USDC
   *   amount: '1000000000', // 1000 USDC (6 decimals)
   *   chainId: 1,
   * }
   *
   * const txHash = await sdk.approveCowProtocol(params)
   * console.log('Approval transaction:', txHash)
   * ```
   */
  async approveCowProtocol(params) {
    const { env, chainId, signer: signerLike } = this.mergeParams(params);
    const adapter = getGlobalAdapter6();
    const signer = resolveSigner(signerLike);
    const vaultRelayerAddress = params.vaultRelayerAddress ?? (env === "staging" ? COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING[chainId] : COW_PROTOCOL_VAULT_RELAYER_ADDRESS[chainId]);
    const txParams = {
      to: params.tokenAddress,
      data: adapter.utils.encodeFunction(ERC20_APPROVE_ABI, "approve", [
        vaultRelayerAddress,
        "0x" + params.amount.toString(16)
      ])
    };
    const txReceipt = await signer.sendTransaction(txParams);
    return txReceipt.hash;
  }
  resolveOrderBookApi(params) {
    const chainId = params.chainId ?? this.traderParams.chainId;
    const env = params.env ?? this.traderParams.env ?? this.options.orderBookApi?.context?.env ?? "prod";
    if (!chainId) {
      throw new Error("Chain ID is missing in getOrder() call");
    }
    return resolveOrderBookApi(chainId, env, this.options.orderBookApi);
  }
  mergeParams(params) {
    const { chainId, signer, appCode, env, settlementContractOverride, ethFlowContractOverride } = params;
    const orderBookContext = this.options.orderBookApi?.context;
    const traderParams = {
      chainId: chainId ?? this.traderParams.chainId ?? orderBookContext?.chainId,
      signer: signer ?? this.traderParams.signer ?? getGlobalAdapter6().signer,
      appCode: appCode ?? this.traderParams.appCode,
      env: env ?? this.traderParams.env ?? orderBookContext?.env,
      settlementContractOverride: settlementContractOverride ?? this.traderParams.settlementContractOverride,
      ethFlowContractOverride: ethFlowContractOverride ?? this.traderParams.ethFlowContractOverride
    };
    assertTraderParams(traderParams);
    return {
      ...params,
      ...traderParams
    };
  }
  /**
   * Merges quoter-specific parameters (chainId, appCode, env, owner) without requiring a signer.
   * Used by quote-only operations that don't need signing capability.
   */
  mergeQuoterParams(params) {
    const chainId = params.chainId ?? this.traderParams.chainId;
    const appCode = params.appCode ?? this.traderParams.appCode;
    const env = params.env ?? this.traderParams.env ?? this.options.orderBookApi?.context?.env ?? "prod";
    const settlementContractOverride = params.settlementContractOverride ?? this.traderParams.settlementContractOverride;
    const ethFlowContractOverride = params.ethFlowContractOverride ?? this.traderParams.ethFlowContractOverride;
    if (!chainId) {
      throw new Error("Missing quoter parameters: chainId");
    }
    if (!appCode) {
      throw new Error("Missing quoter parameters: appCode");
    }
    return {
      ...params,
      chainId,
      appCode,
      env,
      settlementContractOverride,
      ethFlowContractOverride
    };
  }
};
function assertTraderParams(params) {
  if (!isTraderParameters(params)) {
    throw new Error("Missing trader parameters: " + getMissingTraderParams(params).join(", "));
  }
}
function getMissingTraderParams(params) {
  const missingParams = [];
  if (!params.chainId)
    missingParams.push("chainId");
  if (!params.signer)
    missingParams.push("signer");
  if (!params.appCode)
    missingParams.push("appCode");
  return missingParams;
}
function isTraderParameters(params) {
  return getMissingTraderParams(params).length === 0;
}
export {
  TradingSdk,
  buildAppData,
  calculateUniqueOrderId,
  generateAppDataFromDoc,
  getDefaultUtmParams,
  getEthFlowContract,
  getEthFlowTransaction,
  getOrderToSign,
  getPartnerFeeBps,
  getPreSignTransaction,
  getQuote,
  getQuoteWithSigner,
  getQuote as getQuoteWithoutSigner,
  getTradeParametersAfterQuote,
  mapQuoteAmountsAndCosts,
  mergeAppDataDoc,
  postCoWProtocolTrade,
  postLimitOrder,
  postSellNativeCurrencyOrder,
  postSwapOrder,
  postSwapOrderFromQuote,
  suggestSlippageBps,
  swapParamsToLimitOrderParams
};
