// @openocean.finance/limitorder-sdk/lib/utils/chain.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNativeToken = isNativeToken;
exports.getChainById = getChainById;
var chainObj = {
    "eth": {
        chainName: "Ethereum Mainnet",
        chainId: 1,
        blockExplorerUrl: "https://etherscan.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: ["ETH", "USDT", "USDC", "BUSD", "UNI", "C98", "LINK", "MATIC"],
        rpcUrls: [
            "https://1rpc.io/eth"
        ],
        compiler: 'EVM'
    },
    "bsc": {
        chainName: "Binance Smart Chain", chainId: 56, blockExplorerUrl: "https://bscscan.com/tx/",
        popularToken: ["BNB", "USDT", "BUSD", "CAKE", "C98", "BAKE", "MBOX"],
        nativeCurrency: { name: "BNB", symbol: "BNB", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://bsc-dataseed.binance.org", "https://bsc-dataseed1.binance.org"],
        compiler: 'EVM',
    },
    "solana": {
        compiler: 'SOL',
        chainName: "Solana Mainnet",
        blockExplorerUrl: "https://solscan.io/tx/",
        popularToken: ["SOL", "SNY", "USDT", "USDC", "RAY", "STEP"],
        rpcUrls: null,
        chainId: 7565164,
        nativeCurrency: {
            symbol: "SOL",
            id: "solana",
            address: "So11111111111111111111111111111111111111112",
            decimals: 9,
            gasAmount: 1000000000000000
        }
    },
    "starknet": {
        compiler: 'STARKNET',
        chainName: "Starknet Testnet", blockExplorerUrl: "https://starkscan.co/tx/", popularToken: ["ETH", "WBTC", "USDC", "USDT", "DAI"],
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "	0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7" },
        rpcUrls: null
    },
    "flow": {
        compiler: 'FLOW',
        chainName: "Flow Mainnet", blockExplorerUrl: "https://flowscan.org/transaction/", popularToken: ["SOL", "SNY", "USDT", "USDC", "RAY", "STEP"],
        rpcUrls: null,
        nativeCurrency: {}
    },
    "polygon": {
        compiler: 'EVM',
        chainName: "Polygon Mainnet", chainId: 137, blockExplorerUrl: "https://polygonscan.com/tx/", popularToken: ["USDT", "USDC", "MATIC", "AAVE", "DINO", "ADDY", "MIMATIC"],
        nativeCurrency: { name: "MATIC", symbol: "MATIC", decimals: 18, address: "0x0000000000000000000000000000000000001010" },
        rpcUrls: ["https://polygon.llamarpc.com"]
    },
    "polygon_zkevm": {
        compiler: 'EVM',
        chainName: "Polygon zkEVM", chainId: 1101, blockExplorerUrl: "https://zkevm.polygonscan.com/tx/", popularToken: ["USDT", "USDC", "MATIC", "AAVE", "DINO", "ADDY", "MIMATIC"],
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://rpc.ankr.com/polygon_zkevm"]
    },
    "avax": {
        compiler: 'EVM',
        chainName: "Avalanche", chainId: 43114, blockExplorerUrl: "https://cchain.explorer.avax.network/tx/", popularToken: ["AVAX", "PNG", "DAI.E", "ETH", "WAVAX", "JOE", "QI", "USDC.E"],
        nativeCurrency: { name: "AVAX", symbol: "AVAX", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://api.avax.network/ext/bc/C/rpc"]
    },
    "fantom": {
        compiler: 'EVM',
        chainName: "Fantom", chainId: 250, blockExplorerUrl: "https://explorer.fantom.network/tx/", popularToken: ["ETH", "USDT", "USDC", "SPIRIT", "1INCH", "C98", "FTM", "fUSDT", "MIM", "BOO"],
        nativeCurrency: { name: "FTM", symbol: "FTM", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpcapi.fantom.network"]
    },
    "sonic": {
        compiler: 'EVM',
        chainName: "Sonic Mainnet", chainId: 146, blockExplorerUrl: "https://sonicscan.org/tx/", popularToken: ["ETH", "USDT", "USDC"],
        nativeCurrency: { name: "Sonic", symbol: "S", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.soniclabs.com"]
    },
    "arbitrum": {
        compiler: 'EVM',
        chainName: "Arbitrum", chainId: 42161, blockExplorerUrl: "https://arbiscan.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://arb1.arbitrum.io/rpc"]
    },
    "xdai": {
        compiler: 'EVM',
        chainName: "Gnosis Mainnet",
        chainId: 100,
        blockExplorerUrl: "https://blockscout.com/poa/xdai/tx/", popularToken: ["ETH", "USDT", "USDC", "BUSD", "1INCH", "C98"],
        nativeCurrency: { name: "xDai", symbol: "XDAI", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.gnosischain.com", "https://rpc.xdaichain.com/"]
    },
    "boba": {
        compiler: 'EVM',
        chainName: "Boba Mainnet", chainId: 288, blockExplorerUrl: "https://blockexplorer.boba.network/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://mainnet.boba.network"]
    },
    "ont": {
        compiler: 'ONT',
        chainName: "Ontology Mainnet", blockExplorerUrl: "https://explorer.ont.io/transaction/",
        popularToken: [],
        nativeCurrency: {
            symbol: "ONG",
            id: "ong",
            address: "0x0000000000000000000000000000000000000000",
            decimals: 18,
            gasAmount: 1000000000000000
        },
        rpcUrls: null
    },
    "ontevm": {
        compiler: 'EVM',
        chainName: "Ontology Evm", chainId: 5851, blockExplorerUrl: "https://explorer.ont.io/testnet/tx/",
        nativeCurrency: { name: "ONG", symbol: "ONG", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://polaris1.ont.io:10339"]
    },
    "metis": {
        compiler: 'EVM',
        chainName: "Andromeda (Metis)", chainId: 1088, blockExplorerUrl: "https://andromeda-explorer.metis.io/tx/",
        nativeCurrency: { name: "METIS", symbol: "METIS", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://andromeda.metis.io/?owner=1088"]
    },
    "tron": {
        compiler: 'TRON',
        chainName: "Tron Mainnet",
        blockExplorerUrl: "https://tronscan.io/#/transaction/",
        popularToken: ["TRX", "WTRX", "ETH", "BTC", "USDT", "USDJ", "JST"],
        nativeCurrency: {
            symbol: "TRX",
            id: "tron",
            address: "T9yD14Nj9j7xAB4dbGeiX9h8unkKHxuWwb",
            decimals: 6,
            gasAmount: 1000000000000000
        },
        rpcUrls: null
    },
    "heco": {
        compiler: 'EVM',
        chainName: "Heco Mainnet",
        chainId: 128,
        blockExplorerUrl: "https://http-mainnet.hecochain.com/tx/",
        popularToken: ["HT", "ETH", "AAVE", "USDT", "USDC", "MDX", "DEP"],
        nativeCurrency: { name: "HT", symbol: "HT", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://http-mainnet.hecochain.com"]
    },
    "okex": {
        compiler: 'EVM',
        chainName: "OEC Mainnet", chainId: 66, blockExplorerUrl: "https://www.oklink.com/okexchain/",
        popularToken: ["OKT", "OKB", "USDT", "USDC", "BUSD", "BNB", "CHE"],
        nativeCurrency: { name: "OKT", symbol: "OKT", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://exchainrpc.okex.org"]
    },
    "optimism": {
        compiler: 'EVM',
        chainName: "Optimism",
        chainId: 10,
        blockExplorerUrl: "https://optimistic.etherscan.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://rpc.ankr.com/optimism", "https://optimism-mainnet.public.blastapi.io", "https://mainnet.optimism.io"]
    },
    "harmony": {
        compiler: 'EVM',
        chainName: "Harmony",
        chainId: 1666600000,
        blockExplorerUrl: "https://explorer.harmony.one/",
        nativeCurrency: { name: "ONE", symbol: "ONE", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: ["ETH", "USDT", "USDC", "BUSD", "1INCH", "C98"],
        rpcUrls: ["https://api.s0.t.hmny.io", "https://api.harmony.one"]
    },
    "dot": {
        compiler: 'EVM',
        chainName: "Polkadot", blockExplorerUrl: "", popularToken: ["ETH", "USDT", "USDC", "BUSD", "1INCH", "C98"],
        nativeCurrency: {}
    },
    "neo": {
        compiler: 'EVM',
        chainName: "Neo", blockExplorerUrl: "",
        popularToken: ["ETH", "USDT", "USDC", "BUSD", "1INCH", "C98"]
    },
    "aurora": {
        chainName: "Aurora Mainnet",
        chainId: 1313161554,
        blockExplorerUrl: "https://aurorascan.dev/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://mainnet.aurora.dev"
        ],
        compiler: 'EVM'
    },
    "cronos": {
        chainName: "Cronos Mainnet",
        chainId: 25,
        blockExplorerUrl: "https://cronos.org/explorer/tx/",
        nativeCurrency: {
            name: "CRO",
            symbol: "CRO",
            decimals: 18,
            address: "0x0000000000000000000000000000000000000000"
        },
        popularToken: [],
        rpcUrls: [
            "https://evm.cronos.org"
        ],
        compiler: 'EVM'
    },
    "moonriver": {
        chainName: "Moonriver Mainnet",
        chainId: 1285,
        blockExplorerUrl: "https://moonriver.moonscan.io",
        nativeCurrency: { name: "MOVR", symbol: "MOVR", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.moonriver.moonbeam.network"
        ],
        compiler: 'EVM'
    },
    "bsctest": {
        chainName: "BSC Test Network",
        chainId: 1285,
        blockExplorerUrl: "https://testnet.bscscan.com/",
        nativeCurrency: { name: "BNB", symbol: "BNB", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://data-seed-prebsc-1-s1.binance.org:8545"
        ],
        compiler: 'EVM'
    },
    "aptos": {
        compiler: 'APTOS',
        chainName: "Aptos Mainnet",
        blockExplorerUrl: "https://explorer.aptoslabs.com/txn/",
        nativeCurrency: {
            name: "Aptos",
            symbol: "APT",
            decimals: 8,
            address: "0x1"
        },
    },
    "near": {
        compiler: 'NEAR',
        chainName: "Near Mainnet",
        blockExplorerUrl: "https://nearblocks.io/txns/",
        nativeCurrency: {
            name: "Near",
            symbol: "Near",
            decimals: 24,
            address: "near.near"
        },
    },
    cosmos: {
        chainId: 'cosmoshub-4',
        chainName: 'Cosmos',
        rpcUrls: [
            ''
        ],
        blockExplorerUrl: '',
        compiler: 'COSMOS',
        nativeCurrency: {
            symbol: "ATOM",
            id: "cosmos",
            address: "",
            decimals: 18,
            gasAmount: 1000000000000000
        }
    },
    osmosis: {
        chainId: 'osmosis-1',
        chainName: 'Osmosis',
        rpcUrls: [
            'https://rpc.cosmos.directory/osmosis'
        ],
        blockExplorerUrl: 'https://www.mintscan.io/osmosis/txs/',
        compiler: 'COSMOS',
    },
    iris: {
        chainId: 'irishub-1',
        chainName: 'Iris',
        rpcUrls: [
            'https://rpc-iris.keplr.app'
        ],
        blockExplorerUrl: '',
        compiler: 'COSMOS',
    },
    "kava": {
        chainName: "Kava Mainnet",
        chainId: 2222,
        blockExplorerUrl: "https://explorer.kava.io/tx/",
        nativeCurrency: {
            name: "KAVA",
            symbol: "KAVA",
            decimals: 18,
            address: "0x0000000000000000000000000000000000000000"
        },
        popularToken: [],
        rpcUrls: [
            "https://evm.kava.io"
        ],
        compiler: 'EVM'
    },
    "celo": {
        chainName: "Celo Mainnet",
        chainId: 42220,
        blockExplorerUrl: "https://explorer.celo.org/mainnet/tx/",
        nativeCurrency: { name: "CELO", symbol: "CELO", decimals: 18, address: "" },
        popularToken: [],
        rpcUrls: [
            "https://forno.celo.org"
        ],
        compiler: 'EVM'
    },
    "klaytn": {
        chainName: "Klaytn Cypress Mainnet",
        chainId: 8217,
        blockExplorerUrl: "https://explorer.celo.org/tx/",
        nativeCurrency: { name: "KLAY", symbol: "KLAY", decimals: 18, address: "0x471EcE3750Da237f93B8E339c536989b8978a438" },
        popularToken: [],
        rpcUrls: [
            "https://public-node-api.klaytnapi.com/v1/cypress"
        ],
        compiler: 'EVM'
    },
    "zksync": {
        chainName: "zkSync Era Mainnet",
        chainId: 324,
        blockExplorerUrl: "https://explorer.zksync.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://mainnet.era.zksync.io"
        ],
        compiler: 'EVM'
    },
    "lineaTest": {
        chainName: "Linea Testnet",
        chainId: 59140,
        blockExplorerUrl: "https://explorer.goerli.linea.build/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.goerli.linea.build"
        ],
        compiler: 'EVM'
    },
    "linea": {
        chainName: "Linea Mainnet",
        chainId: 59144,
        blockExplorerUrl: "https://lineascan.build/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.linea.build"
        ],
        compiler: 'EVM'
    },
    "telos": {
        chainName: "Telos Mainnet",
        chainId: 40,
        blockExplorerUrl: "https://teloscan.io/tx/",
        nativeCurrency: { name: "TLOS", symbol: "TLOS", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: [],
        rpcUrls: [
            "https://mainnet.telos.net/evm"
        ],
        compiler: 'EVM'
    },
    "scroll": {
        chainName: "Scroll",
        chainId: 534352,
        blockExplorerUrl: "https://blockscout.scroll.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.scroll.io"
        ],
        compiler: 'EVM'
    },
    "base": {
        chainName: "Base",
        chainId: 8453,
        blockExplorerUrl: "https://basescan.org/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://developer-access-mainnet.base.org"
        ],
        compiler: 'EVM'
    },
    "opbnb": {
        chainName: "opBNB",
        chainId: 204,
        blockExplorerUrl: "https://mainnet.opbnbscan.com/tx/",
        nativeCurrency: { name: "BNB", symbol: "BNB", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://opbnb-mainnet-rpc.bnbchain.org"
        ],
        compiler: 'EVM'
    },
    "opbnbtest": {
        chainName: "opBNBTest",
        chainId: 5611,
        blockExplorerUrl: "https://opbnb-testnet.bscscan.com/tx/",
        nativeCurrency: { name: "BNB", symbol: "BNB", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://opbnb-testnet-rpc.bnbchain.org"
        ],
        compiler: 'EVM'
    },
    "mantle": {
        chainName: "Mantle",
        chainId: 5000,
        blockExplorerUrl: "https://explorer.mantle.xyz/tx/",
        nativeCurrency: { name: "MNT", symbol: "MNT", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: [],
        rpcUrls: [
            "https://mantle.publicnode.com"
        ],
        compiler: 'EVM'
    },
    "manta": {
        chainName: "manta",
        chainId: 169,
        blockExplorerUrl: "https://pacific-explorer.manta.network/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://pacific-rpc.manta.network/http"
        ],
        compiler: 'EVM'
    },
    "x1": {
        chainName: "X1",
        chainId: 195,
        blockExplorerUrl: "https://www.oklink.com/x1-test/tx/",
        nativeCurrency: { name: "OKB", symbol: "OKB", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: [],
        rpcUrls: [
            "https://testrpc.x1.tech"
        ],
        compiler: 'EVM'
    },
    "blast": {
        chainName: "Blast",
        chainId: 81457,
        blockExplorerUrl: "https://blastscan.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.blast.io/"
        ],
        compiler: 'EVM'
    },
    "mode": {
        chainName: "Mode",
        chainId: 34443,
        blockExplorerUrl: "https://explorer.mode.network/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://1rpc.io/mode"
        ],
        compiler: 'EVM'
    },
    "pulse": {
        chainName: "PulseChain",
        chainId: 369,
        blockExplorerUrl: "https://scan.pulsechain.com/tx/",
        nativeCurrency: { name: "PLS", symbol: "PLS", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.pulsechain.com"
        ],
        compiler: 'EVM'
    },
    "merlin": {
        chainName: "Merlin Chain",
        chainId: 4200,
        blockExplorerUrl: "https://scan.merlinchain.io/tx/",
        nativeCurrency: { name: "BTC", symbol: "BTC", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: [],
        rpcUrls: [
            "https://rpc.merlinchain.io"
        ],
        compiler: 'EVM'
    },
    "rootstock": {
        compiler: 'EVM',
        chainName: "Rootstock",
        chainId: 30,
        blockExplorerUrl: "https://explorer.rsk.co/tx/",
        nativeCurrency: { name: "RBTC", symbol: "RBTC", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        popularToken: ["RBTC", "RUSDT", "RIF"],
        rpcUrls: ["https://public-node.rsk.co"]
    },
    "sei": {
        compiler: 'EVM',
        chainName: "Sei Network",
        chainId: 1329,
        blockExplorerUrl: "https://seitrace.com/tx/",
        nativeCurrency: { name: "SEI", symbol: "SEI", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        popularToken: ["SEI"],
        rpcUrls: ["https://evm-rpc.sei-apis.com"]
    },
    "btc": {
        compiler: 'BTC',
        chainName: "BTC",
        chainId: 101010,
        blockExplorerUrl: "",
        nativeCurrency: { name: "BTC", symbol: "BTC", decimals: 6, address: "0x101010" },
        popularToken: ["BTC"],
        rpcUrls: []
    },
    "arbitrum_sepolia": {
        compiler: 'EVM',
        chainName: "Arbitrum Sepolia",
        chainId: 421614,
        blockExplorerUrl: "https://sepolia-explorer.arbitrum.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://arbitrum-sepolia.blockpi.network/v1/rpc/public"]
    },
    "gravity": {
        compiler: 'EVM',
        chainName: "Gravity Alpha Mainnet",
        chainId: 1625,
        blockExplorerUrl: "https://explorer.gravity.xyz/tx/",
        nativeCurrency: { name: "G", symbol: "G", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.gravity.xyz"]
    },
    "ape": {
        compiler: 'EVM',
        chainName: "ApeChain",
        chainId: 33139,
        blockExplorerUrl: "https://apescan.io/tx/",
        nativeCurrency: { name: "ApeCoin", symbol: "APE", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.apechain.com"]
    },
    "sui": {
        compiler: 'SUI',
        chainName: "SuiChain",
        chainId: 1000001,
        blockExplorerUrl: "https://suivision.xyz/txblock/",
        nativeCurrency: { name: "SUI", symbol: "SUI", decimals: 9, address: "0x2" },
        rpcUrls: []
    },
    "bera": {
        compiler: 'EVM',
        chainName: "Berachain", chainId: 80094, blockExplorerUrl: "https://berascan.com/tx/", popularToken: ["BERA", "USDT", "USDC"],
        nativeCurrency: { name: "Berachain BERA", symbol: "BERA", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.berachain.com"]
    },
    "uni": {
        compiler: 'EVM',
        chainName: "UniChain",
        chainId: 130,
        blockExplorerUrl: "https://uniscan.xyz/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://unichain.drpc.org"]
    },
    "flare": {
        compiler: 'EVM',
        chainName: "Flare",
        chainId: 14,
        blockExplorerUrl: "https://flarescan.com/tx/",
        nativeCurrency: { name: "FLR", symbol: "FLR", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.ankr.com/flare"]
    },
    "monad": {
        compiler: 'EVM',
        chainName: "Monad Testnet",
        chainId: 10143,
        blockExplorerUrl: "https://testnet.monadexplorer.com/tx/",
        nativeCurrency: { name: "MON", symbol: "MON", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://testnet-rpc.monad.xyz"]
    },
    "swell": {
        compiler: 'EVM',
        chainName: "Swellchain",
        chainId: 1923,
        blockExplorerUrl: "https://swellchainscan.io/tx/",
        nativeCurrency: { name: "ETH", symbol: "ETH", decimals: 18, address: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE" },
        rpcUrls: ["https://rpc.ankr.com/swell"]
    },
    "hyperevm": {
        compiler: 'EVM',
        chainName: "HyperEvm",
        chainId: 999,
        blockExplorerUrl: "https://purrsec.com/tx/",
        nativeCurrency: {
            name: "HYPE",
            symbol: "HYPE",
            decimals: 18,
            address: "0x0000000000000000000000000000000000000000"
        },
        rpcUrls: [
            "https://rpc.hyperliquid.xyz/evm",
            "https://rpc.hyperlend.finance",
        ]
    },
    "plume": {
        compiler: 'EVM',
        chainName: "Plume Mainnet",
        chainId: 98866,
        blockExplorerUrl: "https://explorer.plume.org/tx/",
        nativeCurrency: { name: "PLUME", symbol: "PLUME", decimals: 18, address: "0x0000000000000000000000000000000000000000" },
        rpcUrls: ["https://rpc.plume.org"]
    },
};
function isNativeToken(chainName, address) {
    if (chainObj[chainName] &&
        chainObj[chainName].nativeCurrency &&
        chainObj[chainName].nativeCurrency.address) {
        return (chainObj[chainName].nativeCurrency.address.toUpperCase() ===
            address.toUpperCase());
    }
    return false;
}
function getChainById(chainId) {
    for (var key in chainObj) {
        if (Object.prototype.hasOwnProperty.call(chainObj, key)) {
            var element = chainObj[key];
            if (element.chainId == chainId) {
                return __assign({ chainKey: key }, element);
            }
        }
    }
    return '';
}
