// @openocean.finance/limitorder-sdk/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitOrderNodeSdk = exports.OpenoceanLimitOrderSdk = exports.openoceanLimitOrderSdk = void 0;
var limit_order_sdk_1 = require("./utils/limit-order-sdk");
var ERC20_abi_1 = require("./asset/abi/ERC20_abi");
var utils_1 = require("./utils");
var chain_1 = require("./utils/chain");
var chart_1 = require("./utils/chart");
var limitOrderNodeSdk_1 = __importDefault(require("./limitOrderNodeSdk"));
exports.LimitOrderNodeSdk = limitOrderNodeSdk_1.default;
var ethers6_1 = require("ethers6");
var OpenoceanLimitOrderSdk = /** @class */ (function () {
    function OpenoceanLimitOrderSdk() {
    }
    OpenoceanLimitOrderSdk.prototype.getLimitContractAddress = function (chainName, mode) {
        var contractAddress = null;
        if (mode == 'Fusion' || mode == 'fusion') {
            contractAddress = {
                bsc: "0xe1c264269970bffbc831449f5aafcd4cecfcadff",
            }[chainName];
        }
        else if (mode == 'Dca' || mode == 'dca') {
            contractAddress = {
                eth: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
            }[chainName] || '0x6cBB2598881940D08d5Ea3fA8F557E02996e1031';
        }
        else {
            contractAddress = {
                eth: "0xcC8d695603ce0b43D352891892FcC716c6a7C9f4",
                bsc: "0xA8A0213bb2ce671E457Ec14D08EB9d40E6DA8e2d",
                avax: "0x99b3488Ee3432bB60256140b4BD2488E3b6A705f",
                fantom: "0x44A632dC8ee03ad2cF5d530280a044DaED3E1ec0",
                polygon: "0xFA9B584Bc9543B66BeFdc41fb1DA8636edD7a697",
                arbitrum: "0x23C78B3d85b45BfA6DC8e09b517ba2d9b0ECCA8C",
                optimism: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                moonriver: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                harmony: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                heco: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                okex: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                xdai: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                cronos: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                zksync: "0x6a9115a77598cb9897347a6c31E6a164Ea06F870",
                linea: "0x4045734fe21c7B7E0cE516BE009780Cc2BA39A8f",
                base: "0xb5486f71c902fe0844bb07221fa8f47834d90b1b",
                sonic: "0xb45373129b4220160b92bd2320869f44d48ecd01",
                bera: "0x8D2B7e5501Eb6D92F8e349f2FEbe785DD070bE74",
                sei: "0xfE9a934A8607EF020aDf22D4431d6cE6005Aa4d3",
                hyperevm: "0x4E6b18217AC75A779262c20B3Cc07050cBe7282B"
            }[chainName];
        }
        if (!contractAddress) {
            throw new Error("Contract address not found for chain ".concat(chainName, " and mode ").concat(mode));
        }
        return contractAddress;
    };
    OpenoceanLimitOrderSdk.prototype.createLimitOrder = function (walletParams, orderParams) {
        return __awaiter(this, void 0, void 0, function () {
            var provider, ContractAddress, connector, limitOrderBuilder, limitOrderProtocolFacade, limitOrderPredicateBuilder, timeJson, minutes, expireTime, and, timestampBelow, simplePredicate, receiver, receiverInputData, interaction, limitOrder, limitOrderTypedData, callData, rawAllowanceHex, allowanceBN, _approveAmount, _makerAmount2, signature, orderHash, message, amountRate;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = walletParams.provider;
                        ContractAddress = this.getLimitContractAddress(walletParams.chainKey, walletParams.mode);
                        connector = new limit_order_sdk_1.Web3ProviderConnector(provider, walletParams.signer);
                        limitOrderBuilder = new limit_order_sdk_1.LimitOrderBuilder(ContractAddress, Number(walletParams.chainId), connector);
                        limitOrderProtocolFacade = new limit_order_sdk_1.LimitOrderProtocolFacade(ContractAddress, connector);
                        limitOrderPredicateBuilder = new limit_order_sdk_1.LimitOrderPredicateBuilder(limitOrderProtocolFacade);
                        timeJson = {
                            "10M": 10,
                            "1H": 60,
                            "1D": 1440,
                            "3D": 4320,
                            "7D": 10080,
                            "30D": 43200,
                            "3Month": 129600,
                            "6Month": 259200,
                            "1Y": 525600,
                            "2Y": 1051200,
                            "3Y": 1576800,
                        };
                        minutes = timeJson[orderParams.expire || '1H'];
                        expireTime = minutes * 60 * 1000;
                        and = limitOrderPredicateBuilder.and, timestampBelow = limitOrderPredicateBuilder.timestampBelow;
                        simplePredicate = and(timestampBelow(Math.round(Date.now() / 1000) + minutes * 60) // a limit order is valid only for 1 minute
                        );
                        receiver = orderParams.receiver, receiverInputData = orderParams.receiverInputData;
                        interaction = receiver ? receiver + (receiverInputData || "") : "0x";
                        limitOrder = limitOrderBuilder.buildLimitOrder({
                            makerAssetAddress: orderParams.makerTokenAddress,
                            takerAssetAddress: orderParams.takerTokenAddress,
                            makerAddress: walletParams.account,
                            makerAmount: orderParams.makerAmount,
                            takerAmount: orderParams.takerAmount,
                            predicate: simplePredicate,
                            permit: "0x",
                            interaction: interaction,
                            receiver: receiver
                        });
                        limitOrderTypedData = limitOrderBuilder.buildLimitOrderTypedData(limitOrder);
                        limitOrderTypedData.domain.name = "openocean Limit Order Protocol";
                        limitOrderTypedData.domain.version = "2";
                        if (!!(0, chain_1.isNativeToken)(walletParams.chainKey, orderParams.makerTokenAddress)) return [3 /*break*/, 3];
                        callData = connector.contractEncodeABI(ERC20_abi_1.ERC20_abi, orderParams.makerTokenAddress, "allowance", [walletParams.account, ContractAddress]);
                        return [4 /*yield*/, connector.ethCall(orderParams.makerTokenAddress, callData)];
                    case 1:
                        rawAllowanceHex = _a.sent();
                        allowanceBN = (0, ethers6_1.toBigInt)(rawAllowanceHex);
                        _approveAmount = +utils_1.utils.decimals2Amount(allowanceBN.toString(), orderParams.makerTokenDecimals);
                        _makerAmount2 = +utils_1.utils.decimals2Amount(orderParams.makerAmount, orderParams.makerTokenDecimals);
                        if (!(_approveAmount < _makerAmount2)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.sendEthApprove(orderParams.makerTokenAddress, "", ContractAddress, walletParams.account, connector, orderParams, orderParams.gasPrice)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [4 /*yield*/, limitOrderBuilder.buildOrderSignature(walletParams.account, limitOrderTypedData)];
                    case 4:
                        signature = _a.sent();
                        orderHash = limitOrderBuilder.buildLimitOrderHash(limitOrderTypedData);
                        message = limitOrderTypedData.message;
                        amountRate = (+utils_1.utils.decimals2Amount(orderParams.makerAmount, orderParams.makerTokenDecimals) /
                            +utils_1.utils.decimals2Amount(orderParams.takerAmount, orderParams.takerTokenDecimals)).toFixed(6);
                        return [2 /*return*/, {
                                makerAmount: orderParams.makerAmount,
                                takerAmount: orderParams.takerAmount,
                                signature: signature,
                                orderHash: orderHash,
                                orderMaker: walletParams.account,
                                remainingMakerAmount: orderParams.makerAmount,
                                data: message,
                                isActive: true,
                                chainId: walletParams.chainId,
                                expireTime: expireTime,
                                amountRate: amountRate,
                                interaction: interaction, // Default value 0x
                                receiver: receiver
                            }];
                }
            });
        });
    };
    OpenoceanLimitOrderSdk.prototype.cancelLimitOrder = function (walletParams, orderParams) {
        return __awaiter(this, void 0, void 0, function () {
            var provider, ContractAddress, connector, limitOrderProtocolFacade, callData, swapParams, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = walletParams.provider;
                        ContractAddress = this.getLimitContractAddress(walletParams.chainKey, walletParams.mode);
                        connector = new limit_order_sdk_1.Web3ProviderConnector(provider, walletParams.signer);
                        limitOrderProtocolFacade = new limit_order_sdk_1.LimitOrderProtocolFacade(ContractAddress, connector);
                        callData = limitOrderProtocolFacade.cancelLimitOrder(orderParams.orderData);
                        swapParams = {
                            from: walletParams.account,
                            to: ContractAddress,
                            data: callData,
                            // gas: 5000000,
                            gasPrice: orderParams.gasPrice,
                        };
                        return [4 /*yield*/, connector.sendTransaction(swapParams)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result];
                }
            });
        });
    };
    OpenoceanLimitOrderSdk.prototype.sendEthApprove = function (inToken, amount, toContract, account, connector, orderParams, gasPrice) {
        return __awaiter(this, void 0, void 0, function () {
            var maxAmount, callDataBalanceOf, rawBalanceOfHex, balanceOfBN, callDataTotalSupply, rawTotalSupplyHex, totalSupplyBN, error_1, amountToUse, callData, txParams, txHash;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(inToken === "0xfa17b330bcc4e7f3e2456996d89a5a54ab044831")) return [3 /*break*/, 2];
                        callDataBalanceOf = connector.contractEncodeABI(ERC20_abi_1.ERC20_abi, inToken, "balanceOf", [account]);
                        return [4 /*yield*/, connector.ethCall(inToken, callDataBalanceOf)];
                    case 1:
                        rawBalanceOfHex = _a.sent();
                        balanceOfBN = (0, ethers6_1.toBigInt)(rawBalanceOfHex);
                        maxAmount = balanceOfBN.toString();
                        return [3 /*break*/, 5];
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        callDataTotalSupply = connector.contractEncodeABI(ERC20_abi_1.ERC20_abi, inToken, "totalSupply", []);
                        return [4 /*yield*/, connector.ethCall(inToken, callDataTotalSupply)];
                    case 3:
                        rawTotalSupplyHex = _a.sent();
                        totalSupplyBN = (0, ethers6_1.toBigInt)(rawTotalSupplyHex);
                        maxAmount = totalSupplyBN.toString();
                        return [3 /*break*/, 5];
                    case 4:
                        error_1 = _a.sent();
                        maxAmount = (0, ethers6_1.toBigInt)("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
                        return [3 /*break*/, 5];
                    case 5:
                        amountToUse = amount || maxAmount;
                        callData = connector.contractEncodeABI(ERC20_abi_1.ERC20_abi, inToken, "approve", [toContract, amountToUse]);
                        txParams = {
                            from: account,
                            to: inToken,
                            data: callData,
                            gasPrice: gasPrice
                        };
                        return [4 /*yield*/, connector.sendTransaction(txParams)];
                    case 6:
                        txHash = _a.sent();
                        return [2 /*return*/, txHash];
                }
            });
        });
    };
    OpenoceanLimitOrderSdk.prototype.cancelOrderSignature = function (walletParams, orderParams) {
        return __awaiter(this, void 0, void 0, function () {
            var provider, web3ProviderConnector, message, signer;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = walletParams.provider;
                        web3ProviderConnector = new limit_order_sdk_1.Web3ProviderConnector(provider, walletParams.signer);
                        message = 'orderHash:\n' + orderParams.orderHash;
                        return [4 /*yield*/, web3ProviderConnector.signMessage(walletParams.account, message)];
                    case 1:
                        signer = _a.sent();
                        return [2 /*return*/, signer];
                }
            });
        });
    };
    OpenoceanLimitOrderSdk.prototype.loadChart = function (option) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, chart_1.chart.loadChart(option)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return OpenoceanLimitOrderSdk;
}());
exports.OpenoceanLimitOrderSdk = OpenoceanLimitOrderSdk;
var openoceanLimitOrderSdk = new OpenoceanLimitOrderSdk();
exports.openoceanLimitOrderSdk = openoceanLimitOrderSdk;
exports.default = openoceanLimitOrderSdk;
