// @openocean.finance/limitorder-sdk/lib/limitOrderNodeSdk.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var limit_order_sdk_1 = require("./utils/limit-order-sdk");
var ERC20_abi_1 = require("./asset/abi/ERC20_abi");
var utils_1 = require("./utils");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var chain_1 = require("./utils/chain");
var chart_1 = require("./utils/chart");
var ethers6_1 = require("ethers6");
var ERC20_abi1 = ERC20_abi_1.ERC20_abi;
var LimitOrderNodeSdk = /** @class */ (function () {
    function LimitOrderNodeSdk(chainId, provider, ethersSigner, mode) {
        this.chainId = chainId;
        this.provider = provider;
        this.ethersSigner = ethersSigner;
        this.mode = mode;
        this.isEthers = provider instanceof ethers6_1.JsonRpcProvider;
        if (this.isEthers) {
            this.ethersProvider = provider;
        }
        else {
            var web3 = provider;
            web3.eth.defaultAccount = ethersSigner.address;
        }
        this.signerAddress = ethersSigner.address;
        this.chain = (0, chain_1.getChainById)(chainId);
        this.limitOrderContractAddress = this.getLimitContractAddress(this.chain.chainKey, mode);
        this.builder = this.getLimitOrderBuilder(chainId, mode);
        this.orderFacade = this.getLimitOrderFacade();
    }
    LimitOrderNodeSdk.prototype.getLimitOrderBuilder = function (chainId, mode) {
        var providerConnector = this.getProviderConnector();
        return new limit_order_sdk_1.LimitOrderBuilder(this.getLimitContractAddress(this.chain.chainKey, mode), chainId, providerConnector);
    };
    LimitOrderNodeSdk.prototype.getProviderConnector = function () {
        if (this.isEthers && this.ethersProvider) {
            if (!this.ethersSigner) {
                throw new Error('Ethers signer not initialized');
            }
            return new limit_order_sdk_1.PrivateKeyProviderConnector('', this.ethersProvider, this.ethersSigner);
        }
        else {
            var web3 = this.provider;
            var walletAccount = web3.eth.accounts.wallet[this.signerAddress];
            if (!walletAccount) {
                throw new Error('Wallet account not found');
            }
            return new limit_order_sdk_1.PrivateKeyProviderConnector(walletAccount.privateKey.replace(/^0x/, ''), web3);
        }
    };
    LimitOrderNodeSdk.prototype.getLimitOrderFacade = function () {
        var providerConnector = this.getProviderConnector();
        return new limit_order_sdk_1.LimitOrderProtocolFacade(this.limitOrderContractAddress, providerConnector);
    };
    LimitOrderNodeSdk.prototype.getLimitContractAddress = function (chainName, mode) {
        if (mode == 'Fusion') {
            var address = {
                bsc: "0xe1c264269970bffbc831449f5aafcd4cecfcadff",
                // eth: "0xcC8d695603ce0b43D352891892FcC716c6a7C9f4",
                // avax: "0x99b3488Ee3432bB60256140b4BD2488E3b6A705f",
                // fantom: "0x44A632dC8ee03ad2cF5d530280a044DaED3E1ec0",
                // polygon: "0xFA9B584Bc9543B66BeFdc41fb1DA8636edD7a697",
                // arbitrum: "0x23C78B3d85b45BfA6DC8e09b517ba2d9b0ECCA8C",
                // optimism: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // moonriver: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // harmony: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // heco: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // okex: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // xdai: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // cronos: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                // zksync: "0x6a9115a77598cb9897347a6c31E6a164Ea06F870",
            }[chainName];
            if (address)
                return address;
        }
        else if (mode == 'Dca') {
            var address = {
                eth: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
                bsc: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
                // base: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
                sonic: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
                bera: "0x6cBB2598881940D08d5Ea3fA8F557E02996e1031",
            }[chainName];
            return address || '0x6cBB2598881940D08d5Ea3fA8F557E02996e1031';
        }
        else {
            var address = {
                eth: "0xcC8d695603ce0b43D352891892FcC716c6a7C9f4",
                bsc: "0xA8A0213bb2ce671E457Ec14D08EB9d40E6DA8e2d",
                avax: "0x99b3488Ee3432bB60256140b4BD2488E3b6A705f",
                fantom: "0x44A632dC8ee03ad2cF5d530280a044DaED3E1ec0",
                polygon: "0xFA9B584Bc9543B66BeFdc41fb1DA8636edD7a697",
                arbitrum: "0x23C78B3d85b45BfA6DC8e09b517ba2d9b0ECCA8C",
                optimism: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                moonriver: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                harmony: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                heco: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                okex: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                xdai: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                cronos: "0xc0A62DDCd284020dC883C642A3Ed5D2A1770eb91",
                zksync: "0x6a9115a77598cb9897347a6c31E6a164Ea06F870",
                linea: "0x4045734fe21c7B7E0cE516BE009780Cc2BA39A8f",
                base: "0xb5486f71c902fe0844bb07221fa8f47834d90b1b",
                sonic: "0xb45373129b4220160b92bd2320869f44d48ecd01",
                bera: "0x8D2B7e5501Eb6D92F8e349f2FEbe785DD070bE74",
            }[chainName];
            if (address)
                return address;
        }
        return "";
    };
    LimitOrderNodeSdk.prototype.createLimitOrder = function (orderParams) {
        return __awaiter(this, void 0, void 0, function () {
            var limitOrderPredicateBuilder, timeJson, minutes, expireTime, and, timestampBelow, simplePredicate, receiver, receiverInputData, interaction, limitOrder, limitOrderTypedData, contract, approveAmount, _approveAmount, _makerAmount2, web3, contract, approveAmount, _approveAmount, _makerAmount2, signature, orderHash, message, amountRate;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        limitOrderPredicateBuilder = new limit_order_sdk_1.LimitOrderPredicateBuilder(this.orderFacade);
                        timeJson = {
                            "10M": 10,
                            "1H": 60,
                            "1D": 1440,
                            "3D": 4320,
                            "7D": 10080,
                            "30D": 43200,
                            "3Month": 129600,
                            "6Month": 259200,
                            "1Y": 525600,
                            "2Y": 1051200,
                            "3Y": 1576800,
                        };
                        minutes = timeJson[orderParams.expire];
                        expireTime = minutes * 60 * 1000;
                        and = limitOrderPredicateBuilder.and, timestampBelow = limitOrderPredicateBuilder.timestampBelow;
                        simplePredicate = and(timestampBelow(Math.round(Date.now() / 1000) + minutes * 60) // a limit order is valid only for 1 minute
                        );
                        receiver = orderParams.receiver, receiverInputData = orderParams.receiverInputData;
                        interaction = receiver ? receiver + (receiverInputData || "") : "0x";
                        limitOrder = this.builder.buildLimitOrder({
                            makerAssetAddress: orderParams.makerTokenAddress,
                            takerAssetAddress: orderParams.takerTokenAddress,
                            makerAddress: this.signerAddress,
                            makerAmount: orderParams.makerAmount,
                            takerAmount: orderParams.takerAmount,
                            predicate: simplePredicate,
                            permit: "0x",
                            interaction: interaction,
                            receiver: receiver
                        });
                        limitOrderTypedData = this.builder.buildLimitOrderTypedData(limitOrder);
                        limitOrderTypedData.domain.name = "openocean Limit Order Protocol";
                        limitOrderTypedData.domain.version = "2";
                        if (!!(0, chain_1.isNativeToken)(this.chain.chainKey, orderParams.makerTokenAddress)) return [3 /*break*/, 7];
                        if (!(this.isEthers && this.ethersSigner)) return [3 /*break*/, 4];
                        contract = new ethers6_1.Contract(orderParams.makerTokenAddress, ERC20_abi1, this.ethersSigner);
                        return [4 /*yield*/, contract.allowance(this.signerAddress, this.limitOrderContractAddress)];
                    case 1:
                        approveAmount = _a.sent();
                        _approveAmount = +utils_1.utils.decimals2Amount(approveAmount, orderParams.makerTokenDecimals);
                        _makerAmount2 = +utils_1.utils.decimals2Amount(orderParams.makerAmount, orderParams.makerTokenDecimals);
                        console.log("_approveAmount", _approveAmount, _makerAmount2, orderParams.makerTokenDecimals);
                        if (!(_approveAmount < _makerAmount2)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.sendEthApprove(orderParams.makerTokenAddress, "1", "", orderParams.gasPrice, this.limitOrderContractAddress)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [3 /*break*/, 7];
                    case 4:
                        web3 = this.provider;
                        contract = new web3.eth.Contract(ERC20_abi1, orderParams.makerTokenAddress);
                        return [4 /*yield*/, contract.methods
                                .allowance(this.signerAddress, this.limitOrderContractAddress)
                                .call()];
                    case 5:
                        approveAmount = _a.sent();
                        _approveAmount = +utils_1.utils.decimals2Amount(approveAmount, orderParams.makerTokenDecimals);
                        _makerAmount2 = +utils_1.utils.decimals2Amount(orderParams.makerAmount, orderParams.makerTokenDecimals);
                        console.log("_approveAmount", _approveAmount, _makerAmount2, orderParams.makerTokenDecimals);
                        if (!(_approveAmount < _makerAmount2)) return [3 /*break*/, 7];
                        return [4 /*yield*/, this.sendEthApprove(orderParams.makerTokenAddress, "1", "", orderParams.gasPrice, this.limitOrderContractAddress)];
                    case 6:
                        _a.sent();
                        _a.label = 7;
                    case 7: return [4 /*yield*/, this.builder.buildOrderSignature(this.signerAddress, limitOrderTypedData)];
                    case 8:
                        signature = _a.sent();
                        orderHash = this.builder.buildLimitOrderHash(limitOrderTypedData);
                        message = limitOrderTypedData.message;
                        amountRate = (+utils_1.utils.decimals2Amount(orderParams.makerAmount, orderParams.makerTokenDecimals) /
                            +utils_1.utils.decimals2Amount(orderParams.takerAmount, orderParams.takerTokenDecimals)).toFixed(6);
                        return [2 /*return*/, {
                                makerAmount: orderParams.makerAmount,
                                takerAmount: orderParams.takerAmount,
                                signature: signature,
                                orderHash: orderHash,
                                orderMaker: this.signerAddress,
                                remainingMakerAmount: orderParams.makerAmount,
                                data: message,
                                isActive: true,
                                chainId: this.chain.chainId,
                                expireTime: expireTime,
                                amountRate: amountRate,
                                interaction: interaction
                            }];
                }
            });
        });
    };
    LimitOrderNodeSdk.prototype.cancelLimitOrder = function (orderParams) {
        return __awaiter(this, void 0, void 0, function () {
            var callData, tx, web3, swapParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        callData = this.orderFacade.cancelLimitOrder(orderParams.orderData);
                        if (!(this.isEthers && this.ethersSigner)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.ethersSigner.sendTransaction({
                                to: this.limitOrderContractAddress,
                                data: callData,
                                gasLimit: 5000000,
                                gasPrice: orderParams.gasPrice
                            })];
                    case 1:
                        tx = _a.sent();
                        return [4 /*yield*/, tx.wait()];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        web3 = this.provider;
                        swapParams = {
                            from: this.signerAddress,
                            to: this.limitOrderContractAddress,
                            data: callData,
                            gas: 5000000,
                            gasPrice: orderParams.gasPrice,
                        };
                        return [4 /*yield*/, (0, utils_1.confirmLater)(web3.eth.sendTransaction(swapParams))];
                    case 4: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    LimitOrderNodeSdk.prototype.sendEthApprove = function (inToken, type, amount, gasPrice, toContract) {
        return __awaiter(this, void 0, void 0, function () {
            var contract, maxAmount, approveAmount, gasEstimate, tx, web3, contract, maxAmount, approveAmount, gasAmount;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.isEthers && this.ethersSigner)) return [3 /*break*/, 8];
                        contract = new ethers6_1.Contract(inToken, ERC20_abi1, this.ethersSigner);
                        maxAmount = new bignumber_js_1.default("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff").toFixed(0, 1);
                        if (!(inToken === "0xfa17b330bcc4e7f3e2456996d89a5a54ab044831")) return [3 /*break*/, 2];
                        return [4 /*yield*/, contract.balanceOf(this.signerAddress)];
                    case 1:
                        maxAmount = _a.sent();
                        return [3 /*break*/, 4];
                    case 2:
                        if (!contract.totalSupply) return [3 /*break*/, 4];
                        return [4 /*yield*/, contract.totalSupply()];
                    case 3:
                        maxAmount = _a.sent();
                        console.log("maxAmount", maxAmount);
                        _a.label = 4;
                    case 4:
                        approveAmount = type === "1" ? maxAmount : amount;
                        return [4 /*yield*/, contract.approve.estimateGas(toContract, approveAmount)];
                    case 5:
                        gasEstimate = _a.sent();
                        return [4 /*yield*/, contract.approve(toContract, approveAmount, {
                                gasLimit: Math.floor(Number(gasEstimate) * 1.2),
                                gasPrice: gasPrice
                            })];
                    case 6:
                        tx = _a.sent();
                        return [4 /*yield*/, tx.wait()];
                    case 7: return [2 /*return*/, _a.sent()];
                    case 8:
                        web3 = this.provider;
                        contract = new web3.eth.Contract(ERC20_abi1, inToken);
                        maxAmount = new bignumber_js_1.default("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff").toFixed(0, 1);
                        if (!(inToken === "0xfa17b330bcc4e7f3e2456996d89a5a54ab044831")) return [3 /*break*/, 10];
                        return [4 /*yield*/, contract.methods.balanceOf(this.signerAddress).call()];
                    case 9:
                        maxAmount = _a.sent();
                        return [3 /*break*/, 12];
                    case 10:
                        if (!contract.methods.totalSupply) return [3 /*break*/, 12];
                        return [4 /*yield*/, contract.methods.totalSupply().call()];
                    case 11:
                        maxAmount = _a.sent();
                        console.log("maxAmount", maxAmount);
                        _a.label = 12;
                    case 12:
                        approveAmount = type === "1" ? maxAmount : amount;
                        return [4 /*yield*/, contract.methods
                                .approve(toContract, approveAmount)
                                .estimateGas({
                                from: this.signerAddress,
                            })];
                    case 13:
                        gasAmount = _a.sent();
                        return [4 /*yield*/, (0, utils_1.confirmLater)(contract.methods.approve(toContract, approveAmount).send({
                                from: this.signerAddress,
                                gasPrice: gasPrice,
                                gas: Math.floor(Number(gasAmount) * 1.2),
                            }))];
                    case 14: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    LimitOrderNodeSdk.prototype.loadChart = function (option) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, chart_1.chart.loadChart(option)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return LimitOrderNodeSdk;
}());
exports.default = LimitOrderNodeSdk;
