// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/limit-order-predicate.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitOrderPredicateBuilder = void 0;
var limit_order_protocol_model_1 = require("./model/limit-order-protocol.model");
var limit_order_protocol_const_1 = require("./limit-order-protocol.const");
var LimitOrderPredicateBuilder = /** @class */ (function () {
    function LimitOrderPredicateBuilder(facade) {
        var _this = this;
        this.facade = facade;
        this.and = function () {
            var predicates = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                predicates[_i] = arguments[_i];
            }
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.and, [
                predicates.map(function () { return _this.facade.contractAddress; }),
                predicates,
            ]);
        };
        this.or = function () {
            var predicates = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                predicates[_i] = arguments[_i];
            }
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.or, [
                predicates.map(function () { return _this.facade.contractAddress; }),
                predicates,
            ]);
        };
        this.eq = function (value, address, callData) {
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.eq, [
                value,
                address,
                callData,
            ]);
        };
        this.lt = function (value, address, callData) {
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.lt, [
                value,
                address,
                callData,
            ]);
        };
        this.gt = function (value, address, callData) {
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.gt, [
                value,
                address,
                callData,
            ]);
        };
        this.nonceEquals = function (makerAddress, makerNonce) {
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.nonceEquals, [makerAddress, makerNonce]);
        };
        this.timestampBelow = function (timestamp) {
            return _this.facade.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.timestampBelow, [limit_order_protocol_const_1.ZX + timestamp.toString(16)]);
        };
    }
    return LimitOrderPredicateBuilder;
}());
exports.LimitOrderPredicateBuilder = LimitOrderPredicateBuilder;
