// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/limit-order.builder.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitOrderBuilder = exports.generateRFQOrderInfo = exports.generateOrderSalt = void 0;
var limit_order_protocol_const_1 = require("./limit-order-protocol.const");
var limit_order_protocol_model_1 = require("./model/limit-order-protocol.model");
var ethers = require("ethers6");
function generateOrderSalt() {
    return Math.round(Math.random() * Date.now()) + '';
}
exports.generateOrderSalt = generateOrderSalt;
function generateRFQOrderInfo(id, expiresInTimestamp, wrapEth) {
    var info = (BigInt(expiresInTimestamp) << BigInt(64)) | BigInt(id);
    if (wrapEth) {
        return (info | (BigInt(1) << BigInt(255))).toString(10);
    }
    return info.toString(10);
}
exports.generateRFQOrderInfo = generateRFQOrderInfo;
var LimitOrderBuilder = /** @class */ (function () {
    function LimitOrderBuilder(contractAddress, chainId, providerConnector, generateSalt) {
        if (generateSalt === void 0) { generateSalt = generateOrderSalt; }
        this.contractAddress = contractAddress;
        this.chainId = chainId;
        this.providerConnector = providerConnector;
        this.generateSalt = generateSalt;
    }
    LimitOrderBuilder.prototype.buildOrderSignature = function (walletAddress, typedData) {
        return this.providerConnector.signTypedData(walletAddress, typedData);
    };
    LimitOrderBuilder.prototype.buildLimitOrderHash = function (orderTypedData) {
        var domain = orderTypedData.domain, types = orderTypedData.types, message = orderTypedData.message;
        var EIP712Domain = types.EIP712Domain, restTypes = __rest(types, ["EIP712Domain"]);
        var dataHash = ethers.TypedDataEncoder.hash(domain, restTypes, message);
        return dataHash;
    };
    LimitOrderBuilder.prototype.buildLimitOrderTypedData = function (order, domainName) {
        if (domainName === void 0) { domainName = limit_order_protocol_const_1.PROTOCOL_NAME; }
        return {
            primaryType: 'Order',
            types: {
                EIP712Domain: limit_order_protocol_const_1.EIP712_DOMAIN,
                Order: limit_order_protocol_const_1.ORDER_STRUCTURE,
            },
            domain: {
                name: domainName,
                version: limit_order_protocol_const_1.PROTOCOL_VERSION,
                chainId: this.chainId,
                verifyingContract: this.contractAddress,
            },
            message: order,
        };
    };
    LimitOrderBuilder.prototype.buildRFQOrderTypedData = function (order, domainName) {
        if (domainName === void 0) { domainName = limit_order_protocol_const_1.PROTOCOL_NAME; }
        return {
            primaryType: 'OrderRFQ',
            types: {
                EIP712Domain: limit_order_protocol_const_1.EIP712_DOMAIN,
                OrderRFQ: limit_order_protocol_const_1.RFQ_ORDER_STRUCTURE,
            },
            domain: {
                name: domainName,
                version: limit_order_protocol_const_1.PROTOCOL_VERSION,
                chainId: this.chainId,
                verifyingContract: this.contractAddress,
            },
            message: order,
        };
    };
    /* eslint-disable max-lines-per-function */
    LimitOrderBuilder.prototype.buildRFQOrder = function (_a) {
        var id = _a.id, _b = _a.wrapEth, wrapEth = _b === void 0 ? false : _b, expiresInTimestamp = _a.expiresInTimestamp, makerAssetAddress = _a.makerAssetAddress, takerAssetAddress = _a.takerAssetAddress, makerAddress = _a.makerAddress, _c = _a.takerAddress, takerAddress = _c === void 0 ? limit_order_protocol_const_1.ZERO_ADDRESS : _c, makerAmount = _a.makerAmount, takerAmount = _a.takerAmount;
        return {
            info: generateRFQOrderInfo(id, expiresInTimestamp, wrapEth),
            makerAsset: makerAssetAddress,
            takerAsset: takerAssetAddress,
            maker: makerAddress,
            allowedSender: takerAddress,
            makingAmount: makerAmount,
            takingAmount: takerAmount,
        };
    };
    /* eslint-enable max-lines-per-function */
    /* eslint-disable max-lines-per-function */
    LimitOrderBuilder.prototype.buildLimitOrder = function (_a) {
        var makerAssetAddress = _a.makerAssetAddress, takerAssetAddress = _a.takerAssetAddress, makerAddress = _a.makerAddress, _b = _a.receiver, receiver = _b === void 0 ? limit_order_protocol_const_1.ZERO_ADDRESS : _b, _c = _a.takerAddress, takerAddress = _c === void 0 ? limit_order_protocol_const_1.ZERO_ADDRESS : _c, makerAmount = _a.makerAmount, takerAmount = _a.takerAmount, _d = _a.predicate, predicate = _d === void 0 ? limit_order_protocol_const_1.ZX : _d, _e = _a.permit, permit = _e === void 0 ? limit_order_protocol_const_1.ZX : _e, _f = _a.interaction, interaction = _f === void 0 ? limit_order_protocol_const_1.ZX : _f;
        return {
            salt: this.generateSalt(),
            makerAsset: makerAssetAddress,
            takerAsset: takerAssetAddress,
            maker: makerAddress,
            receiver: receiver,
            allowedSender: takerAddress,
            makingAmount: makerAmount,
            takingAmount: takerAmount,
            makerAssetData: limit_order_protocol_const_1.ZX,
            takerAssetData: limit_order_protocol_const_1.ZX,
            getMakerAmount: this.getAmountData(limit_order_protocol_model_1.LimitOrderProtocolMethods.getMakerAmount, makerAmount, takerAmount),
            getTakerAmount: this.getAmountData(limit_order_protocol_model_1.LimitOrderProtocolMethods.getTakerAmount, makerAmount, takerAmount),
            predicate: predicate,
            permit: permit,
            interaction: interaction,
        };
    };
    /* eslint-enable max-lines-per-function */
    LimitOrderBuilder.prototype.getContractCallData = function (methodName, methodParams) {
        if (methodParams === void 0) { methodParams = []; }
        return this.providerConnector.contractEncodeABI(limit_order_protocol_const_1.LIMIT_ORDER_PROTOCOL_ABI, this.contractAddress, methodName, methodParams);
    };
    // Get nonce from contract (nonce method) and put it to predicate on order creating
    LimitOrderBuilder.prototype.getAmountData = function (methodName, makerAmount, takerAmount, swapTakerAmount) {
        if (swapTakerAmount === void 0) { swapTakerAmount = '0'; }
        return this.getContractCallData(methodName, [
            makerAmount,
            takerAmount,
            swapTakerAmount,
        ]).substr(0, 2 + 68 * 2);
    };
    return LimitOrderBuilder;
}());
exports.LimitOrderBuilder = LimitOrderBuilder;
