// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/limit-order-protocol.facade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitOrderProtocolFacade = void 0;
var limit_order_protocol_const_1 = require("./limit-order-protocol.const");
var limit_order_protocol_model_1 = require("./model/limit-order-protocol.model");
var bignumber_1 = require("@ethersproject/bignumber");
var get_rpc_code_1 = require("./utils/get-rpc-code");
var LimitOrderProtocolFacade = /** @class */ (function () {
    function LimitOrderProtocolFacade(contractAddress, providerConnector) {
        this.contractAddress = contractAddress;
        this.providerConnector = providerConnector;
    }
    LimitOrderProtocolFacade.prototype.fillLimitOrder = function (order, signature, makerAmount, takerAmount, thresholdAmount) {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.fillOrder, [
            order,
            signature,
            makerAmount,
            takerAmount,
            thresholdAmount,
        ]);
    };
    LimitOrderProtocolFacade.prototype.fillOrderToWithPermit = function (params) {
        var order = params.order, signature = params.signature, makerAmount = params.makerAmount, takerAmount = params.takerAmount, thresholdAmount = params.thresholdAmount, targetAddress = params.targetAddress, permit = params.permit;
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.fillOrderToWithPermit, [
            order,
            signature,
            makerAmount,
            takerAmount,
            thresholdAmount,
            targetAddress,
            permit,
        ]);
    };
    LimitOrderProtocolFacade.prototype.fillRFQOrder = function (order, signature, makerAmount, takerAmount) {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.fillOrderRFQ, [order, signature, makerAmount, takerAmount]);
    };
    LimitOrderProtocolFacade.prototype.cancelLimitOrder = function (order) {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.cancelOrder, [
            order,
        ]);
    };
    LimitOrderProtocolFacade.prototype.cancelRFQOrder = function (orderInfo) {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.cancelOrderRFQ, [orderInfo]);
    };
    LimitOrderProtocolFacade.prototype.nonce = function (makerAddress) {
        var callData = this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.nonce, [makerAddress]);
        return this.providerConnector
            .ethCall(this.contractAddress, callData)
            .then(function (nonce) { return bignumber_1.BigNumber.from(nonce).toNumber(); });
    };
    LimitOrderProtocolFacade.prototype.advanceNonce = function (count) {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.advanceNonce, [count]);
    };
    LimitOrderProtocolFacade.prototype.increaseNonce = function () {
        return this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.increaseNonce);
    };
    LimitOrderProtocolFacade.prototype.checkPredicate = function (order) {
        var callData = this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.checkPredicate, [order]);
        return this.providerConnector
            .ethCall(this.contractAddress, callData)
            .catch(function (error) {
            console.error(error);
            return false;
        })
            .then(function (result) {
            try {
                return bignumber_1.BigNumber.from(result).toNumber() === 1;
            }
            catch (e) {
                console.error(e);
                return false;
            }
        });
    };
    LimitOrderProtocolFacade.prototype.remaining = function (orderHash) {
        var _this = this;
        var callData = this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.remaining, [orderHash]);
        return this.providerConnector
            .ethCall(this.contractAddress, callData)
            .then(function (result) {
            var response = _this.parseRemainingResponse(result);
            if (response !== null) {
                return response;
            }
            // Parse error
            var parsed = _this.parseContractResponse(result);
            return Promise.reject(parsed);
        });
    };
    LimitOrderProtocolFacade.prototype.simulateCalls = function (tokens, data) {
        var _this = this;
        var callData = this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.simulateCalls, [tokens, data]);
        return this.providerConnector
            .ethCall(this.contractAddress, callData)
            .then(function (result) {
            var parsed = _this.parseSimulateTransferResponse(result);
            if (parsed !== null)
                return parsed;
            return Promise.reject(result);
        })
            .catch(function (error) {
            var parsed = _this.parseSimulateTransferError(error);
            if (parsed !== null)
                return parsed;
            return Promise.reject(error);
        });
    };
    LimitOrderProtocolFacade.prototype.domainSeparator = function () {
        var callData = this.getContractCallData(limit_order_protocol_model_1.LimitOrderProtocolMethods.DOMAIN_SEPARATOR);
        return this.providerConnector.ethCall(this.contractAddress, callData);
    };
    LimitOrderProtocolFacade.prototype.getContractCallData = function (methodName, methodParams) {
        if (methodParams === void 0) { methodParams = []; }
        return this.providerConnector.contractEncodeABI(limit_order_protocol_const_1.LIMIT_ORDER_PROTOCOL_ABI, this.contractAddress, methodName, methodParams);
    };
    LimitOrderProtocolFacade.prototype.parseRemainingResponse = function (response) {
        if (response.length === 66) {
            return bignumber_1.BigNumber.from(response);
        }
        return null;
    };
    LimitOrderProtocolFacade.prototype.parseSimulateTransferResponse = function (response) {
        var parsed = this.parseContractResponse(response);
        if (parsed.startsWith(limit_order_protocol_const_1.CALL_RESULTS_PREFIX)) {
            var data = parsed.replace(limit_order_protocol_const_1.CALL_RESULTS_PREFIX, '');
            return !data.includes('0');
        }
        return null;
    };
    LimitOrderProtocolFacade.prototype.parseSimulateTransferError = function (error) {
        var message = this.stringifyError(error);
        var isCorrectCode = this.isMsgContainsCorrectCode(message);
        if (isCorrectCode !== null) {
            return isCorrectCode;
        }
        try {
            var code = get_rpc_code_1.getRPCCode(message);
            return code ? this.isMsgContainsCorrectCode(code) : null;
        }
        catch (e) {
            return null;
        }
    };
    LimitOrderProtocolFacade.prototype.parseContractResponse = function (response) {
        // Aurora network wraps revert message into Revert()
        var matched = response.match(/Revert\(([^)]+)\)/);
        var message = matched && matched[1] ? matched[1] : response;
        return this.providerConnector.decodeABIParameter('string', limit_order_protocol_const_1.ZX + message.slice(10));
    };
    LimitOrderProtocolFacade.prototype.isMsgContainsCorrectCode = function (message) {
        var regex = new RegExp('(' + limit_order_protocol_const_1.CALL_RESULTS_PREFIX + '\\d+)');
        var matched = message.match(regex);
        if (matched) {
            return !matched[0].includes('0');
        }
        else {
            try {
                var matchParsed = this.parseContractResponse(message).match(regex);
                if (matchParsed) {
                    return !matchParsed[0].includes('0');
                }
            }
            catch (e) {
                console.error(e);
            }
        }
        return null;
    };
    LimitOrderProtocolFacade.prototype.stringifyError = function (error) {
        if (typeof error === 'string') {
            return error;
        }
        if (error instanceof Error) {
            return error.toString();
        }
        return JSON.stringify(error);
    };
    return LimitOrderProtocolFacade;
}());
exports.LimitOrderProtocolFacade = LimitOrderProtocolFacade;
