// @cowprotocol/sdk-app-data/dist/index.mjs
import {
  __glob
} from "./chunk-3563FJHN.mjs";

// src/api/index.ts
import { setGlobalAdapter } from "@cowprotocol/sdk-common";

// src/api/appDataHexToCid.ts
import { getGlobalAdapter } from "@cowprotocol/sdk-common";

// src/consts.ts
var DEFAULT_IPFS_READ_URI = "https://cloudflare-ipfs.com/ipfs";
var DEFAULT_IPFS_WRITE_URI = "https://api.pinata.cloud";
var MetaDataError = class extends Error {
};

// src/api/appDataHexToCid.ts
import { CID } from "multiformats/cid";
async function appDataHexToCid(appDataHex) {
  const cid = await _appDataHexToCid(appDataHex);
  _assertCid(cid, appDataHex);
  return cid;
}
async function appDataHexToCidLegacy(appDataHex) {
  const cid = await _appDataHexToCidLegacy(appDataHex);
  _assertCid(cid, appDataHex);
  return cid;
}
async function _assertCid(cid, appDataHex) {
  if (!cid)
    throw new MetaDataError("Error getting CID from appDataHex: " + appDataHex);
}
async function _appDataHexToCid(appDataHex) {
  const cidBytes = await _toCidBytes({
    version: 1,
    // CIDv1
    multicodec: 85,
    // Raw codec
    hashingAlgorithm: 27,
    // keccak hash algorithm
    hashingLength: 32,
    // keccak hash length (0x20 = 32)
    multihashHex: appDataHex
    // 32 bytes of the keccak256 hash
  });
  const { base16 } = await import("multiformats/bases/base16");
  return base16.encode(cidBytes);
}
async function _appDataHexToCidLegacy(appDataHex) {
  const cidBytes = await _toCidBytes({
    version: 1,
    // CIDv1
    multicodec: 112,
    // dag-pb
    hashingAlgorithm: 18,
    // sha2-256 hash algorithm
    hashingLength: 32,
    //  SHA-256 length (0x20 = 32)
    multihashHex: appDataHex
    // 32 bytes of the sha2-256 hash
  });
  return CID.decode(cidBytes).toV0().toString();
}
async function _toCidBytes(params) {
  const hashBytes = getGlobalAdapter().utils.arrayify(params.multihashHex);
  const cidPrefix = Uint8Array.from([params.version, params.multicodec, params.hashingAlgorithm, params.hashingLength]);
  const cidBytes = new Uint8Array(cidPrefix.length + hashBytes.length);
  cidBytes.set(cidPrefix);
  cidBytes.set(hashBytes, cidPrefix.length);
  return cidBytes;
}

// src/utils/ipfs.ts
var MULTIBASE_BASE16 = "f";
async function parseCid(ipfsHash) {
  const { CID: CID2 } = await import("multiformats/cid");
  const decoder = await getDecoder(ipfsHash);
  return CID2.parse(ipfsHash, decoder);
}
async function getDecoder(ipfsHash) {
  if (ipfsHash[0] === MULTIBASE_BASE16) {
    const { base16 } = await import("multiformats/bases/base16");
    return base16;
  }
  return void 0;
}
async function extractDigest(cid) {
  const cidDetails = await parseCid(cid);
  const { digest } = cidDetails.multihash;
  return `0x${Buffer.from(digest).toString("hex")}`;
}

// src/api/cidToAppDataHex.ts
async function cidToAppDataHex(cid) {
  return extractDigest(cid);
}

// src/api/fetchDocFromCid.ts
async function fetchDocFromCid(cid, ipfsUri = DEFAULT_IPFS_READ_URI) {
  const { default: fetch } = await import("cross-fetch");
  const response = await fetch(`${ipfsUri}/${cid}`);
  return await response.json();
}

// src/api/fetchDocFromAppData.ts
async function fetchDocFromAppDataHex(appDataHex, ipfsUri) {
  return _fetchDocFromCidAux(appDataHexToCid, appDataHex, ipfsUri);
}
async function fetchDocFromAppDataHexLegacy(appDataHex, ipfsUri) {
  return _fetchDocFromCidAux(appDataHexToCidLegacy, appDataHex, ipfsUri);
}
async function _fetchDocFromCidAux(hexToCid, appDataHex, ipfsUri) {
  try {
    const cid = await hexToCid(appDataHex);
    if (!cid)
      throw new MetaDataError("Error getting serialized CID");
    return fetchDocFromCid(cid, ipfsUri);
  } catch (e) {
    const error = e;
    throw new MetaDataError(`Error decoding AppData: appDataHex=${appDataHex}, message=${error.message}`);
  }
}

// src/generatedTypes/v0.1.0.ts
var v0_1_0_exports = {};

// src/generatedTypes/v0.10.0.ts
var v0_10_0_exports = {};

// src/generatedTypes/v0.11.0.ts
var v0_11_0_exports = {};

// src/generatedTypes/v0.2.0.ts
var v0_2_0_exports = {};

// src/generatedTypes/v0.3.0.ts
var v0_3_0_exports = {};

// src/generatedTypes/v0.4.0.ts
var v0_4_0_exports = {};

// src/generatedTypes/v0.5.0.ts
var v0_5_0_exports = {};

// src/generatedTypes/v0.6.0.ts
var v0_6_0_exports = {};

// src/generatedTypes/v0.7.0.ts
var v0_7_0_exports = {};

// src/generatedTypes/v0.8.0.ts
var v0_8_0_exports = {};

// src/generatedTypes/v0.9.0.ts
var v0_9_0_exports = {};

// src/generatedTypes/v1.0.0.ts
var v1_0_0_exports = {};

// src/generatedTypes/v1.1.0.ts
var v1_1_0_exports = {};

// src/generatedTypes/v1.10.0.ts
var v1_10_0_exports = {};

// src/generatedTypes/v1.11.0.ts
var v1_11_0_exports = {};

// src/generatedTypes/v1.12.0.ts
var v1_12_0_exports = {};

// src/generatedTypes/v1.13.0.ts
var v1_13_0_exports = {};

// src/generatedTypes/v1.14.0.ts
var v1_14_0_exports = {};

// src/generatedTypes/v1.2.0.ts
var v1_2_0_exports = {};

// src/generatedTypes/v1.3.0.ts
var v1_3_0_exports = {};

// src/generatedTypes/v1.4.0.ts
var v1_4_0_exports = {};

// src/generatedTypes/v1.5.0.ts
var v1_5_0_exports = {};

// src/generatedTypes/v1.6.0.ts
var v1_6_0_exports = {};

// src/generatedTypes/v1.7.0.ts
var v1_7_0_exports = {};

// src/generatedTypes/v1.8.0.ts
var v1_8_0_exports = {};

// src/generatedTypes/v1.9.0.ts
var v1_9_0_exports = {};

// src/generatedTypes/index.ts
var LATEST_APP_DATA_VERSION = "1.14.0";
var LATEST_QUOTE_METADATA_VERSION = "1.1.0";
var LATEST_REFERRER_METADATA_VERSION = "1.0.0";
var LATEST_ORDER_CLASS_METADATA_VERSION = "0.3.0";
var LATEST_UTM_METADATA_VERSION = "0.3.0";
var LATEST_HOOKS_METADATA_VERSION = "0.2.0";
var LATEST_SIGNER_METADATA_VERSION = "0.1.0";
var LATEST_WIDGET_METADATA_VERSION = "0.1.0";
var LATEST_PARTNER_FEE_METADATA_VERSION = "1.0.0";
var LATEST_REPLACED_ORDER_METADATA_VERSION = "0.1.0";
var LATEST_WRAPPERS_METADATA_VERSION = "0.2.0";
var LATEST_USER_CONSENTS_METADATA_VERSION = "0.1.0";

// src/api/generateAppDataDoc.ts
var DEFAULT_APP_CODE = "CoW Swap";
var DEFAULT_APP_DATA_DOC = {
  appCode: DEFAULT_APP_CODE,
  metadata: {},
  version: LATEST_APP_DATA_VERSION
};
async function generateAppDataDoc(params) {
  return {
    ...DEFAULT_APP_DATA_DOC,
    ...params,
    version: LATEST_APP_DATA_VERSION
  };
}

// src/api/getAppDataInfo.ts
import { getGlobalAdapter as getGlobalAdapter2 } from "@cowprotocol/sdk-common";

// src/utils/stringify.ts
import stringify from "json-stringify-deterministic";
async function stringifyDeterministic(obj) {
  return stringify(obj);
}

// src/importSchema.ts
import { CowError } from "@cowprotocol/sdk-common";

// import("../schemas/v*.json") in src/importSchema.ts
var globImport_schemas_v_json = __glob({
  "../schemas/v0.1.0.json": () => import("./v0.1.0-452JMA5G.mjs"),
  "../schemas/v0.10.0.json": () => import("./v0.10.0-OHB5VFCY.mjs"),
  "../schemas/v0.11.0.json": () => import("./v0.11.0-YC4BNGSG.mjs"),
  "../schemas/v0.2.0.json": () => import("./v0.2.0-RZSAEYWJ.mjs"),
  "../schemas/v0.3.0.json": () => import("./v0.3.0-QEEYHNXA.mjs"),
  "../schemas/v0.4.0.json": () => import("./v0.4.0-XWYIU6OQ.mjs"),
  "../schemas/v0.5.0.json": () => import("./v0.5.0-6JWER4SU.mjs"),
  "../schemas/v0.6.0.json": () => import("./v0.6.0-DCUHOIX6.mjs"),
  "../schemas/v0.7.0.json": () => import("./v0.7.0-WUIYE67P.mjs"),
  "../schemas/v0.8.0.json": () => import("./v0.8.0-MA4RCSXJ.mjs"),
  "../schemas/v0.9.0.json": () => import("./v0.9.0-MODYXNNJ.mjs"),
  "../schemas/v1.0.0.json": () => import("./v1.0.0-LO3INJ7G.mjs"),
  "../schemas/v1.1.0.json": () => import("./v1.1.0-VDZLWY4U.mjs"),
  "../schemas/v1.10.0.json": () => import("./v1.10.0-HV5QAMRI.mjs"),
  "../schemas/v1.11.0.json": () => import("./v1.11.0-7OORRXZT.mjs"),
  "../schemas/v1.12.0.json": () => import("./v1.12.0-JRP6T6PK.mjs"),
  "../schemas/v1.13.0.json": () => import("./v1.13.0-F46Z6ASK.mjs"),
  "../schemas/v1.14.0.json": () => import("./v1.14.0-4GFOYRNY.mjs"),
  "../schemas/v1.2.0.json": () => import("./v1.2.0-J7HBJ6X3.mjs"),
  "../schemas/v1.3.0.json": () => import("./v1.3.0-YE5Z6DA7.mjs"),
  "../schemas/v1.4.0.json": () => import("./v1.4.0-XW4S4BWH.mjs"),
  "../schemas/v1.5.0.json": () => import("./v1.5.0-RPBY3HCV.mjs"),
  "../schemas/v1.6.0.json": () => import("./v1.6.0-WTOLS5Y6.mjs"),
  "../schemas/v1.7.0.json": () => import("./v1.7.0-7V2ZPHKV.mjs"),
  "../schemas/v1.8.0.json": () => import("./v1.8.0-UKWSFP36.mjs"),
  "../schemas/v1.9.0.json": () => import("./v1.9.0-SQYL3FIK.mjs")
});

// src/importSchema.ts
async function importSchema(version) {
  if (!/^\d+\.\d+\.\d+$/.test(version)) {
    throw new Error(`AppData version ${version} is not a valid version`);
  }
  try {
    return await globImport_schemas_v_json(`../schemas/v${version}.json`);
  } catch (e) {
    console.log(e);
    throw new CowError(`AppData version ${version} doesn't exist`);
  }
}

// src/api/validateAppDataDoc.ts
var _ajvPromise;
var _validatorPromises = {};
async function getAjv() {
  if (!_ajvPromise) {
    _ajvPromise = import("ajv").then(({ default: Ajv }) => new Ajv());
  }
  return _ajvPromise;
}
async function _createValidator(ajv, version) {
  let validator = ajv.getSchema(version);
  if (!validator) {
    const schema = await importSchema(version);
    ajv.addSchema(schema, version);
    validator = ajv.getSchema(version);
  }
  return validator;
}
async function getValidator(ajv, version) {
  let validatorPromise = _validatorPromises[version];
  if (!validatorPromise) {
    validatorPromise = _createValidator(ajv, version);
    _validatorPromises[version] = validatorPromise;
  }
  return validatorPromise;
}
async function validateAppDataDoc(appDataDoc) {
  const { version } = appDataDoc;
  try {
    const ajv = await getAjv();
    const validator = await getValidator(ajv, version);
    const success = !!validator(appDataDoc);
    const errors = validator.errors ? ajv.errorsText(validator.errors) : void 0;
    return { success, errors };
  } catch (e) {
    if (e instanceof Error) {
      return {
        success: false,
        errors: e.message
      };
    } else {
      throw e;
    }
  }
}

// src/api/getAppDataInfo.ts
async function getAppDataInfo(appDataAux) {
  return _appDataToCidAux(appDataAux, _appDataToCid);
}
async function getAppDataInfoLegacy(appDataAux) {
  const fullAppData = JSON.stringify(appDataAux);
  return _appDataToCidAux(fullAppData, _appDataToCidLegacy);
}
async function _appDataToCidAux(appDataAux, deriveCid) {
  const [appDataDoc, fullAppData] = typeof appDataAux === "string" ? [JSON.parse(appDataAux), appDataAux] : [appDataAux, await stringifyDeterministic(appDataAux)];
  const validation = await validateAppDataDoc(appDataDoc);
  if (!validation?.success) {
    throw new MetaDataError(`Invalid appData provided: ${validation?.errors}`);
  }
  try {
    const cid = await deriveCid(fullAppData);
    const appDataHex = await extractDigest(cid);
    if (!appDataHex) {
      throw new MetaDataError(`Could not extract appDataHex from calculated cid ${cid}`);
    }
    return { cid, appDataHex, appDataContent: fullAppData };
  } catch (e) {
    const error = e;
    console.error("Failed to calculate appDataHex", error);
    throw new MetaDataError(`Failed to calculate appDataHex: ${error.message}`);
  }
}
async function _appDataToCid(fullAppDataJson) {
  const adapter = getGlobalAdapter2();
  const appDataHex = adapter.utils.keccak256(adapter.utils.toUtf8Bytes(fullAppDataJson));
  return appDataHexToCid(appDataHex);
}
async function _appDataToCidLegacy(doc) {
  const fullAppData = typeof doc === "string" ? doc : await stringifyDeterministic(doc);
  const { of } = await import("ipfs-only-hash");
  return of(fullAppData, { cidVersion: 0 });
}

// src/api/getAppDataSchema.ts
async function getAppDataSchema(version) {
  try {
    return await importSchema(version);
  } catch (e) {
    const error = e;
    throw new MetaDataError(error.message);
  }
}

// src/api/uploadMetadataDocToIpfsLegacy.ts
async function uploadMetadataDocToIpfsLegacy(appDataDoc, ipfsConfig) {
  const { IpfsHash: cid } = await _pinJsonInPinataIpfs(appDataDoc, ipfsConfig);
  return {
    appData: await extractDigest(cid),
    cid
  };
}
async function _pinJsonInPinataIpfs(file, { writeUri = DEFAULT_IPFS_WRITE_URI, pinataApiKey = "", pinataApiSecret = "" }) {
  const { default: fetch } = await import("cross-fetch");
  if (!pinataApiKey || !pinataApiSecret) {
    throw new MetaDataError("You need to pass IPFS api credentials.");
  }
  const body = await stringifyDeterministic({
    pinataContent: file,
    pinataMetadata: { name: "appData" }
  });
  const pinataUrl = `${writeUri}/pinning/pinJSONToIPFS`;
  const response = await fetch(pinataUrl, {
    method: "POST",
    body,
    headers: {
      "Content-Type": "application/json",
      pinata_api_key: pinataApiKey,
      pinata_secret_api_key: pinataApiSecret
    }
  });
  const data = await response.json();
  if (response.status !== 200) {
    throw new Error(data.error.details || data.error);
  }
  return data;
}

// src/api/index.ts
var MetadataApi = class {
  /**
   * Creates a new MetadataApi instance
   *
   * @param adapter Provider adapter implementation
   */
  constructor(adapter) {
    if (adapter) {
      setGlobalAdapter(adapter);
    }
  }
  // Schema & Doc generation/validation
  getAppDataSchema = getAppDataSchema;
  generateAppDataDoc = generateAppDataDoc;
  validateAppDataDoc = validateAppDataDoc;
  // appData / CID conversion
  getAppDataInfo = getAppDataInfo;
  // (appData | fullAppData) -->  { cid, appDataHex, appDataContent }
  appDataHexToCid = appDataHexToCid;
  // appDataHex --> cid
  cidToAppDataHex = cidToAppDataHex;
  // cid --> appDataHex
  // Fetch from IPFS
  fetchDocFromAppDataHex = fetchDocFromAppDataHex;
  // appDataHex --> appData
  // Legacy methods
  legacy = {
    // Fetch appData document from IPFS (deprecated)
    fetchDocFromCid,
    // cid --> document
    // Upload to IPFS (deprecated)
    uploadMetadataDocToIpfs: uploadMetadataDocToIpfsLegacy,
    //  appData --> cid + publish IPFS
    appDataToCid: getAppDataInfoLegacy,
    // (appData | fullAppData) --> cid
    appDataHexToCid: appDataHexToCidLegacy,
    // appDataHex --> cid
    fetchDocFromAppDataHex: fetchDocFromAppDataHexLegacy
    // appDataHex --> appData
  };
};
export {
  LATEST_APP_DATA_VERSION,
  LATEST_HOOKS_METADATA_VERSION,
  LATEST_ORDER_CLASS_METADATA_VERSION,
  LATEST_PARTNER_FEE_METADATA_VERSION,
  LATEST_QUOTE_METADATA_VERSION,
  LATEST_REFERRER_METADATA_VERSION,
  LATEST_REPLACED_ORDER_METADATA_VERSION,
  LATEST_SIGNER_METADATA_VERSION,
  LATEST_USER_CONSENTS_METADATA_VERSION,
  LATEST_UTM_METADATA_VERSION,
  LATEST_WIDGET_METADATA_VERSION,
  LATEST_WRAPPERS_METADATA_VERSION,
  MetadataApi,
  v1_14_0_exports as cowAppDataLatestScheme,
  stringifyDeterministic,
  v0_10_0_exports as v0_10_0,
  v0_11_0_exports as v0_11_0,
  v0_1_0_exports as v0_1_0,
  v0_2_0_exports as v0_2_0,
  v0_3_0_exports as v0_3_0,
  v0_4_0_exports as v0_4_0,
  v0_5_0_exports as v0_5_0,
  v0_6_0_exports as v0_6_0,
  v0_7_0_exports as v0_7_0,
  v0_8_0_exports as v0_8_0,
  v0_9_0_exports as v0_9_0,
  v1_0_0_exports as v1_0_0,
  v1_10_0_exports as v1_10_0,
  v1_11_0_exports as v1_11_0,
  v1_12_0_exports as v1_12_0,
  v1_13_0_exports as v1_13_0,
  v1_14_0_exports as v1_14_0,
  v1_1_0_exports as v1_1_0,
  v1_2_0_exports as v1_2_0,
  v1_3_0_exports as v1_3_0,
  v1_4_0_exports as v1_4_0,
  v1_5_0_exports as v1_5_0,
  v1_6_0_exports as v1_6_0,
  v1_7_0_exports as v1_7_0,
  v1_8_0_exports as v1_8_0,
  v1_9_0_exports as v1_9_0
};
