// @cowprotocol/sdk-config/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ADDITIONAL_TARGET_CHAINS_MAP: () => ADDITIONAL_TARGET_CHAINS_MAP,
  ALL_ADDITIONAL_TARGET_CHAINS: () => ALL_ADDITIONAL_TARGET_CHAINS,
  ALL_ADDITIONAL_TARGET_CHAIN_IDS: () => ALL_ADDITIONAL_TARGET_CHAIN_IDS,
  ALL_CHAINS: () => ALL_CHAINS,
  ALL_CHAINS_IDS: () => ALL_CHAINS_IDS,
  ALL_SUPPORTED_CHAINS: () => ALL_SUPPORTED_CHAINS,
  ALL_SUPPORTED_CHAINS_MAP: () => ALL_SUPPORTED_CHAINS_MAP,
  ALL_SUPPORTED_CHAIN_IDS: () => ALL_SUPPORTED_CHAIN_IDS,
  AdditionalTargetChainId: () => AdditionalTargetChainId,
  BARN_ETH_FLOW_ADDRESSES: () => BARN_ETH_FLOW_ADDRESSES,
  BTC_CURRENCY_ADDRESS: () => BTC_CURRENCY_ADDRESS,
  COMPOSABLE_COW: () => COMPOSABLE_COW,
  COMPOSABLE_COW_CONTRACT_ADDRESS: () => COMPOSABLE_COW_CONTRACT_ADDRESS,
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS: () => COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS,
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING: () => COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING,
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS: () => COW_PROTOCOL_VAULT_RELAYER_ADDRESS,
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING: () => COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING,
  DEFAULT_COW_API_CONTEXT: () => DEFAULT_COW_API_CONTEXT,
  DEFAULT_IPFS_READ_URI: () => DEFAULT_IPFS_READ_URI,
  DEFAULT_IPFS_WRITE_URI: () => DEFAULT_IPFS_WRITE_URI,
  ENVS_LIST: () => ENVS_LIST,
  ETH_ADDRESS: () => ETH_ADDRESS,
  ETH_FLOW_ADDRESSES: () => ETH_FLOW_ADDRESSES,
  EVM_NATIVE_CURRENCY_ADDRESS: () => EVM_NATIVE_CURRENCY_ADDRESS,
  EXTENSIBLE_FALLBACK_HANDLER: () => EXTENSIBLE_FALLBACK_HANDLER,
  EXTENSIBLE_FALLBACK_HANDLER_CONTRACT_ADDRESS: () => EXTENSIBLE_FALLBACK_HANDLER_CONTRACT_ADDRESS,
  EvmChains: () => EvmChains,
  MAX_VALID_TO_EPOCH: () => MAX_VALID_TO_EPOCH,
  NonEvmChains: () => NonEvmChains,
  RAW_CHAINS_FILES_PATH: () => RAW_CHAINS_FILES_PATH,
  RAW_FILES_PATH: () => RAW_FILES_PATH,
  SOL_NATIVE_CURRENCY_ADDRESS: () => SOL_NATIVE_CURRENCY_ADDRESS,
  SupportedChainId: () => SupportedChainId,
  TOKEN_LIST_IMAGES_PATH: () => TOKEN_LIST_IMAGES_PATH,
  TRADABLE_SUPPORTED_CHAINS: () => TRADABLE_SUPPORTED_CHAINS,
  TRADABLE_SUPPORTED_CHAIN_IDS: () => TRADABLE_SUPPORTED_CHAIN_IDS,
  WRAPPED_NATIVE_CURRENCIES: () => WRAPPED_NATIVE_CURRENCIES,
  arbitrumOne: () => arbitrumOne,
  avalanche: () => avalanche,
  base: () => base,
  bitcoin: () => bitcoin,
  bnb: () => bnb,
  getChainInfo: () => getChainInfo,
  getWrappedTokenForChain: () => getWrappedTokenForChain,
  gnosisChain: () => gnosisChain,
  ink: () => ink,
  isAdditionalTargetChain: () => isAdditionalTargetChain,
  isBtcChain: () => isBtcChain,
  isChainDeprecated: () => isChainDeprecated,
  isChainUnderDevelopment: () => isChainUnderDevelopment,
  isEvmChain: () => isEvmChain,
  isEvmChainInfo: () => isEvmChainInfo,
  isNonEvmChain: () => isNonEvmChain,
  isNonEvmChainInfo: () => isNonEvmChainInfo,
  isSupportedChain: () => isSupportedChain,
  isTargetChainId: () => isTargetChainId,
  isZkSyncChain: () => isZkSyncChain,
  linea: () => linea,
  mainnet: () => mainnet,
  mapAddressToSupportedNetworks: () => mapAddressToSupportedNetworks,
  mapAllNetworks: () => mapAllNetworks,
  mapSupportedNetworks: () => mapSupportedNetworks,
  nativeCurrencyTemplate: () => nativeCurrencyTemplate,
  optimism: () => optimism,
  plasma: () => plasma,
  polygon: () => polygon,
  sepolia: () => sepolia,
  solana: () => solana
});
module.exports = __toCommonJS(src_exports);

// src/chains/types.ts
var MAINNET_ID = 1;
var BNB_ID = 56;
var GNOSIS_CHAIN_ID = 100;
var POLYGON_ID = 137;
var BASE_ID = 8453;
var PLASMA_ID = 9745;
var ARBITRUM_ONE_ID = 42161;
var AVALANCHE_ID = 43114;
var INK_ID = 57073;
var LINEA_ID = 59144;
var SEPOLIA_ID = 11155111;
var OPTIMISM_ID = 10;
var BITCOIN_ID = 1e9;
var SOLANA_ID = 1000000001;
var EvmChains = /* @__PURE__ */ ((EvmChains2) => {
  EvmChains2[EvmChains2["MAINNET"] = MAINNET_ID] = "MAINNET";
  EvmChains2[EvmChains2["BNB"] = BNB_ID] = "BNB";
  EvmChains2[EvmChains2["GNOSIS_CHAIN"] = GNOSIS_CHAIN_ID] = "GNOSIS_CHAIN";
  EvmChains2[EvmChains2["POLYGON"] = POLYGON_ID] = "POLYGON";
  EvmChains2[EvmChains2["BASE"] = BASE_ID] = "BASE";
  EvmChains2[EvmChains2["PLASMA"] = PLASMA_ID] = "PLASMA";
  EvmChains2[EvmChains2["ARBITRUM_ONE"] = ARBITRUM_ONE_ID] = "ARBITRUM_ONE";
  EvmChains2[EvmChains2["AVALANCHE"] = AVALANCHE_ID] = "AVALANCHE";
  EvmChains2[EvmChains2["INK"] = INK_ID] = "INK";
  EvmChains2[EvmChains2["LINEA"] = LINEA_ID] = "LINEA";
  EvmChains2[EvmChains2["SEPOLIA"] = SEPOLIA_ID] = "SEPOLIA";
  EvmChains2[EvmChains2["OPTIMISM"] = OPTIMISM_ID] = "OPTIMISM";
  return EvmChains2;
})(EvmChains || {});
var NonEvmChains = /* @__PURE__ */ ((NonEvmChains2) => {
  NonEvmChains2[NonEvmChains2["BITCOIN"] = BITCOIN_ID] = "BITCOIN";
  NonEvmChains2[NonEvmChains2["SOLANA"] = SOLANA_ID] = "SOLANA";
  return NonEvmChains2;
})(NonEvmChains || {});
var SupportedChainId = /* @__PURE__ */ ((SupportedChainId2) => {
  SupportedChainId2[SupportedChainId2["MAINNET"] = MAINNET_ID] = "MAINNET";
  SupportedChainId2[SupportedChainId2["BNB"] = BNB_ID] = "BNB";
  SupportedChainId2[SupportedChainId2["GNOSIS_CHAIN"] = GNOSIS_CHAIN_ID] = "GNOSIS_CHAIN";
  SupportedChainId2[SupportedChainId2["POLYGON"] = POLYGON_ID] = "POLYGON";
  SupportedChainId2[SupportedChainId2["BASE"] = BASE_ID] = "BASE";
  SupportedChainId2[SupportedChainId2["PLASMA"] = PLASMA_ID] = "PLASMA";
  SupportedChainId2[SupportedChainId2["ARBITRUM_ONE"] = ARBITRUM_ONE_ID] = "ARBITRUM_ONE";
  SupportedChainId2[SupportedChainId2["AVALANCHE"] = AVALANCHE_ID] = "AVALANCHE";
  SupportedChainId2[SupportedChainId2["INK"] = INK_ID] = "INK";
  SupportedChainId2[SupportedChainId2["LINEA"] = LINEA_ID] = "LINEA";
  SupportedChainId2[SupportedChainId2["SEPOLIA"] = SEPOLIA_ID] = "SEPOLIA";
  return SupportedChainId2;
})(SupportedChainId || {});
var AdditionalTargetChainId = /* @__PURE__ */ ((AdditionalTargetChainId2) => {
  AdditionalTargetChainId2[AdditionalTargetChainId2["OPTIMISM"] = OPTIMISM_ID] = "OPTIMISM";
  AdditionalTargetChainId2[AdditionalTargetChainId2["BITCOIN"] = BITCOIN_ID] = "BITCOIN";
  AdditionalTargetChainId2[AdditionalTargetChainId2["SOLANA"] = SOLANA_ID] = "SOLANA";
  return AdditionalTargetChainId2;
})(AdditionalTargetChainId || {});

// src/constants/addresses.ts
var EVM_NATIVE_CURRENCY_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
var SOL_NATIVE_CURRENCY_ADDRESS = "11111111111111111111111111111111";
var BTC_CURRENCY_ADDRESS = "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa";

// src/constants/paths.ts
var RAW_FILES_PATH = "https://files.cow.fi/cow-sdk";
var RAW_CHAINS_FILES_PATH = `${RAW_FILES_PATH}/chains`;
var TOKEN_LIST_IMAGES_PATH = "https://files.cow.fi/token-lists/images";

// src/constants/nativeCurrencyTemplate.ts
var nativeCurrencyTemplate = {
  address: EVM_NATIVE_CURRENCY_ADDRESS,
  decimals: 18,
  name: "Ether",
  symbol: "ETH",
  logoUrl: `${TOKEN_LIST_IMAGES_PATH}/1/${EVM_NATIVE_CURRENCY_ADDRESS.toLowerCase()}/logo.png`
};

// src/chains/details/mainnet.ts
var ethereumLogo = `${RAW_CHAINS_FILES_PATH}/images/mainnet-logo.svg`;
var mainnet = {
  id: SupportedChainId.MAINNET,
  label: "Ethereum",
  eip155Label: "Ethereum Mainnet",
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.MAINNET
  },
  addressPrefix: "eth",
  isTestnet: false,
  contracts: {
    ensRegistry: {
      address: "0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e"
    },
    ensUniversalResolver: {
      address: "0xce01f8eee7E479C928F8919abD53E553a36CeF67",
      blockCreated: 19258213
    },
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 14353601
    }
  },
  rpcUrls: {
    default: {
      http: ["https://eth.merkle.io"]
    }
  },
  color: "#62688F",
  logo: { light: ethereumLogo, dark: ethereumLogo },
  website: {
    name: "Ethereum",
    url: "https://ethereum.org"
  },
  docs: {
    name: "Ethereum Docs",
    url: "https://ethereum.org/en/developers/docs"
  },
  blockExplorer: {
    name: "Etherscan",
    url: "https://etherscan.io"
  }
};

// src/chains/details/gnosis.ts
var gnosisChainLogo = `${RAW_CHAINS_FILES_PATH}/images/gnosis-logo.svg`;
var gnosisChain = {
  id: SupportedChainId.GNOSIS_CHAIN,
  label: "Gnosis",
  eip155Label: "Gnosis",
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.GNOSIS_CHAIN,
    name: "xDAI",
    symbol: "xDAI",
    logoUrl: `${TOKEN_LIST_IMAGES_PATH}/100/${EVM_NATIVE_CURRENCY_ADDRESS.toLowerCase()}/logo.png`
  },
  addressPrefix: "gno",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 21022491
    }
  },
  rpcUrls: {
    default: {
      http: ["https://rpc.gnosischain.com"],
      webSocket: ["wss://rpc.gnosischain.com/wss"]
    }
  },
  color: "#07795B",
  logo: { light: gnosisChainLogo, dark: gnosisChainLogo },
  website: {
    name: "Gnosis Chain",
    url: "https://www.gnosischain.com"
  },
  docs: {
    name: "Gnosis Chain Docs",
    url: "https://docs.gnosischain.com"
  },
  blockExplorer: {
    name: "Gnosisscan",
    url: "https://gnosisscan.io"
  },
  bridges: [
    {
      name: "Gnosis Chain Bridge",
      url: "https://bridge.gnosischain.com"
    }
  ]
};

// src/chains/details/arbitrum.ts
var arbitrumLogo = `${RAW_CHAINS_FILES_PATH}/images/arbitrum-logo.svg`;
var arbitrumOne = {
  id: SupportedChainId.ARBITRUM_ONE,
  label: "Arbitrum",
  eip155Label: "Arbitrum One",
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.ARBITRUM_ONE
  },
  addressPrefix: "arb1",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 7654707
    }
  },
  rpcUrls: {
    default: {
      http: ["https://arb1.arbitrum.io/rpc"]
    }
  },
  color: "#1B4ADD",
  logo: { light: arbitrumLogo, dark: arbitrumLogo },
  website: {
    name: "Arbitrum",
    url: "https://arbitrum.io"
  },
  docs: {
    name: "Arbitrum Docs",
    url: "https://docs.arbitrum.io"
  },
  blockExplorer: {
    name: "Arbiscan",
    url: "https://arbiscan.io"
  },
  bridges: [
    {
      name: "Arbitrum Bridge",
      url: "https://bridge.arbitrum.io"
    }
  ]
};

// src/chains/details/base.ts
var baseLogo = `${RAW_CHAINS_FILES_PATH}/images/base-logo.svg`;
var base = {
  id: SupportedChainId.BASE,
  label: "Base",
  eip155Label: "Base",
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.BASE
  },
  addressPrefix: "base",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 5022
    }
  },
  rpcUrls: {
    default: {
      http: ["https://mainnet.base.org"]
    }
  },
  color: "#0052FF",
  logo: { light: baseLogo, dark: baseLogo },
  website: {
    name: "Base",
    url: "https://base.org"
  },
  docs: {
    name: "Base Docs",
    url: "https://docs.base.org"
  },
  blockExplorer: {
    name: "Basescan",
    url: "https://basescan.org"
  },
  bridges: [
    {
      name: "Superchain Bridges",
      url: "https://bridge.base.org/deposit"
    }
  ]
};

// src/constants/ipfs.ts
var DEFAULT_IPFS_READ_URI = "https://gnosis.mypinata.cloud/ipfs";
var DEFAULT_IPFS_WRITE_URI = "https://api.pinata.cloud";

// src/constants/wrappedTokens.ts
var wrappedNativeCurrencyEth = {
  decimals: 18,
  name: "Wrapped Ether",
  symbol: "WETH",
  logoUrl: `${TOKEN_LIST_IMAGES_PATH}/1/0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2/logo.png`
};
var WRAPPED_NATIVE_CURRENCIES = {
  [SupportedChainId.MAINNET]: createWrappedTokenForChain(
    SupportedChainId.MAINNET,
    "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    wrappedNativeCurrencyEth
  ),
  [SupportedChainId.GNOSIS_CHAIN]: createWrappedTokenForChain(
    SupportedChainId.GNOSIS_CHAIN,
    "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
    {
      decimals: 18,
      name: "Wrapped XDAI",
      symbol: "WXDAI"
    }
  ),
  [SupportedChainId.ARBITRUM_ONE]: createWrappedTokenForChain(
    SupportedChainId.ARBITRUM_ONE,
    "0x82aF49447D8a07e3bd95BD0d56f35241523fBab1",
    wrappedNativeCurrencyEth
  ),
  [SupportedChainId.BASE]: createWrappedTokenForChain(
    SupportedChainId.BASE,
    "0x4200000000000000000000000000000000000006",
    wrappedNativeCurrencyEth
  ),
  [SupportedChainId.SEPOLIA]: createWrappedTokenForChain(
    SupportedChainId.SEPOLIA,
    "0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14",
    wrappedNativeCurrencyEth
  ),
  [SupportedChainId.POLYGON]: createWrappedTokenForChain(
    SupportedChainId.POLYGON,
    "0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270",
    {
      decimals: 18,
      name: "Wrapped POL",
      symbol: "WPOL"
    }
  ),
  [SupportedChainId.AVALANCHE]: createWrappedTokenForChain(
    SupportedChainId.AVALANCHE,
    "0xb31f66aa3c1e785363f0875a1b74e27b85fd66c7",
    {
      decimals: 18,
      name: "Wrapped AVAX",
      symbol: "WAVAX"
    }
  ),
  [SupportedChainId.BNB]: createWrappedTokenForChain(SupportedChainId.BNB, "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c", {
    decimals: 18,
    name: "Wrapped BNB",
    symbol: "WBNB"
  }),
  [SupportedChainId.PLASMA]: createWrappedTokenForChain(
    SupportedChainId.PLASMA,
    "0x6100e367285b01f48d07953803a2d8dca5d19873",
    {
      decimals: 18,
      name: "Wrapped XPL",
      symbol: "WXPL"
    }
  ),
  [SupportedChainId.LINEA]: createWrappedTokenForChain(
    SupportedChainId.LINEA,
    "0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f",
    wrappedNativeCurrencyEth
  ),
  [SupportedChainId.INK]: createWrappedTokenForChain(
    SupportedChainId.INK,
    "0x4200000000000000000000000000000000000006",
    wrappedNativeCurrencyEth
  )
};
function createWrappedTokenForChain(chainId, address, info) {
  const { logoUrl, ...rest } = info;
  return {
    chainId,
    address,
    ...rest,
    logoUrl: logoUrl || `${TOKEN_LIST_IMAGES_PATH}/${chainId}/${address.toLowerCase()}/logo.png`
  };
}
function getWrappedTokenForChain(chainId) {
  return WRAPPED_NATIVE_CURRENCIES[chainId];
}

// src/chains/details/sepolia.ts
var sepoliaLogo = `${RAW_CHAINS_FILES_PATH}/images/sepolia-logo.svg`;
var sepolia = {
  id: SupportedChainId.SEPOLIA,
  label: "Sepolia",
  eip155Label: "Ethereum Sepolia",
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.SEPOLIA
  },
  addressPrefix: "sep",
  isTestnet: true,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 751532
    },
    ensRegistry: { address: "0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e" },
    ensUniversalResolver: {
      address: "0xc8Af999e38273D658BE1b921b88A9Ddf005769cC",
      blockCreated: 5317080
    }
  },
  rpcUrls: {
    default: {
      http: ["https://sepolia.drpc.org"]
    }
  },
  color: "#C12FF2",
  logo: { light: sepoliaLogo, dark: sepoliaLogo },
  website: {
    name: "Ethereum",
    url: "https://sepolia.dev"
  },
  docs: {
    name: "Sepolia Docs",
    url: "https://ethereum.org/en/developers/docs/networks/#sepolia"
  },
  blockExplorer: {
    name: "Etherscan",
    url: "https://sepolia.etherscan.io"
  }
};

// src/chains/details/avalanche.ts
var avaxLogo = `${RAW_CHAINS_FILES_PATH}/images/avax-logo.svg`;
var avalanche = {
  id: SupportedChainId.AVALANCHE,
  label: "Avalanche",
  eip155Label: "Avalanche C-Chain",
  logo: { light: avaxLogo, dark: avaxLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.AVALANCHE,
    name: "Avalanche",
    symbol: "AVAX",
    logoUrl: avaxLogo
  },
  addressPrefix: "avax",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 11907934
    }
  },
  rpcUrls: {
    default: {
      http: ["https://api.avax.network/ext/bc/C/rpc"]
    }
  },
  color: "#ff3944",
  website: {
    name: "Avalanche",
    url: "https://www.avax.network/"
  },
  docs: {
    name: "Avalanche Docs",
    url: "https://build.avax.network/docs"
  },
  blockExplorer: {
    name: "Snowscan",
    url: "https://snowscan.xyz"
  }
};

// src/chains/details/polygon.ts
var polygonLogo = `${RAW_CHAINS_FILES_PATH}/images/polygon-logo.svg`;
var polygon = {
  id: SupportedChainId.POLYGON,
  label: "Polygon",
  eip155Label: "Polygon Mainnet",
  logo: { light: polygonLogo, dark: polygonLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.POLYGON,
    name: "POL",
    symbol: "POL",
    logoUrl: polygonLogo
  },
  addressPrefix: "matic",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 25770160
    }
  },
  rpcUrls: {
    default: {
      http: ["https://polygon-rpc.com"]
    }
  },
  color: "#8247e5",
  website: {
    name: "Polygon",
    url: "https://polygon.technology"
  },
  docs: {
    name: "Polygon Docs",
    url: "https://docs.polygon.technology"
  },
  blockExplorer: {
    name: "Polygonscan",
    url: "https://polygonscan.com"
  }
};

// src/chains/details/bnb.ts
var bnbLogo = `${RAW_CHAINS_FILES_PATH}/images/bnb-logo.svg`;
var bnb = {
  id: SupportedChainId.BNB,
  label: "BNB",
  eip155Label: "BNB Chain Mainnet",
  logo: { light: bnbLogo, dark: bnbLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.BNB,
    name: "BNB Chain Native Token",
    symbol: "BNB",
    logoUrl: bnbLogo
  },
  addressPrefix: "bnb",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 15921452
    }
  },
  rpcUrls: {
    default: {
      http: ["https://bsc-dataseed1.bnbchain.org"]
    }
  },
  color: "#F0B90B",
  website: {
    name: "BNB Chain",
    url: "https://www.bnbchain.org"
  },
  docs: {
    name: "BNB Chain Docs",
    url: "https://docs.bnbchain.org"
  },
  blockExplorer: {
    name: "Bscscan",
    url: "https://bscscan.com"
  },
  bridges: [
    {
      name: "BNB Chain Cross-Chain Bridge",
      url: "https://www.bnbchain.org/en/bnb-chain-bridge"
    }
  ]
};

// src/chains/details/optimism.ts
var optimismLogo = `${RAW_CHAINS_FILES_PATH}/images/optimism-logo.svg`;
var optimism = {
  id: 10,
  label: "Optimism",
  eip155Label: "OP Mainnet",
  logo: { light: optimismLogo, dark: optimismLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: 10
  },
  addressPrefix: "op",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 4286263
    }
  },
  rpcUrls: {
    default: {
      http: ["https://mainnet.optimism.io"]
    }
  },
  color: "#ff0420",
  website: {
    name: "Optimism",
    url: "https://optimism.io"
  },
  docs: {
    name: "Optimism Docs",
    url: "https://docs.optimism.io"
  },
  blockExplorer: {
    name: "Etherscan",
    url: "https://optimistic.etherscan.io"
  }
};

// src/chains/details/linea.ts
var lineaLogo = `${RAW_CHAINS_FILES_PATH}/images/linea-logo.svg`;
var linea = {
  id: SupportedChainId.LINEA,
  label: "Linea",
  eip155Label: "Linea Mainnet",
  logo: { light: lineaLogo, dark: lineaLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.LINEA
  },
  addressPrefix: "linea",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 42
    }
  },
  rpcUrls: {
    default: {
      http: ["https://rpc.linea.build"]
    }
  },
  color: "#61dfff",
  // Brand assets https://linea.build/assets
  website: {
    name: "Linea",
    url: "https://linea.build"
  },
  docs: {
    name: "Linea Docs",
    url: "https://docs.linea.build"
  },
  blockExplorer: {
    name: "Lineascan",
    url: "https://lineascan.build"
  },
  bridges: [
    {
      name: "Linea Bridge",
      url: "https://linea.build/hub/bridge"
    }
  ]
};

// src/chains/details/plasma.ts
var light = `${RAW_CHAINS_FILES_PATH}/images/plasma-logo.svg`;
var dark = light;
var plasma = {
  id: SupportedChainId.PLASMA,
  label: "Plasma",
  eip155Label: "Plasma Mainnet",
  logo: { light, dark },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.PLASMA,
    name: "Plasma",
    symbol: "XPL",
    logoUrl: light
  },
  addressPrefix: "plasma",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xca11bde05977b3631167028862be2a173976ca11",
      blockCreated: 0
    }
  },
  rpcUrls: {
    default: {
      http: ["https://rpc.plasma.to"]
    }
  },
  color: "#569F8C",
  // brand kit https://www.plasma.to/brand
  website: {
    name: "Plasma",
    url: "https://www.plasma.to"
  },
  docs: {
    name: "Plasma Docs",
    url: "https://docs.plasma.to"
  },
  blockExplorer: {
    name: "Plasmascan",
    url: "https://plasmascan.to"
  }
  // No native bridge available AFAICT
  // bridges: [
  //   {
  //     name: 'Plasma Bridge',
  //     url: '',
  //   },
  // ],
};

// src/chains/details/ink.ts
var inkLogo = `${RAW_CHAINS_FILES_PATH}/images/ink-logo.svg`;
var ink = {
  id: SupportedChainId.INK,
  label: "Ink",
  eip155Label: "Ink Chain Mainnet",
  logo: { light: inkLogo, dark: inkLogo },
  nativeCurrency: {
    ...nativeCurrencyTemplate,
    chainId: SupportedChainId.INK
  },
  addressPrefix: "ink",
  isTestnet: false,
  contracts: {
    multicall3: {
      address: "0xcA11bde05977b3631167028862bE2a173976CA11",
      blockCreated: 0
    }
  },
  rpcUrls: {
    default: {
      http: ["https://rpc-ten.inkonchain.com"]
    }
  },
  color: "#7132f5",
  // Brand assets https://docs.inkonchain.com/work-with-ink/brand-kit
  website: {
    name: "Ink",
    url: "https://inkonchain.com/"
  },
  docs: {
    name: "Ink Docs",
    url: "https://docs.inkonchain.com"
  },
  blockExplorer: {
    name: "Ink Explorer",
    url: "https://explorer.inkonchain.com"
    // Note: doesn't seem to be Etherscan derived. Will need to consider it in the code.
    // TODO: add a flag to indicate that the explorer is not Etherscan derived and handle it in the code..
  },
  bridges: [
    {
      name: "Ink Bridge",
      url: "https://inkonchain.com/bridge"
    }
  ]
};

// src/chains/details/bitcoin.ts
var bitcoinLogo = `${RAW_CHAINS_FILES_PATH}/images/bitcoin-logo.svg`;
var bitcoin = {
  id: NonEvmChains.BITCOIN,
  label: "Bitcoin",
  logo: { light: bitcoinLogo, dark: bitcoinLogo },
  nativeCurrency: {
    chainId: NonEvmChains.BITCOIN,
    address: BTC_CURRENCY_ADDRESS,
    name: "Bitcoin",
    symbol: "BTC",
    decimals: 8,
    logoUrl: bitcoinLogo
  },
  addressPrefix: "btc",
  isTestnet: false,
  color: "#f7931a",
  website: {
    name: "Bitcoin",
    url: "https://bitcoin.org"
  },
  docs: {
    name: "Bitcoin Docs",
    url: "https://bitcoin.org/en/developer-documentation"
  },
  blockExplorer: {
    name: "Blockstream Explorer",
    url: "https://blockstream.info"
  }
};

// src/chains/details/solana.ts
var solanaLogo = `${RAW_CHAINS_FILES_PATH}/images/solana-logo.svg`;
var solana = {
  id: NonEvmChains.SOLANA,
  label: "Solana",
  logo: { light: solanaLogo, dark: solanaLogo },
  nativeCurrency: {
    chainId: NonEvmChains.SOLANA,
    address: SOL_NATIVE_CURRENCY_ADDRESS,
    name: "Solana",
    symbol: "SOL",
    decimals: 9,
    logoUrl: solanaLogo
  },
  addressPrefix: "sol",
  isTestnet: false,
  color: "#9945FF",
  website: {
    name: "Solana",
    url: "https://solana.com"
  },
  docs: {
    name: "Solana Docs",
    url: "https://docs.solana.com"
  },
  blockExplorer: {
    name: "Solana Explorer",
    url: "https://explorer.solana.com"
  }
};

// src/chains/const/chainIds.ts
var ALL_SUPPORTED_CHAINS_MAP = {
  [SupportedChainId.MAINNET]: mainnet,
  [SupportedChainId.GNOSIS_CHAIN]: gnosisChain,
  [SupportedChainId.ARBITRUM_ONE]: arbitrumOne,
  [SupportedChainId.BASE]: base,
  [SupportedChainId.AVALANCHE]: avalanche,
  [SupportedChainId.POLYGON]: polygon,
  [SupportedChainId.BNB]: bnb,
  [SupportedChainId.PLASMA]: plasma,
  [SupportedChainId.LINEA]: linea,
  [SupportedChainId.INK]: ink,
  [SupportedChainId.SEPOLIA]: sepolia
};
var ALL_SUPPORTED_CHAINS = Object.values(ALL_SUPPORTED_CHAINS_MAP);
var ALL_SUPPORTED_CHAIN_IDS = ALL_SUPPORTED_CHAINS.map(
  (chain) => chain.id
);
var TRADABLE_SUPPORTED_CHAINS = Object.values(ALL_SUPPORTED_CHAINS_MAP).filter(
  (chain) => !chain.isDeprecated
);
var TRADABLE_SUPPORTED_CHAIN_IDS = TRADABLE_SUPPORTED_CHAINS.map(
  (chain) => chain.id
);
var ADDITIONAL_TARGET_CHAINS_MAP = {
  [AdditionalTargetChainId.OPTIMISM]: optimism,
  [AdditionalTargetChainId.BITCOIN]: bitcoin,
  [AdditionalTargetChainId.SOLANA]: solana
};
var ALL_ADDITIONAL_TARGET_CHAINS = Object.values(ADDITIONAL_TARGET_CHAINS_MAP);
var ALL_ADDITIONAL_TARGET_CHAIN_IDS = ALL_ADDITIONAL_TARGET_CHAINS.map(
  (chain) => chain.id
);
var ALL_CHAINS = ALL_SUPPORTED_CHAINS.concat(ALL_ADDITIONAL_TARGET_CHAINS);
var ALL_CHAINS_IDS = ALL_CHAINS.map((chain) => chain.id);

// src/chains/const/utils.ts
function mapSupportedNetworks(value) {
  return ALL_SUPPORTED_CHAIN_IDS.reduce(
    (acc, chainId) => ({
      ...acc,
      [chainId]: typeof value === "function" ? value(chainId) : value
    }),
    {}
  );
}
function mapAllNetworks(value) {
  return ALL_CHAINS_IDS.reduce(
    (acc, chainId) => ({
      ...acc,
      [chainId]: typeof value === "function" ? value(chainId) : value
    }),
    {}
  );
}
function mapAddressToSupportedNetworks(address) {
  return mapSupportedNetworks(address);
}

// src/chains/const/contracts.ts
var ETH_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
var EXTENSIBLE_FALLBACK_HANDLER = "0x2f55e8b20D0B9FEFA187AA7d00B6Cbe563605bF5";
var COMPOSABLE_COW = "0xfdaFc9d1902f4e0b84f65F49f244b32b31013b74";
var VAULT_RELAYER = "0xC92E8bdf79f0507f65a392b0ab4667716BFE0110";
var VAULT_RELAYER_STAGING = "0xC7242d167563352E2BCA4d71C043fbe542DB8FB2";
var SETTLEMENT_CONTRACT = "0x9008D19f58AAbD9eD0D60971565AA8510560ab41";
var SETTLEMENT_CONTRACT_STAGING = "0xf553d092b50bdcbddeD1A99aF2cA29FBE5E2CB13";
var COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS = mapAddressToSupportedNetworks(SETTLEMENT_CONTRACT);
var COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING = mapAddressToSupportedNetworks(SETTLEMENT_CONTRACT_STAGING);
var COW_PROTOCOL_VAULT_RELAYER_ADDRESS = mapAddressToSupportedNetworks(VAULT_RELAYER);
var COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING = mapAddressToSupportedNetworks(VAULT_RELAYER_STAGING);
var EXTENSIBLE_FALLBACK_HANDLER_CONTRACT_ADDRESS = mapAddressToSupportedNetworks(EXTENSIBLE_FALLBACK_HANDLER);
var COMPOSABLE_COW_CONTRACT_ADDRESS = mapAddressToSupportedNetworks(COMPOSABLE_COW);
var ETH_FLOW_ADDRESS = "0xba3cb449bd2b4adddbc894d8697f5170800eadec";
var BARN_ETH_FLOW_ADDRESS = "0xb37aDD6AC288BD3825a901Cba6ec65A89f31B8CC";
var ETH_FLOW_ADDRESSES = mapAddressToSupportedNetworks(ETH_FLOW_ADDRESS);
var BARN_ETH_FLOW_ADDRESSES = mapAddressToSupportedNetworks(BARN_ETH_FLOW_ADDRESS);
var MAX_VALID_TO_EPOCH = 4294967295;

// src/chains/utils.ts
function isEvmChain(chainId) {
  return chainId in EvmChains;
}
function isNonEvmChain(chainId) {
  return chainId in NonEvmChains;
}
function isBtcChain(chainId) {
  return chainId === NonEvmChains.BITCOIN;
}
function isEvmChainInfo(chainInfo) {
  return isEvmChain(chainInfo.id);
}
function isNonEvmChainInfo(chainInfo) {
  return isNonEvmChain(chainInfo.id);
}
function getChainInfo(chainId) {
  if (isSupportedChain(chainId)) {
    return ALL_SUPPORTED_CHAINS_MAP[chainId];
  }
  if (isAdditionalTargetChain(chainId)) {
    return ADDITIONAL_TARGET_CHAINS_MAP[chainId];
  }
  return void 0;
}
function isSupportedChain(chainId) {
  return chainId in SupportedChainId;
}
function isAdditionalTargetChain(chainId) {
  return chainId in AdditionalTargetChainId;
}
function isTargetChainId(chainId) {
  return isSupportedChain(chainId) || isAdditionalTargetChain(chainId);
}
function isZkSyncChain(chainId) {
  if (!isEvmChain(chainId))
    return false;
  return Boolean(getChainInfo(chainId)?.isZkSync);
}
function isChainUnderDevelopment(chainId) {
  return Boolean(getChainInfo(chainId)?.isUnderDevelopment);
}
function isChainDeprecated(chainId) {
  return Boolean(getChainInfo(chainId)?.isDeprecated);
}

// src/types/configs.ts
var ENVS_LIST = ["prod", "staging"];
var DEFAULT_COW_API_CONTEXT = {
  env: "prod",
  chainId: SupportedChainId.MAINNET
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ADDITIONAL_TARGET_CHAINS_MAP,
  ALL_ADDITIONAL_TARGET_CHAINS,
  ALL_ADDITIONAL_TARGET_CHAIN_IDS,
  ALL_CHAINS,
  ALL_CHAINS_IDS,
  ALL_SUPPORTED_CHAINS,
  ALL_SUPPORTED_CHAINS_MAP,
  ALL_SUPPORTED_CHAIN_IDS,
  AdditionalTargetChainId,
  BARN_ETH_FLOW_ADDRESSES,
  BTC_CURRENCY_ADDRESS,
  COMPOSABLE_COW,
  COMPOSABLE_COW_CONTRACT_ADDRESS,
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS,
  COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING,
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS,
  COW_PROTOCOL_VAULT_RELAYER_ADDRESS_STAGING,
  DEFAULT_COW_API_CONTEXT,
  DEFAULT_IPFS_READ_URI,
  DEFAULT_IPFS_WRITE_URI,
  ENVS_LIST,
  ETH_ADDRESS,
  ETH_FLOW_ADDRESSES,
  EVM_NATIVE_CURRENCY_ADDRESS,
  EXTENSIBLE_FALLBACK_HANDLER,
  EXTENSIBLE_FALLBACK_HANDLER_CONTRACT_ADDRESS,
  EvmChains,
  MAX_VALID_TO_EPOCH,
  NonEvmChains,
  RAW_CHAINS_FILES_PATH,
  RAW_FILES_PATH,
  SOL_NATIVE_CURRENCY_ADDRESS,
  SupportedChainId,
  TOKEN_LIST_IMAGES_PATH,
  TRADABLE_SUPPORTED_CHAINS,
  TRADABLE_SUPPORTED_CHAIN_IDS,
  WRAPPED_NATIVE_CURRENCIES,
  arbitrumOne,
  avalanche,
  base,
  bitcoin,
  bnb,
  getChainInfo,
  getWrappedTokenForChain,
  gnosisChain,
  ink,
  isAdditionalTargetChain,
  isBtcChain,
  isChainDeprecated,
  isChainUnderDevelopment,
  isEvmChain,
  isEvmChainInfo,
  isNonEvmChain,
  isNonEvmChainInfo,
  isSupportedChain,
  isTargetChainId,
  isZkSyncChain,
  linea,
  mainnet,
  mapAddressToSupportedNetworks,
  mapAllNetworks,
  mapSupportedNetworks,
  nativeCurrencyTemplate,
  optimism,
  plasma,
  polygon,
  sepolia,
  solana
});
