// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/utils/limit-order-rfq.const.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.explorersUrls = exports.contractAddresses = exports.rpcUrls = exports.cancelOrderSchema = exports.fillOrderSchema = exports.createOrderSchema = exports.operationSchema = void 0;
var limit_order_protocol_model_1 = require("../model/limit-order-protocol.model");
var commonProperties = [
    {
        type: 'select',
        name: 'chainId',
        message: 'Select network',
        choices: [
            { title: 'Ethereum', value: limit_order_protocol_model_1.ChainId.etherumMainnet },
            { title: 'BSC', value: limit_order_protocol_model_1.ChainId.binanceMainnet },
            { title: 'Polygon', value: limit_order_protocol_model_1.ChainId.polygonMainnet },
            { title: 'Optimism', value: limit_order_protocol_model_1.ChainId.optimismMainnet },
            { title: 'Arbitrum', value: limit_order_protocol_model_1.ChainId.arbitrumMainnet },
        ],
    },
    {
        type: 'password',
        name: 'privateKey',
        message: 'Enter your private key',
    },
];
exports.operationSchema = [
    {
        type: 'select',
        name: 'operation',
        choices: [
            { title: 'create', value: 'create' },
            { title: 'fill', value: 'fill' },
            { title: 'cancel', value: 'cancel' },
        ],
        message: 'Choose operation for limit order RFQ: create, fill, cancel',
    },
];
exports.createOrderSchema = __spreadArray(__spreadArray([], commonProperties, true), [
    {
        type: 'number',
        name: 'orderId',
        message: 'Limit order RFQ id',
    },
    {
        type: 'number',
        name: 'expiresIn',
        message: 'Expires in (seconds, for example: 300 - order will expired in 5 mins)',
        initial: 300,
    },
    {
        type: 'text',
        name: 'makerAssetAddress',
        message: 'Maker asset address',
    },
    {
        type: 'text',
        name: 'takerAssetAddress',
        message: 'Taker asset address',
    },
    {
        type: 'text',
        name: 'makerAmount',
        message: 'Maker asset amount',
    },
    {
        type: 'text',
        name: 'takerAmount',
        message: 'Taker asset amount',
    },
    {
        type: 'text',
        name: 'takerAddress',
        message: 'Taker address (optional)',
    },
], false);
exports.fillOrderSchema = __spreadArray(__spreadArray([], commonProperties, true), [
    {
        type: 'number',
        name: 'gasPrice',
        message: 'Gas price (GWEI)',
        initial: 10,
    },
    {
        type: 'text',
        name: 'order',
        message: 'Limit order RFQ json',
    },
    {
        type: 'text',
        name: 'makerAmount',
        message: 'Maker asset fill amount (set 0 if you will use taker asset amount)',
    },
    {
        type: 'text',
        name: 'takerAmount',
        message: 'Taker asset amount (set 0 if has set maker asset amount)',
    },
], false);
exports.cancelOrderSchema = __spreadArray(__spreadArray([], commonProperties, true), [
    {
        type: 'number',
        name: 'gasPrice',
        message: 'Gas price (GWEI)',
        initial: 10,
    },
    {
        type: 'text',
        name: 'orderInfo',
        message: 'Order info',
    },
], false);
exports.rpcUrls = (_a = {},
    _a[limit_order_protocol_model_1.ChainId.etherumMainnet] = 'https://mainnet.infura.io/v3',
    _a[limit_order_protocol_model_1.ChainId.binanceMainnet] = 'https://bsc-dataseed.binance.org',
    _a[limit_order_protocol_model_1.ChainId.polygonMainnet] = 'https://polygon-rpc.com',
    _a[limit_order_protocol_model_1.ChainId.optimismMainnet] = 'https://rpc.ankr.com/optimism',
    _a[limit_order_protocol_model_1.ChainId.arbitrumMainnet] = 'https://arb1.arbitrum.io/rpc',
    _a);
exports.contractAddresses = (_b = {},
    _b[limit_order_protocol_model_1.ChainId.etherumMainnet] = '0x119c71d3bbac22029622cbaec24854d3d32d2828',
    _b[limit_order_protocol_model_1.ChainId.binanceMainnet] = '0x1e38eff998df9d3669e32f4ff400031385bf6362',
    _b[limit_order_protocol_model_1.ChainId.polygonMainnet] = '0x94bc2a1c732bcad7343b25af48385fe76e08734f',
    _b[limit_order_protocol_model_1.ChainId.optimismMainnet] = '0x11431a89893025d2a48dca4eddc396f8c8117187',
    _b[limit_order_protocol_model_1.ChainId.arbitrumMainnet] = '0x7f069df72b7a39bce9806e3afaf579e54d8cf2b9',
    _b);
exports.explorersUrls = (_c = {},
    _c[limit_order_protocol_model_1.ChainId.etherumMainnet] = 'https://etherscan.io',
    _c[limit_order_protocol_model_1.ChainId.binanceMainnet] = 'https://bscscan.com',
    _c[limit_order_protocol_model_1.ChainId.polygonMainnet] = 'https://polygonscan.com',
    _c[limit_order_protocol_model_1.ChainId.optimismMainnet] = 'https://optimistic.etherscan.io',
    _c[limit_order_protocol_model_1.ChainId.arbitrumMainnet] = 'https://arbiscan.io',
    _c);
