// @openocean.finance/limitorder-sdk/lib/utils/limit-order-sdk/utils/limit-order-rfq.helpers.js
#!/usr/bin/env node
"use strict";
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.gweiToWei = exports.sendSignedTransaction = exports.cancelOrder = exports.fillOrder = exports.createOrder = exports.cancelOrderOperation = exports.fillOrderOperation = exports.createOrderOperation = void 0;
var tslib_1 = require("tslib");
var prompts_1 = tslib_1.__importDefault(require("prompts"));
var kleur_1 = tslib_1.__importDefault(require("kleur"));
var web3_1 = tslib_1.__importDefault(require("web3"));
var limit_order_builder_1 = require("../limit-order.builder");
var limit_order_protocol_facade_1 = require("../limit-order-protocol.facade");
var limit_order_rfq_const_1 = require("./limit-order-rfq.const");
var private_key_provider_connector_1 = require("../connector/private-key-provider.connector");
function createOrderOperation(isRunningWithArgv, params) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var creatingParams, _a, newOrder;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = params;
                    if (_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, prompts_1.default(limit_order_rfq_const_1.createOrderSchema)];
                case 1:
                    _a = (_b.sent());
                    _b.label = 2;
                case 2:
                    creatingParams = _a;
                    newOrder = createOrder(creatingParams);
                    if (isRunningWithArgv) {
                        console.log(JSON.stringify(newOrder));
                        return [2 /*return*/];
                    }
                    console.log(kleur_1.default.green().bold('New limit order RFQ: '));
                    console.log(kleur_1.default.white().underline(JSON.stringify(newOrder, null, 4)));
                    return [2 /*return*/];
            }
        });
    });
}
exports.createOrderOperation = createOrderOperation;
function fillOrderOperation(isRunningWithArgv, params) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var fillingParams, _a, orderForFill, txHash;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = params;
                    if (_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, prompts_1.default(limit_order_rfq_const_1.fillOrderSchema)];
                case 1:
                    _a = (_b.sent());
                    _b.label = 2;
                case 2:
                    fillingParams = _a;
                    orderForFill = JSON.parse(fillingParams.order);
                    console.log(kleur_1.default.green().bold('Order for filling: '));
                    console.log(kleur_1.default.white().underline(JSON.stringify(orderForFill, null, 4)));
                    return [4 /*yield*/, fillOrder(fillingParams, orderForFill)];
                case 3:
                    txHash = _b.sent();
                    if (isRunningWithArgv) {
                        console.log(txHash);
                        return [2 /*return*/];
                    }
                    console.log(kleur_1.default.green().bold('Order filling transaction: '));
                    printTransactionLink(explorerTxLink(fillingParams.chainId, txHash));
                    return [2 /*return*/];
            }
        });
    });
}
exports.fillOrderOperation = fillOrderOperation;
function cancelOrderOperation(isRunningWithArgv, params) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var cancelingParams, _a, cancelingTxHash;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = params;
                    if (_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, prompts_1.default(limit_order_rfq_const_1.cancelOrderSchema)];
                case 1:
                    _a = (_b.sent());
                    _b.label = 2;
                case 2:
                    cancelingParams = _a;
                    return [4 /*yield*/, cancelOrder(cancelingParams)];
                case 3:
                    cancelingTxHash = _b.sent();
                    if (isRunningWithArgv) {
                        console.log(cancelingTxHash);
                        return [2 /*return*/];
                    }
                    console.log(kleur_1.default.green().bold('Order canceling transaction: '));
                    printTransactionLink(explorerTxLink(cancelingParams.chainId, cancelingTxHash));
                    return [2 /*return*/];
            }
        });
    });
}
exports.cancelOrderOperation = cancelOrderOperation;
/* eslint-disable max-lines-per-function */
function createOrder(params) {
    var contractAddress = limit_order_rfq_const_1.contractAddresses[params.chainId];
    var web3 = new web3_1.default(limit_order_rfq_const_1.rpcUrls[params.chainId]);
    var providerConnector = new private_key_provider_connector_1.PrivateKeyProviderConnector(params.privateKey, web3);
    var walletAddress = web3.eth.accounts.privateKeyToAccount(params.privateKey).address;
    var limitOrderBuilder = new limit_order_builder_1.LimitOrderBuilder(contractAddress, params.chainId, providerConnector);
    return limitOrderBuilder.buildRFQOrder({
        id: params.orderId,
        expiresInTimestamp: Math.ceil(Date.now() / 1000) + params.expiresIn,
        makerAddress: walletAddress,
        makerAssetAddress: params.makerAssetAddress,
        takerAssetAddress: params.takerAssetAddress,
        makerAmount: params.makerAmount,
        takerAmount: params.takerAmount,
        takerAddress: params.takerAddress || undefined,
    });
}
exports.createOrder = createOrder;
/* eslint-enable max-lines-per-function */
/* eslint-disable max-lines-per-function */
function fillOrder(params, order) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var contractAddress, web3, providerConnector, walletAddress, limitOrderBuilder, limitOrderProtocolFacade, typedData, signature, callData, txConfig, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    contractAddress = limit_order_rfq_const_1.contractAddresses[params.chainId];
                    web3 = new web3_1.default(limit_order_rfq_const_1.rpcUrls[params.chainId]);
                    providerConnector = new private_key_provider_connector_1.PrivateKeyProviderConnector(params.privateKey, web3);
                    walletAddress = web3.eth.accounts.privateKeyToAccount(params.privateKey).address;
                    limitOrderBuilder = new limit_order_builder_1.LimitOrderBuilder(contractAddress, params.chainId, providerConnector);
                    limitOrderProtocolFacade = new limit_order_protocol_facade_1.LimitOrderProtocolFacade(contractAddress, providerConnector);
                    typedData = limitOrderBuilder.buildRFQOrderTypedData(order, params.domainName || undefined);
                    return [4 /*yield*/, limitOrderBuilder.buildOrderSignature(walletAddress, typedData)];
                case 1:
                    signature = _b.sent();
                    callData = limitOrderProtocolFacade.fillRFQOrder(order, signature, params.makerAmount, params.takerAmount);
                    _a = {
                        to: contractAddress,
                        from: walletAddress,
                        data: callData,
                        value: '0',
                        gas: 120000,
                        gasPrice: gweiToWei(params.gasPrice)
                    };
                    return [4 /*yield*/, web3.eth.getTransactionCount(walletAddress)];
                case 2:
                    txConfig = (_a.nonce = _b.sent(),
                        _a);
                    return [2 /*return*/, sendSignedTransaction(web3, txConfig, params.privateKey)];
            }
        });
    });
}
exports.fillOrder = fillOrder;
/* eslint-enable max-lines-per-function */
/* eslint-disable max-lines-per-function */
function cancelOrder(params) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var contractAddress, web3, providerConnector, walletAddress, limitOrderProtocolFacade, callData, txConfig, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    contractAddress = limit_order_rfq_const_1.contractAddresses[params.chainId];
                    web3 = new web3_1.default(new web3_1.default.providers.HttpProvider(limit_order_rfq_const_1.rpcUrls[params.chainId]));
                    providerConnector = new private_key_provider_connector_1.PrivateKeyProviderConnector(params.privateKey, web3);
                    walletAddress = web3.eth.accounts.privateKeyToAccount(params.privateKey).address;
                    limitOrderProtocolFacade = new limit_order_protocol_facade_1.LimitOrderProtocolFacade(contractAddress, providerConnector);
                    callData = limitOrderProtocolFacade.cancelRFQOrder(params.orderInfo);
                    _a = {
                        to: contractAddress,
                        from: walletAddress,
                        data: callData,
                        value: '0',
                        gas: 50000,
                        gasPrice: gweiToWei(params.gasPrice)
                    };
                    return [4 /*yield*/, web3.eth.getTransactionCount(walletAddress)];
                case 1:
                    txConfig = (_a.nonce = _b.sent(),
                        _a);
                    return [2 /*return*/, sendSignedTransaction(web3, txConfig, params.privateKey)];
            }
        });
    });
}
exports.cancelOrder = cancelOrder;
/* eslint-enable max-lines-per-function */
function sendSignedTransaction(web3, txConfig, privateKey) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var sign;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, web3.eth.accounts.signTransaction(txConfig, privateKey)];
                case 1:
                    sign = _a.sent();
                    return [4 /*yield*/, new Promise(function (resolve, reject) {
                            web3.eth.sendSignedTransaction(sign.rawTransaction, function (error, hash) {
                                if (error) {
                                    reject(error);
                                    return;
                                }
                                resolve(hash);
                            });
                        })];
                case 2: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.sendSignedTransaction = sendSignedTransaction;
function explorerTxLink(chainId, txHash) {
    var explorerUrl = limit_order_rfq_const_1.explorersUrls[chainId];
    return "".concat(explorerUrl, "/tx/").concat(txHash);
}
function gweiToWei(value) {
    return value + '000000000';
}
exports.gweiToWei = gweiToWei;
function printTransactionLink(text) {
    console.log(kleur_1.default.white('************************************************'));
    console.log(kleur_1.default.white('   '));
    console.log(kleur_1.default.white().underline(text));
    console.log(kleur_1.default.white('   '));
    console.log(kleur_1.default.white('************************************************'));
}
