// @cowprotocol/sdk-order-signing/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  COW_EIP712_TYPES: () => COW_EIP712_TYPES,
  ORDER_PRIMARY_TYPE: () => ORDER_PRIMARY_TYPE,
  OrderSigningUtils: () => OrderSigningUtils
});
module.exports = __toCommonJS(src_exports);

// src/orderSigningUtils.ts
var import_sdk_common2 = require("@cowprotocol/sdk-common");

// src/utils.ts
var import_sdk_contracts_ts = require("@cowprotocol/sdk-contracts-ts");
var import_sdk_common = require("@cowprotocol/sdk-common");
var import_sdk_order_book = require("@cowprotocol/sdk-order-book");
var import_sdk_config = require("@cowprotocol/sdk-config");
var METAMASK_SIGNATURE_ERROR_CODE = -32603;
var METHOD_NOT_FOUND_ERROR_CODE = -32601;
var METHOD_NOT_FOUND_ERROR_MSG_REGEX = /Method not found/i;
var V4_ERROR_MSG_REGEX = /eth_signTypedData_v4 does not exist/i;
var V3_ERROR_MSG_REGEX = /eth_signTypedData_v3 does not exist/i;
var RPC_REQUEST_FAILED_REGEX = /RPC request failed/i;
var METAMASK_STRING_CHAINID_REGEX = /provided chainid .* must match the active chainid/i;
var mapSigningSchema = {
  [import_sdk_order_book.EcdsaSigningScheme.EIP712]: import_sdk_contracts_ts.ContractsSigningScheme.EIP712,
  [import_sdk_order_book.EcdsaSigningScheme.ETHSIGN]: import_sdk_contracts_ts.ContractsSigningScheme.ETHSIGN
};
function isProviderRpcError(error) {
  return error.code !== void 0 || error.message !== void 0;
}
async function _signOrder(params) {
  const { chainId, signer, order, signingScheme, env, settlementContractOverride } = params;
  const domain = getDomain(chainId, { env, settlementContractOverride });
  return (0, import_sdk_contracts_ts.signOrder)(domain, order, mapSigningSchema[signingScheme], signer);
}
async function _signOrderCancellation(params) {
  const { chainId, signer, signingScheme, orderUid, env, settlementContractOverride } = params;
  const domain = getDomain(chainId, { env, settlementContractOverride });
  return (0, import_sdk_contracts_ts.signOrderCancellation)(domain, orderUid, mapSigningSchema[signingScheme], signer);
}
async function _signOrderCancellations(params) {
  const { chainId, signer, signingScheme, orderUids, env, settlementContractOverride } = params;
  const domain = getDomain(chainId, { env, settlementContractOverride });
  return (0, import_sdk_contracts_ts.signOrderCancellations)(domain, orderUids, mapSigningSchema[signingScheme], signer);
}
async function _signPayload(payload, signFn, signer, signingMethod = "v4") {
  const signingScheme = signingMethod === "eth_sign" ? import_sdk_order_book.EcdsaSigningScheme.ETHSIGN : import_sdk_order_book.EcdsaSigningScheme.EIP712;
  let signature = null;
  let _signer;
  try {
    switch (signingMethod) {
      case "default":
      case "v3":
        _signer = (0, import_sdk_contracts_ts.getTypedDataVersionedSigner)(signer, "v3");
        break;
      case "int_v4":
        _signer = (0, import_sdk_contracts_ts.getIntChainIdTypedDataV4Signer)(signer);
        break;
      default:
        _signer = signer;
    }
  } catch (e) {
    console.error("Wallet not supported:", e);
    throw new import_sdk_common.CowError("Wallet not supported");
  }
  try {
    signature = await signFn({ ...payload, signer: _signer, signingScheme });
  } catch (e) {
    if (!isProviderRpcError(e)) {
      console.error(e);
      throw e;
    }
    const regexErrorCheck = [METHOD_NOT_FOUND_ERROR_MSG_REGEX, RPC_REQUEST_FAILED_REGEX].some(
      (regex) => (
        // for example 1Inch error doesn't have e.message so we will check the output of toString()
        [e.message, e.toString()].some((msg) => regex.test(msg))
      )
    );
    if (e.code === METHOD_NOT_FOUND_ERROR_CODE || regexErrorCheck) {
      switch (signingMethod) {
        case "v4":
          return _signPayload(payload, signFn, signer, "default");
        case "default":
          return _signPayload(payload, signFn, signer, "v3");
        case "v3":
          return _signPayload(payload, signFn, signer, "eth_sign");
        default:
          throw e;
      }
    } else if (METAMASK_STRING_CHAINID_REGEX.test(e.message)) {
      return _signPayload(payload, signFn, signer, "int_v4");
    } else if (e.code === METAMASK_SIGNATURE_ERROR_CODE) {
      return _signPayload(payload, signFn, signer, "eth_sign");
    } else if (V4_ERROR_MSG_REGEX.test(e.message)) {
      return _signPayload(payload, signFn, signer, "v3");
    } else if (V3_ERROR_MSG_REGEX.test(e.message)) {
      return _signPayload(payload, signFn, signer, "eth_sign");
    } else {
      console.error(e);
      throw e;
    }
  }
  const data = signature?.data;
  return { signature: data?.toString() || "", signingScheme };
}
async function signOrder(order, chainId, signer, options) {
  const { env, settlementContractOverride } = options ?? {};
  return _signPayload({ order, chainId, env, settlementContractOverride }, _signOrder, signer);
}
async function signOrderCancellation(orderUid, chainId, signer, options) {
  const { env, settlementContractOverride } = options ?? {};
  return _signPayload({ orderUid, chainId, env, settlementContractOverride }, _signOrderCancellation, signer);
}
async function signOrderCancellations(orderUids, chainId, signer, options) {
  const { env, settlementContractOverride } = options ?? {};
  return _signPayload({ orderUids, chainId, env, settlementContractOverride }, _signOrderCancellations, signer);
}
function getDomain(chainId, options) {
  const { env, settlementContractOverride } = options ?? {};
  const settlementContract = settlementContractOverride?.[chainId] ?? (env === "staging" ? import_sdk_config.COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS_STAGING[chainId] : import_sdk_config.COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS[chainId]);
  if (!settlementContract) {
    throw new import_sdk_common.CowError("Unsupported network. Settlement contract is not deployed");
  }
  return import_sdk_contracts_ts.ContractsTs.domain(chainId, settlementContract);
}
async function generateOrderId(chainId, order, params, options) {
  const domain = await getDomain(chainId, options);
  const orderDigest = (0, import_sdk_contracts_ts.hashOrder)(domain, order);
  const orderId = (0, import_sdk_contracts_ts.packOrderUidParams)({
    ...params,
    orderDigest,
    // Different validTo when signing because EthFlow contract expects it to be max for all orders
    validTo: order.validTo
  });
  return { orderId, orderDigest };
}

// src/orderSigningUtils.ts
var import_sdk_order_book2 = require("@cowprotocol/sdk-order-book");
var ORDER_PRIMARY_TYPE = "Order";
var COW_EIP712_TYPES = {
  [ORDER_PRIMARY_TYPE]: [
    { name: "sellToken", type: "address" },
    { name: "buyToken", type: "address" },
    { name: "receiver", type: "address" },
    { name: "sellAmount", type: "uint256" },
    { name: "buyAmount", type: "uint256" },
    { name: "validTo", type: "uint32" },
    { name: "appData", type: "bytes32" },
    { name: "feeAmount", type: "uint256" },
    { name: "kind", type: "string" },
    { name: "partiallyFillable", type: "bool" },
    { name: "sellTokenBalance", type: "string" },
    { name: "buyTokenBalance", type: "string" }
  ]
};
var OrderSigningUtils = class _OrderSigningUtils {
  /**
   * Sign the order intent with the specified signer.
   *
   * @remarks If the API reports an error with the signature, it is likely to be due to an incorrectly
   *          specified `chainId`. Please ensure that the `chainId` is correct for the network you are
   *          using.
   * @param {UnsignedOrder} order The unsigned order intent to be placed.
   * @param {SupportedChainId} chainId The CoW Protocol `chainId` context that's being used.
   * @param {Signer} signer The signer who is placing the order intent.
   * @returns {Promise<SigningResult>} Encoded signature including signing scheme for the order.
   */
  static async signOrder(order, chainId, signer, options) {
    return signOrder(order, chainId, signer, options);
  }
  /**
   * Sign a cancellation message of an order intent with the specified signer.
   * @param {string} orderUid The unique identifier of the order to cancel.
   * @param {SupportedChainId} chainId The CoW Protocol `chainid` context that's being used.
   * @param {Signer} signer The signer who initially placed the order intent.
   * @returns {Promise<SigningResult>} Encoded signature including signing scheme for the cancellation.
   */
  static async signOrderCancellation(orderUid, chainId, signer, options) {
    return signOrderCancellation(orderUid, chainId, signer, options);
  }
  /**
   * Sign a cancellation message of multiple order intents with the specified signer.
   * @param {string[]} orderUids An array of `orderUid` to cancel.
   * @param {SupportedChainId} chainId The CoW Protocol `chainId` context that's being used.
   * @param {Signer} signer The signer who initially placed the order intents.
   * @returns {Promise<SigningResult>} Encoded signature including signing scheme for the cancellation.
   */
  static async signOrderCancellations(orderUids, chainId, signer, options) {
    return signOrderCancellations(orderUids, chainId, signer, options);
  }
  /**
   * Get the EIP-712 typed domain data being used for signing.
   * @param {SupportedChainId} chainId The CoW Protocol `chainId` context that's being used.
   * @return The EIP-712 typed domain data.
   * @see https://eips.ethereum.org/EIPS/eip-712
   */
  static async getDomain(chainId, options) {
    return getDomain(chainId, options);
  }
  /**
   * Hashes the order intent and generate deterministic order ID.
   * @param {SupportedChainId} chainId The CoW Protocol `chainId` context that's being used.
   * @param {Order} order order to sign
   * @param {Pick<OrderUidParams, 'owner'>} params order unique identifier parameters.
   */
  static async generateOrderId(chainId, order, params, options) {
    return generateOrderId(chainId, order, params, options);
  }
  /**
   * Get the domain separator hash for the EIP-712 typed domain data being used for signing.
   * @param chainId {SupportedChainId} chainId The CoW Protocol protocol `chainId` context that's being used.
   * @returns A string representation of the EIP-712 typed domain data hash.
   */
  static async getDomainSeparator(chainId, options) {
    const adapter = (0, import_sdk_common2.getGlobalAdapter)();
    const types = [
      { name: "name", type: "string" },
      { name: "version", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" }
    ];
    return adapter.utils.hashDomain(getDomain(chainId, options), types);
  }
  static getEIP712Types() {
    return COW_EIP712_TYPES;
  }
  /**
   * Encodes an order and ECDSA signature for EIP-1271 smart contract signature verification.
   *
   * @remarks This method encodes the order data and ECDSA signature into a format suitable for
   *          EIP-1271 signature verification by smart contracts. The order struct is ABI-encoded
   *          as a tuple along with the ECDSA signature bytes. String fields in the order are
   *          hashed using keccak256 before encoding.
   *
   * @param {UnsignedOrder} orderToSign The unsigned order to encode for EIP-1271 verification.
   * @param {string} ecdsaSignature The ECDSA signature (typically 65 bytes hex-encoded) to include in the encoding.
   * @returns {string} The ABI-encoded order struct and signature, ready for EIP-1271 verification.
   *
   * @see https://eips.ethereum.org/EIPS/eip-1271
   *
   * @example
   * ```typescript
   * const orderToSign: UnsignedOrder = { ... }
   * const ecdsaSignature = '0x...' // 65 bytes signature from signing the order
   *
   * const eip1271Signature = OrderSigningUtils.getEip1271Signature(orderToSign, ecdsaSignature)
   * // Use eip1271Signature with a smart contract wallet that implements EIP-1271
   * ```
   */
  static getEip1271Signature(orderToSign, ecdsaSignature) {
    const adapter = (0, import_sdk_common2.getGlobalAdapter)();
    const EIP712Types = COW_EIP712_TYPES[ORDER_PRIMARY_TYPE];
    const components = EIP712Types.map((component) => ({
      name: component.name,
      type: component.type === "string" ? "bytes32" : component.type
    }));
    const values = Object.values(_OrderSigningUtils.encodeUnsignedOrder(orderToSign));
    return adapter.utils.encodeAbi(
      [
        {
          components,
          type: "tuple"
        },
        { type: "bytes" }
      ],
      [values, ecdsaSignature]
    );
  }
  static encodeUnsignedOrder(orderToSign) {
    const order = {
      ...orderToSign,
      sellTokenBalance: orderToSign.sellTokenBalance ?? import_sdk_order_book2.SellTokenSource.ERC20,
      buyTokenBalance: orderToSign.buyTokenBalance ?? import_sdk_order_book2.BuyTokenDestination.ERC20
    };
    const adapter = (0, import_sdk_common2.getGlobalAdapter)();
    const EIP712Types = COW_EIP712_TYPES[ORDER_PRIMARY_TYPE];
    return EIP712Types.reduce((acc, { name, type }) => {
      const value = order[name];
      acc[name] = type === "string" ? String(adapter.utils.id(value)) : value;
      return acc;
    }, {});
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  COW_EIP712_TYPES,
  ORDER_PRIMARY_TYPE,
  OrderSigningUtils
});
